{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparing annotations of *Crepis callicephala* and *C. purpurea*\n",
    "The manually corrected genbank files are used. The annotations should be corrected before submitting to databases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnH-GUG</td>\n",
       "      <td>LSC</td>\n",
       "      <td>2</td>\n",
       "      <td>77</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>psbA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>474</td>\n",
       "      <td>1536</td>\n",
       "      <td>1062</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK</td>\n",
       "      <td>LSC</td>\n",
       "      <td>1758</td>\n",
       "      <td>4358</td>\n",
       "      <td>2600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>matK</td>\n",
       "      <td>LSC</td>\n",
       "      <td>2079</td>\n",
       "      <td>3591</td>\n",
       "      <td>1512</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK-UUU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>4320</td>\n",
       "      <td>4357</td>\n",
       "      <td>37</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>ycf2</td>\n",
       "      <td>IRA</td>\n",
       "      <td>140899</td>\n",
       "      <td>147754</td>\n",
       "      <td>6855</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU | trnM</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147865</td>\n",
       "      <td>147940</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rpl23</td>\n",
       "      <td>IRA</td>\n",
       "      <td>148104</td>\n",
       "      <td>148386</td>\n",
       "      <td>282</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rpl2</td>\n",
       "      <td>IRA</td>\n",
       "      <td>148404</td>\n",
       "      <td>149894</td>\n",
       "      <td>1490</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rps19-fragment</td>\n",
       "      <td>IRA</td>\n",
       "      <td>149952</td>\n",
       "      <td>150007</td>\n",
       "      <td>55</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>271 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism             gene region   start     end  length  \\\n",
       "0    Crepis callicephala         trnH-GUG    LSC       2      77      75   \n",
       "1    Crepis callicephala             psbA    LSC     474    1536    1062   \n",
       "2    Crepis callicephala             trnK    LSC    1758    4358    2600   \n",
       "3    Crepis callicephala             matK    LSC    2079    3591    1512   \n",
       "4    Crepis callicephala         trnK-UUU    LSC    4320    4357      37   \n",
       "..                   ...              ...    ...     ...     ...     ...   \n",
       "266      Crepis purpurea             ycf2    IRA  140899  147754    6855   \n",
       "267      Crepis purpurea  trnM-CAU | trnM    IRA  147865  147940      75   \n",
       "268      Crepis purpurea            rpl23    IRA  148104  148386     282   \n",
       "269      Crepis purpurea             rpl2    IRA  148404  149894    1490   \n",
       "270      Crepis purpurea   rps19-fragment    IRA  149952  150007      55   \n",
       "\n",
       "     spanned  \n",
       "0      False  \n",
       "1      False  \n",
       "2      False  \n",
       "3      False  \n",
       "4      False  \n",
       "..       ...  \n",
       "266    False  \n",
       "267    False  \n",
       "268    False  \n",
       "269    False  \n",
       "270     True  \n",
       "\n",
       "[271 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO, SeqFeature\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/in/C_callicephala_curated_1_cleaned_meta.gb\"\n",
    "c_pur = \"plastomes/in/Crepis_purpurea_edited+.gb\"\n",
    "\n",
    "regions = {}\n",
    "dataset = []\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in c_cal, c_pur:\n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            features = record.features\n",
    "            # check if there features other than 'source'\n",
    "            if len(features) > 1:\n",
    "                for feature in features:\n",
    "                    # check for genome regions\n",
    "                    if feature.type in ['misc_feature', 'repeat_region']:\n",
    "                        label = feature.qualifiers.get('note', \"\")[0].split()[-1].strip(\"()\")\n",
    "                        regions[label] = [int(feature.location.start), int(feature.location.end)]\n",
    "                # filling the dataframe\n",
    "                organism = record.annotations.get('organism', \"\")\n",
    "                for feature in features:\n",
    "                    if feature.type == 'gene':\n",
    "                        gene = feature.qualifiers.get('gene', \"\")\n",
    "                        start = int(feature.location.start)\n",
    "                        end = int(feature.location.end)\n",
    "\n",
    "                        for region in regions.items():\n",
    "                            ranges = region[-1]\n",
    "                            if ranges[0] <= start < ranges[-1]:\n",
    "                                #print(f'{region[0]}\\t{gene}')\n",
    "                                spanned = False\n",
    "                                if end >= ranges[-1]:\n",
    "                                    spanned = True\n",
    "                                dataset.append((organism, gene[0], region[0], start, end, len(feature), spanned))\n",
    "\n",
    "df = pd.DataFrame(dataset, columns=['organism', 'gene', 'region', 'start', 'end', 'length', 'spanned'])\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comparing repeated features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>organism</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Crepis callicephala</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>region</th>\n",
       "      <th>IRA</th>\n",
       "      <th>IRB</th>\n",
       "      <th>IRA</th>\n",
       "      <th>IRB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ndhB</th>\n",
       "      <td>2202.0</td>\n",
       "      <td>2202.0</td>\n",
       "      <td>2202.0</td>\n",
       "      <td>2202.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ndhF</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2226.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2226.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl2</th>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl23</th>\n",
       "      <td>282.0</td>\n",
       "      <td>282.0</td>\n",
       "      <td>282.0</td>\n",
       "      <td>282.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps19-fragment</th>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps7</th>\n",
       "      <td>468.0</td>\n",
       "      <td>468.0</td>\n",
       "      <td>468.0</td>\n",
       "      <td>468.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn16</th>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn23</th>\n",
       "      <td>2810.0</td>\n",
       "      <td>2810.0</td>\n",
       "      <td>2810.0</td>\n",
       "      <td>2810.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn4.5</th>\n",
       "      <td>103.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>103.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>116.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn5S</th>\n",
       "      <td>122.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>122.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnA-UGC</th>\n",
       "      <td>893.0</td>\n",
       "      <td>893.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnA-UGC | trnA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>893.0</td>\n",
       "      <td>893.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnI-GAU</th>\n",
       "      <td>834.0</td>\n",
       "      <td>834.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnI-GAU | trnE-UUC | trnE | trnI</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>836.0</td>\n",
       "      <td>836.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL-CAA</th>\n",
       "      <td>81.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL-CAA | trnL-fragment</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>83.0</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnM</th>\n",
       "      <td>75.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnM-CAU</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnM-CAU | trnM</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnN-GUU</th>\n",
       "      <td>72.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnR-ACG</th>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnR-ACG | trnR</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnV-GAC</th>\n",
       "      <td>73.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnV-GAC | trnV</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>480.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf1 fragment</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>480.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf15</th>\n",
       "      <td>192.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf2</th>\n",
       "      <td>6855.0</td>\n",
       "      <td>6855.0</td>\n",
       "      <td>6855.0</td>\n",
       "      <td>6855.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism                          Crepis callicephala         Crepis purpurea  \\\n",
       "region                                            IRA     IRB             IRA   \n",
       "gene                                                                            \n",
       "ndhB                                           2202.0  2202.0          2202.0   \n",
       "ndhF                                              NaN  2226.0             NaN   \n",
       "rpl2                                           1490.0  1490.0          1490.0   \n",
       "rpl23                                           282.0   282.0           282.0   \n",
       "rps19-fragment                                   60.0     NaN            55.0   \n",
       "rps7                                            468.0   468.0           468.0   \n",
       "rrn16                                          1490.0  1490.0          1490.0   \n",
       "rrn23                                          2810.0  2810.0          2810.0   \n",
       "rrn4.5                                          103.0   103.0           103.0   \n",
       "rrn5                                              NaN   116.0           121.0   \n",
       "rrn5S                                           122.0     NaN             NaN   \n",
       "trnA-UGC                                        893.0   893.0             NaN   \n",
       "trnA-UGC | trnA                                   NaN     NaN           893.0   \n",
       "trnI-GAU                                        834.0   834.0             NaN   \n",
       "trnI-GAU | trnE-UUC | trnE | trnI                 NaN     NaN           836.0   \n",
       "trnL-CAA                                         81.0    81.0             NaN   \n",
       "trnL-CAA | trnL-fragment                          NaN     NaN            83.0   \n",
       "trnM                                             75.0    74.0             NaN   \n",
       "trnM-CAU                                          NaN     NaN             NaN   \n",
       "trnM-CAU | trnM                                   NaN     NaN            75.0   \n",
       "trnN-GUU                                         72.0    72.0            74.0   \n",
       "trnR-ACG                                         74.0    74.0            75.0   \n",
       "trnR-ACG | trnR                                   NaN     NaN             NaN   \n",
       "trnV-GAC                                         73.0    72.0            72.0   \n",
       "trnV-GAC | trnV                                   NaN     NaN             NaN   \n",
       "ycf1                                              NaN   480.0             NaN   \n",
       "ycf1 fragment                                     NaN     NaN             NaN   \n",
       "ycf15                                           192.0   192.0           192.0   \n",
       "ycf2                                           6855.0  6855.0          6855.0   \n",
       "\n",
       "organism                                   \n",
       "region                                IRB  \n",
       "gene                                       \n",
       "ndhB                               2202.0  \n",
       "ndhF                               2226.0  \n",
       "rpl2                               1490.0  \n",
       "rpl23                               282.0  \n",
       "rps19-fragment                        NaN  \n",
       "rps7                                468.0  \n",
       "rrn16                              1490.0  \n",
       "rrn23                              2810.0  \n",
       "rrn4.5                              103.0  \n",
       "rrn5                                  NaN  \n",
       "rrn5S                               122.0  \n",
       "trnA-UGC                              NaN  \n",
       "trnA-UGC | trnA                     893.0  \n",
       "trnI-GAU                              NaN  \n",
       "trnI-GAU | trnE-UUC | trnE | trnI   836.0  \n",
       "trnL-CAA                              NaN  \n",
       "trnL-CAA | trnL-fragment             83.0  \n",
       "trnM                                  NaN  \n",
       "trnM-CAU                             75.0  \n",
       "trnM-CAU | trnM                       NaN  \n",
       "trnN-GUU                             74.0  \n",
       "trnR-ACG                              NaN  \n",
       "trnR-ACG | trnR                      75.0  \n",
       "trnV-GAC                              NaN  \n",
       "trnV-GAC | trnV                      72.0  \n",
       "ycf1                                  NaN  \n",
       "ycf1 fragment                       480.0  \n",
       "ycf15                               192.0  \n",
       "ycf2                               6855.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_repeats = df[df['region'].isin(['IRA', 'IRB'])]\n",
    "\n",
    "pivot_table = df_repeats.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism', 'region'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Reorder levels in the columns\n",
    "#pivot_table = df2.reorder_levels([1, 0], axis=1)\n",
    "\n",
    "# Sort the columns\n",
    "pivot_table = pivot_table.sort_index(axis=1)\n",
    "\n",
    "# Display the pivot table\n",
    "display(pivot_table)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare features within LSC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with different values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>organism</th>\n",
       "      <th>Crepis callicephala</th>\n",
       "      <th>Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>petD</th>\n",
       "      <td>525</td>\n",
       "      <td>1183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>psaA</th>\n",
       "      <td>2253</td>\n",
       "      <td>1776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl16</th>\n",
       "      <td>1445</td>\n",
       "      <td>1440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl22</th>\n",
       "      <td>465</td>\n",
       "      <td>540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC1</th>\n",
       "      <td>2799</td>\n",
       "      <td>2520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC2</th>\n",
       "      <td>4125</td>\n",
       "      <td>2274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps16</th>\n",
       "      <td>1120</td>\n",
       "      <td>1125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism  Crepis callicephala  Crepis purpurea\n",
       "gene                                          \n",
       "petD                      525             1183\n",
       "psaA                     2253             1776\n",
       "rpl16                    1445             1440\n",
       "rpl22                     465              540\n",
       "rpoC1                    2799             2520\n",
       "rpoC2                    4125             2274\n",
       "rps16                    1120             1125"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA genes with different values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>organism</th>\n",
       "      <th>Crepis callicephala</th>\n",
       "      <th>Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>trnC-GCA</th>\n",
       "      <td>72.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnC-GCA | trnC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnD-GUC</th>\n",
       "      <td>74.0</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnE-UUC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnE-UUC | trnE</th>\n",
       "      <td>72.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnF-GAA</th>\n",
       "      <td>70.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnF-GAA | trnF</th>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnG-GCC | trnG</th>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnG-GCC | trnG-UCC</th>\n",
       "      <td>71.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnG-UCC</th>\n",
       "      <td>781.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnK</th>\n",
       "      <td>2600.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnK-UUU</th>\n",
       "      <td>37.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnK-UUU | trnK</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2600.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL-UAA</th>\n",
       "      <td>524.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL-UAA | trnL</th>\n",
       "      <td>NaN</td>\n",
       "      <td>525.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnM-CAU</th>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnM-CAU | trnM</th>\n",
       "      <td>NaN</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnP-UGG</th>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnP-UGG | trnP</th>\n",
       "      <td>NaN</td>\n",
       "      <td>76.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnQ-UUG</th>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnQ-UUG | trnQ</th>\n",
       "      <td>73.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnS-CGA | trnG | trnS-SER | trnS</th>\n",
       "      <td>NaN</td>\n",
       "      <td>782.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnS-GGA</th>\n",
       "      <td>87.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnS-GGA | trnS</th>\n",
       "      <td>NaN</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnS-UGA</th>\n",
       "      <td>83.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnS-UGA | trnS</th>\n",
       "      <td>NaN</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnT-GGU</th>\n",
       "      <td>72.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnT-GGU | trnT-fragment</th>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnT-UGU</th>\n",
       "      <td>73.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnV-UAC</th>\n",
       "      <td>643.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnV-UAC | trnV</th>\n",
       "      <td>NaN</td>\n",
       "      <td>643.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnW-CCA</th>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnW-CCA | trnW</th>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnY-GUA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnY-GUA | trnY</th>\n",
       "      <td>84.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnfM-CAU | trnM-CAU | trnM</th>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism                           Crepis callicephala  Crepis purpurea\n",
       "gene                                                                   \n",
       "trnC-GCA                                          72.0              NaN\n",
       "trnC-GCA | trnC                                    NaN             72.0\n",
       "trnD-GUC                                          74.0             77.0\n",
       "trnE-UUC                                           NaN             74.0\n",
       "trnE-UUC | trnE                                   72.0              NaN\n",
       "trnF-GAA                                          70.0              NaN\n",
       "trnF-GAA | trnF                                    NaN             74.0\n",
       "trnG-GCC | trnG                                    NaN             72.0\n",
       "trnG-GCC | trnG-UCC                               71.0              NaN\n",
       "trnG-UCC                                         781.0              NaN\n",
       "trnK                                            2600.0              NaN\n",
       "trnK-UUU                                          37.0              NaN\n",
       "trnK-UUU | trnK                                    NaN           2600.0\n",
       "trnL-UAA                                         524.0              NaN\n",
       "trnL-UAA | trnL                                    NaN            525.0\n",
       "trnM-CAU                                          74.0              NaN\n",
       "trnM-CAU | trnM                                    NaN             73.0\n",
       "trnP-UGG                                          75.0              NaN\n",
       "trnP-UGG | trnP                                    NaN             76.0\n",
       "trnQ-UUG                                           NaN             72.0\n",
       "trnQ-UUG | trnQ                                   73.0              NaN\n",
       "trnS-CGA | trnG | trnS-SER | trnS                  NaN            782.0\n",
       "trnS-GGA                                          87.0              NaN\n",
       "trnS-GGA | trnS                                    NaN             87.0\n",
       "trnS-UGA                                          83.0              NaN\n",
       "trnS-UGA | trnS                                    NaN             83.0\n",
       "trnT-GGU                                          72.0              NaN\n",
       "trnT-GGU | trnT-fragment                           NaN             72.0\n",
       "trnT-UGU                                          73.0             75.0\n",
       "trnV-UAC                                         643.0              NaN\n",
       "trnV-UAC | trnV                                    NaN            643.0\n",
       "trnW-CCA                                          74.0              NaN\n",
       "trnW-CCA | trnW                                    NaN             74.0\n",
       "trnY-GUA                                           NaN             87.0\n",
       "trnY-GUA | trnY                                   84.0              NaN\n",
       "trnfM-CAU | trnM-CAU | trnM                        NaN             74.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_target = df[df['region'].isin(['LSC'])]\n",
    "\n",
    "non_tRNA = df_target[~df_target['gene'].str.startswith('trn')]\n",
    "tRNA = df_target[df_target['gene'].str.startswith('trn')]\n",
    "\n",
    "pt_tRNA = tRNA.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Sort the columns\n",
    "pt_tRNA = pt_tRNA.sort_index(axis=1)\n",
    "\n",
    "pt_nontRNA = non_tRNA.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Sort the columns\n",
    "pt_nontRNA = pt_nontRNA.sort_index(axis=1)\n",
    "\n",
    "\n",
    "# Display the pivot table\n",
    "#display(pt_tRNA, pt_nontRNA)\n",
    "#with pd.option_context('display.max_rows', None,):\n",
    "#    display(pt_nontRNA)\n",
    "\n",
    "# Find differences\n",
    "diff_nontRNA = pt_nontRNA['Crepis callicephala'] != pt_nontRNA['Crepis purpurea']\n",
    "print(\"Genes with different values:\")\n",
    "display(pt_nontRNA[diff_nontRNA])\n",
    "\n",
    "diff_tRNA = pt_tRNA['Crepis callicephala'] != pt_tRNA['Crepis purpurea']\n",
    "print(\"tRNA genes with different values:\")\n",
    "display(pt_tRNA[diff_tRNA])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SSC features with different values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with different values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>organism</th>\n",
       "      <th>Crepis callicephala</th>\n",
       "      <th>Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ndhA</th>\n",
       "      <td>2147</td>\n",
       "      <td>2146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ndhD</th>\n",
       "      <td>1503</td>\n",
       "      <td>1521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf1</th>\n",
       "      <td>4970</td>\n",
       "      <td>4973</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism  Crepis callicephala  Crepis purpurea\n",
       "gene                                          \n",
       "ndhA                     2147             2146\n",
       "ndhD                     1503             1521\n",
       "ycf1                     4970             4973"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA genes with different values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>organism</th>\n",
       "      <th>Crepis callicephala</th>\n",
       "      <th>Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>trnL-UAG</th>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL-UAG | trnL</th>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism         Crepis callicephala  Crepis purpurea\n",
       "gene                                                 \n",
       "trnL-UAG                        80.0              NaN\n",
       "trnL-UAG | trnL                  NaN             80.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_target = df[df['region'].isin(['SSC'])]\n",
    "\n",
    "non_tRNA = df_target[~df_target['gene'].str.startswith('trn')]\n",
    "tRNA = df_target[df_target['gene'].str.startswith('trn')]\n",
    "\n",
    "pt_tRNA = tRNA.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Sort the columns\n",
    "pt_tRNA = pt_tRNA.sort_index(axis=1)\n",
    "\n",
    "pt_nontRNA = non_tRNA.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Sort the columns\n",
    "pt_nontRNA = pt_nontRNA.sort_index(axis=1)\n",
    "\n",
    "\n",
    "# Display the pivot table\n",
    "#display(pt_tRNA, pt_nontRNA)\n",
    "#with pd.option_context('display.max_rows', None,):\n",
    "#    display(pt_nontRNA)\n",
    "\n",
    "# Find differences\n",
    "diff_nontRNA = pt_nontRNA['Crepis callicephala'] != pt_nontRNA['Crepis purpurea']\n",
    "print(\"Genes with different values:\")\n",
    "display(pt_nontRNA[diff_nontRNA])\n",
    "\n",
    "diff_tRNA = pt_tRNA['Crepis callicephala'] != pt_tRNA['Crepis purpurea']\n",
    "print(\"tRNA genes with different values:\")\n",
    "display(pt_tRNA[diff_tRNA])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_trna_labels(in_label: str):\n",
    "    label = None\n",
    "    #print(\"initial:\", in_label)\n",
    "    if \"trn\" in str(in_label).lower():\n",
    "        if not \" | \" in str(in_label):\n",
    "            label = in_label\n",
    "        else:\n",
    "            labels = in_label.split(\" | \")\n",
    "            triplet_flag = False\n",
    "            fragment_flag = False\n",
    "            marked_by_triplet = []\n",
    "            for i, x in enumerate(labels):\n",
    "                #print(i, \"x:\", x)\n",
    "                if len(x.split(\"-\")[-1]) == 3:\n",
    "                    triplet_flag = True\n",
    "                    marked_by_triplet.append(x)\n",
    "                if \"fragment\" in x:\n",
    "                    fragment_flag = True\n",
    "            if len(marked_by_triplet) == 1:\n",
    "                label = labels[0]\n",
    "            else:\n",
    "                concise = []\n",
    "                for idx, y in enumerate(marked_by_triplet):\n",
    "                    #print(idx, \"y\", y)\n",
    "                    if len(y.split(\"-\")[0]) > 4:\n",
    "                        pass\n",
    "                    else:\n",
    "                        concise.append(y)\n",
    "                if len(concise) == 1:\n",
    "                    label = concise[0]\n",
    "                else:\n",
    "                    #print(f'Manual checking needed: {concise}')\n",
    "                    # fixing case of trnS-SER\n",
    "                    for z, a in enumerate(concise):\n",
    "                        #print(z, \"a:\", a)\n",
    "                        suf = a.split(\"-\")[-1]\n",
    "                        for b in suf:\n",
    "                            if not b.upper() in \"ATGCU\":\n",
    "                                #print(\"omitted:\", b, a)\n",
    "                                pass\n",
    "                            else:\n",
    "                                # ensure assigning the first instanse \n",
    "                                # from the concise list to label var\n",
    "                                if not label == None:\n",
    "                                    pass\n",
    "                                else:\n",
    "                                    label = a\n",
    "                                    break\n",
    "\n",
    "            if fragment_flag == True:\n",
    "                if not \"fragment\" in label:\n",
    "                    label = f'{label}-fragment'\n",
    "\n",
    "        #print(label)\n",
    "        return label\n",
    "    \n",
    "    else:\n",
    "        return in_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trnI-GAU\n"
     ]
    }
   ],
   "source": [
    "print(clean_trna_labels(\"trnI-GAU | trnE-UUC | trnE | trnI\t\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rra4.5S\n"
     ]
    }
   ],
   "source": [
    "def normalize_rrna_label(in_label: str):\n",
    "    letters = list(in_label)\n",
    "    terminal = letters[-1]\n",
    "    if terminal.upper() == \"S\" and \"rrn\" in in_label:\n",
    "        label = in_label[:-1]\n",
    "    else:\n",
    "        label = in_label\n",
    "    \n",
    "    return label\n",
    "\n",
    "print(normalize_rrna_label(\"rra4.5S\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "prs19 fragment\n"
     ]
    }
   ],
   "source": [
    "def normalize_fragment_label(in_label: str):\n",
    "    label = None\n",
    "    if \"fragment\" in in_label:\n",
    "        if \"-fragment\" in in_label:\n",
    "            shorten = in_label.split(\"-fragment\")[0]\n",
    "            label = f'{shorten} fragment'\n",
    "        elif not \" \" in in_label:\n",
    "            shorten = in_label.split(\"fragment\")[0]\n",
    "            label = f'{shorten} fragment'\n",
    "        else:\n",
    "            label = in_label\n",
    "    else:\n",
    "        label = in_label\n",
    "\n",
    "    return label\n",
    "\n",
    "print(normalize_fragment_label(\"prs19 fragment\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ycf1\n"
     ]
    }
   ],
   "source": [
    "def normalize_fragment_ycf(in_label: str, in_region: str):\n",
    "    if not \"5\" in in_label and \"ycf\" in in_label:\n",
    "        if in_region == \"SSC\" or in_region == \"LSC\":\n",
    "            label = in_label\n",
    "        else:\n",
    "            if \"fragment\" in in_label:\n",
    "                label = normalize_fragment_label(in_label)\n",
    "            elif \"ycf1\" in in_label:\n",
    "                label = f'{in_label} fragment'\n",
    "            else:\n",
    "                label = in_label\n",
    "    else:\n",
    "        label = in_label\n",
    "\n",
    "    return label\n",
    "\n",
    "test = \"ycf1\"\n",
    "region = \"SSC\"\n",
    "print(normalize_fragment_ycf(test, region))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalize labels.\n",
    "#### Iteration 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trnT-GGU\n"
     ]
    }
   ],
   "source": [
    "def normalize_labels(in_label: str | None = None, region: str | None = None, exception: str | list | None = None) -> str:\n",
    "    label = None\n",
    "    #if \"trn\" in in_label:\n",
    "    #    label = clean_trna_labels(in_label)\n",
    "    #elif \"rrn\" in in_label:\n",
    "    #    label = normalize_rrna_label(in_label)\n",
    "    #elif \"fragment\" in in_label:\n",
    "    #    label = normalize_fragment_label(in_label)\n",
    "    #elif \"ycf\" in in_label:\n",
    "    #    label = normalize_fragment_ycf(in_label, region)\n",
    "    label = clean_trna_labels(in_label)\n",
    "    label = normalize_rrna_label(label)\n",
    "    label = normalize_fragment_label(label)\n",
    "    label = normalize_fragment_ycf(label, region)\n",
    "\n",
    "    if not exception:\n",
    "        pass\n",
    "    else:\n",
    "        if type(exception) == list:\n",
    "            for exp in exception:\n",
    "                if exp == in_label or exp == label:\n",
    "                    label = label.split(\" fragment\")[0]\n",
    "                else:\n",
    "                    pass\n",
    "        elif type(exception) == str:\n",
    "            exp = exception\n",
    "            if exp == in_label or exp == label:\n",
    "                label = label.split(\" fragment\")[0]\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "             pass\n",
    "\n",
    "\n",
    "    if not label:\n",
    "        return in_label\n",
    "    else:\n",
    "        return label\n",
    "    \n",
    "#test = \"trnS-CGA | trnG | trnS-SER | trnS\"\n",
    "test = \"trnT-GGU | trnT-fragment\"\n",
    "region = \"IRB\"\n",
    "exception = [\"trnL-CAA fragment\", \"trnT-GGU | trnT-fragment\"]\n",
    "print(normalize_labels(test, region, exception))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnH-GUG</td>\n",
       "      <td>LSC</td>\n",
       "      <td>2</td>\n",
       "      <td>77</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>psbA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>474</td>\n",
       "      <td>1536</td>\n",
       "      <td>1062</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK</td>\n",
       "      <td>LSC</td>\n",
       "      <td>1758</td>\n",
       "      <td>4358</td>\n",
       "      <td>2600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>matK</td>\n",
       "      <td>LSC</td>\n",
       "      <td>2079</td>\n",
       "      <td>3591</td>\n",
       "      <td>1512</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK-UUU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>4320</td>\n",
       "      <td>4357</td>\n",
       "      <td>37</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>ycf2</td>\n",
       "      <td>IRA</td>\n",
       "      <td>140899</td>\n",
       "      <td>147754</td>\n",
       "      <td>6855</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147865</td>\n",
       "      <td>147940</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rpl23</td>\n",
       "      <td>IRA</td>\n",
       "      <td>148104</td>\n",
       "      <td>148386</td>\n",
       "      <td>282</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rpl2</td>\n",
       "      <td>IRA</td>\n",
       "      <td>148404</td>\n",
       "      <td>149894</td>\n",
       "      <td>1490</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rps19 fragment</td>\n",
       "      <td>IRA</td>\n",
       "      <td>149952</td>\n",
       "      <td>150007</td>\n",
       "      <td>55</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>271 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism            gene region   start     end  length  \\\n",
       "0    Crepis callicephala        trnH-GUG    LSC       2      77      75   \n",
       "1    Crepis callicephala            psbA    LSC     474    1536    1062   \n",
       "2    Crepis callicephala            trnK    LSC    1758    4358    2600   \n",
       "3    Crepis callicephala            matK    LSC    2079    3591    1512   \n",
       "4    Crepis callicephala        trnK-UUU    LSC    4320    4357      37   \n",
       "..                   ...             ...    ...     ...     ...     ...   \n",
       "266      Crepis purpurea            ycf2    IRA  140899  147754    6855   \n",
       "267      Crepis purpurea        trnM-CAU    IRA  147865  147940      75   \n",
       "268      Crepis purpurea           rpl23    IRA  148104  148386     282   \n",
       "269      Crepis purpurea            rpl2    IRA  148404  149894    1490   \n",
       "270      Crepis purpurea  rps19 fragment    IRA  149952  150007      55   \n",
       "\n",
       "     spanned  \n",
       "0      False  \n",
       "1      False  \n",
       "2      False  \n",
       "3      False  \n",
       "4      False  \n",
       "..       ...  \n",
       "266    False  \n",
       "267    False  \n",
       "268    False  \n",
       "269    False  \n",
       "270     True  \n",
       "\n",
       "[271 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO, SeqFeature\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/in/C_callicephala_curated_1_cleaned_meta.gb\"\n",
    "c_pur = \"plastomes/in/Crepis_purpurea_edited+.gb\"\n",
    "\n",
    "exception = [\n",
    "    \"trnL-CAA fragment\",\n",
    "    \"trnT-GGU | trnT-fragment\",\n",
    "]\n",
    "\n",
    "regions = {}\n",
    "dataset = []\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in c_cal, c_pur:\n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            features = record.features\n",
    "            # check if there features other than 'source'\n",
    "            if len(features) > 1:\n",
    "                for feature in features:\n",
    "                    # check for genome regions\n",
    "                    if feature.type in ['misc_feature', 'repeat_region']:\n",
    "                        label = feature.qualifiers.get('note', \"\")[0].split()[-1].strip(\"()\")\n",
    "                        regions[label] = [int(feature.location.start), int(feature.location.end)]\n",
    "                # filling the dataframe\n",
    "                organism = record.annotations.get('organism', \"\")\n",
    "                for feature in features:\n",
    "                    if feature.type == 'gene':\n",
    "                        gene = feature.qualifiers.get('gene', \"\")\n",
    "                        start = int(feature.location.start)\n",
    "                        end = int(feature.location.end)\n",
    "\n",
    "                        for region in regions.items():\n",
    "                            ranges = region[-1]\n",
    "                            if ranges[0] <= start < ranges[-1]:\n",
    "                                #print(f'{region[0]}\\t{gene}')\n",
    "                                spanned = False\n",
    "                                if end >= ranges[-1]:\n",
    "                                    spanned = True\n",
    "                                dataset.append((organism, normalize_labels(gene[0], region[0], exception), region[0], start, end, len(feature), spanned))\n",
    "\n",
    "df = pd.DataFrame(dataset, columns=['organism', 'gene', 'region', 'start', 'end', 'length', 'spanned'])\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge synonymous labeles\n",
    "#### Checking label to merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK</td>\n",
       "      <td>LSC</td>\n",
       "      <td>1758</td>\n",
       "      <td>4358</td>\n",
       "      <td>2600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnG-UCC</td>\n",
       "      <td>LSC</td>\n",
       "      <td>30094</td>\n",
       "      <td>30875</td>\n",
       "      <td>781</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM</td>\n",
       "      <td>IRB</td>\n",
       "      <td>83887</td>\n",
       "      <td>83961</td>\n",
       "      <td>74</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147833</td>\n",
       "      <td>147908</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnS-CGA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>30129</td>\n",
       "      <td>30911</td>\n",
       "      <td>782</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRB</td>\n",
       "      <td>83900</td>\n",
       "      <td>83975</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147865</td>\n",
       "      <td>147940</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism      gene region   start     end  length  spanned\n",
       "2    Crepis callicephala      trnK    LSC    1758    4358    2600    False\n",
       "25   Crepis callicephala  trnG-UCC    LSC   30094   30875     781    False\n",
       "88   Crepis callicephala      trnM    IRB   83887   83961      74    False\n",
       "132  Crepis callicephala      trnM    IRA  147833  147908      75    False\n",
       "160      Crepis purpurea  trnS-CGA    LSC   30129   30911     782    False\n",
       "223      Crepis purpurea  trnM-CAU    IRB   83900   83975      75    False\n",
       "267      Crepis purpurea  trnM-CAU    IRA  147865  147940      75    False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK</td>\n",
       "      <td>LSC</td>\n",
       "      <td>1758</td>\n",
       "      <td>4358</td>\n",
       "      <td>2600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK-UUU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>4320</td>\n",
       "      <td>4357</td>\n",
       "      <td>37</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnK-UUU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>1760</td>\n",
       "      <td>4360</td>\n",
       "      <td>2600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism      gene region  start   end  length  spanned\n",
       "2    Crepis callicephala      trnK    LSC   1758  4358    2600    False\n",
       "4    Crepis callicephala  trnK-UUU    LSC   4320  4357      37    False\n",
       "138      Crepis purpurea  trnK-UUU    LSC   1760  4360    2600    False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>36111</td>\n",
       "      <td>36185</td>\n",
       "      <td>74</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>51913</td>\n",
       "      <td>51986</td>\n",
       "      <td>73</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM</td>\n",
       "      <td>IRB</td>\n",
       "      <td>83887</td>\n",
       "      <td>83961</td>\n",
       "      <td>74</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147833</td>\n",
       "      <td>147908</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>36147</td>\n",
       "      <td>36221</td>\n",
       "      <td>74</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>51948</td>\n",
       "      <td>52021</td>\n",
       "      <td>73</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRB</td>\n",
       "      <td>83900</td>\n",
       "      <td>83975</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147865</td>\n",
       "      <td>147940</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism      gene region   start     end  length  spanned\n",
       "32   Crepis callicephala  trnM-CAU    LSC   36111   36185      74    False\n",
       "47   Crepis callicephala  trnM-CAU    LSC   51913   51986      73    False\n",
       "88   Crepis callicephala      trnM    IRB   83887   83961      74    False\n",
       "132  Crepis callicephala      trnM    IRA  147833  147908      75    False\n",
       "167      Crepis purpurea  trnM-CAU    LSC   36147   36221      74    False\n",
       "182      Crepis purpurea  trnM-CAU    LSC   51948   52021      73    False\n",
       "223      Crepis purpurea  trnM-CAU    IRB   83900   83975      75    False\n",
       "267      Crepis purpurea  trnM-CAU    IRA  147865  147940      75    False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnS-GCU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>8489</td>\n",
       "      <td>8577</td>\n",
       "      <td>88</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnS-UGA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>34946</td>\n",
       "      <td>35029</td>\n",
       "      <td>83</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnS-GGA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>44777</td>\n",
       "      <td>44864</td>\n",
       "      <td>87</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnS-GCU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>8489</td>\n",
       "      <td>8577</td>\n",
       "      <td>88</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnS-CGA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>30129</td>\n",
       "      <td>30911</td>\n",
       "      <td>782</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnS-UGA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>34982</td>\n",
       "      <td>35065</td>\n",
       "      <td>83</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnS-GGA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>44812</td>\n",
       "      <td>44899</td>\n",
       "      <td>87</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism      gene region  start    end  length  spanned\n",
       "9    Crepis callicephala  trnS-GCU    LSC   8489   8577      88    False\n",
       "29   Crepis callicephala  trnS-UGA    LSC  34946  35029      83    False\n",
       "37   Crepis callicephala  trnS-GGA    LSC  44777  44864      87    False\n",
       "144      Crepis purpurea  trnS-GCU    LSC   8489   8577      88    False\n",
       "160      Crepis purpurea  trnS-CGA    LSC  30129  30911     782    False\n",
       "164      Crepis purpurea  trnS-UGA    LSC  34982  35065      83    False\n",
       "172      Crepis purpurea  trnS-GGA    LSC  44812  44899      87    False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnG-UCC</td>\n",
       "      <td>LSC</td>\n",
       "      <td>30094</td>\n",
       "      <td>30875</td>\n",
       "      <td>781</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnG-GCC</td>\n",
       "      <td>LSC</td>\n",
       "      <td>35854</td>\n",
       "      <td>35925</td>\n",
       "      <td>71</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnG-GCC</td>\n",
       "      <td>LSC</td>\n",
       "      <td>35890</td>\n",
       "      <td>35962</td>\n",
       "      <td>72</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism      gene region  start    end  length  spanned\n",
       "25   Crepis callicephala  trnG-UCC    LSC  30094  30875     781    False\n",
       "31   Crepis callicephala  trnG-GCC    LSC  35854  35925      71    False\n",
       "166      Crepis purpurea  trnG-GCC    LSC  35890  35962      72    False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnR-UCU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>29817</td>\n",
       "      <td>29889</td>\n",
       "      <td>72</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnG-UCC</td>\n",
       "      <td>LSC</td>\n",
       "      <td>30094</td>\n",
       "      <td>30875</td>\n",
       "      <td>781</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnT-GGU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>31010</td>\n",
       "      <td>31082</td>\n",
       "      <td>72</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>psbD</td>\n",
       "      <td>LSC</td>\n",
       "      <td>32293</td>\n",
       "      <td>33355</td>\n",
       "      <td>1062</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>psbC</td>\n",
       "      <td>LSC</td>\n",
       "      <td>33302</td>\n",
       "      <td>34724</td>\n",
       "      <td>1422</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnR-UCU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>29853</td>\n",
       "      <td>29925</td>\n",
       "      <td>72</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnS-CGA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>30129</td>\n",
       "      <td>30911</td>\n",
       "      <td>782</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnT-GGU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>31046</td>\n",
       "      <td>31118</td>\n",
       "      <td>72</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>psbD</td>\n",
       "      <td>LSC</td>\n",
       "      <td>32329</td>\n",
       "      <td>33391</td>\n",
       "      <td>1062</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>psbC</td>\n",
       "      <td>LSC</td>\n",
       "      <td>33338</td>\n",
       "      <td>34760</td>\n",
       "      <td>1422</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism      gene region  start    end  length  spanned\n",
       "24   Crepis callicephala  trnR-UCU    LSC  29817  29889      72    False\n",
       "25   Crepis callicephala  trnG-UCC    LSC  30094  30875     781    False\n",
       "26   Crepis callicephala  trnT-GGU    LSC  31010  31082      72    False\n",
       "27   Crepis callicephala      psbD    LSC  32293  33355    1062    False\n",
       "28   Crepis callicephala      psbC    LSC  33302  34724    1422    False\n",
       "159      Crepis purpurea  trnR-UCU    LSC  29853  29925      72    False\n",
       "160      Crepis purpurea  trnS-CGA    LSC  30129  30911     782    False\n",
       "161      Crepis purpurea  trnT-GGU    LSC  31046  31118      72    False\n",
       "162      Crepis purpurea      psbD    LSC  32329  33391    1062    False\n",
       "163      Crepis purpurea      psbC    LSC  33338  34760    1422    False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "organisms = df['organism'].unique()\n",
    "\n",
    "differed = df.drop_duplicates(subset=['gene', 'region'], keep=False)\n",
    "display(differed)\n",
    "\n",
    "display(df[df['gene'].str.slice(0,4) == 'trnK'])\n",
    "display(df[df['gene'].str.slice(0,4) == 'trnM'])\n",
    "display(df[df['gene'].str.slice(0,4) == 'trnS'])\n",
    "display(df[df['gene'].str.slice(0,4) == 'trnG'])\n",
    "display(df[df['start'].between(29800, 34000)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Conclusion:\n",
    "- 'trnK'-case: the additional instance of the trnK-UUU is due to the mistake on annottation cleaning - an alternative tRNA gene annotation was left in one of the exones. It should be removed manually or programmatically.\n",
    "- 'trnG-trnS'-case reflects the annotation problem of the same sequence. Which is correct is difficult to judge.\n",
    "- 'trnM'-case: trnM within the IR regions should be renamed to 'trnM-CAU'."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Another approach (nighbor matching)\n",
    "\n",
    "Tested, but the problem seems to be easily resolved by only two corrections."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Crepis callicephala' 'Crepis purpurea']\n",
      "0 trnH-GUG\n",
      "None psbA\n",
      "1 psbA\n",
      "2 trnK\n",
      "psbA matK\n",
      "3 matK\n",
      "4 trnK-UUU\n",
      "matK rps16\n",
      "5 rps16\n",
      "6 trnQ-UUG\n",
      "rps16 psbK\n",
      "7 psbK\n",
      "8 psbI\n",
      "9 trnS-GCU\n",
      "psbI trnC-GCA\n",
      "10 trnC-GCA\n",
      "trnS-GCU petN\n",
      "11 petN\n",
      "12 psbM\n",
      "13 trnD-GUC\n",
      "psbM trnY-GUA\n",
      "14 trnY-GUA\n",
      "trnD-GUC trnE-UUC\n",
      "15 trnE-UUC\n",
      "trnY-GUA rpoB\n",
      "16 rpoB\n",
      "17 rpoC1\n",
      "18 rpoC2\n",
      "19 rps2\n",
      "20 atpI\n",
      "21 atpH\n",
      "22 atpF\n",
      "23 atpA\n",
      "24 trnR-UCU\n",
      "atpA trnG-UCC\n",
      "25 trnG-UCC\n",
      "trnR-UCU trnT-GGU\n",
      "26 trnT-GGU\n",
      "trnG-UCC psbD\n",
      "27 psbD\n",
      "28 psbC\n",
      "29 trnS-UGA\n",
      "psbC psbZ\n",
      "30 psbZ\n",
      "31 trnG-GCC\n",
      "psbZ trnM-CAU\n",
      "32 trnM-CAU\n",
      "trnG-GCC rps14\n",
      "33 rps14\n",
      "34 psaB\n",
      "35 psaA\n",
      "36 pafI\n",
      "37 trnS-GGA\n",
      "pafI rps4\n",
      "38 rps4\n",
      "39 trnT-UGU\n",
      "rps4 trnL-UAA\n",
      "40 trnL-UAA\n",
      "trnT-UGU trnF-GAA\n",
      "41 trnF-GAA\n",
      "trnL-UAA ndhJ\n",
      "42 ndhJ\n",
      "43 ndhK\n",
      "44 psbG\n",
      "45 ndhC\n",
      "46 trnV-UAC\n",
      "ndhC trnM-CAU\n",
      "47 trnM-CAU\n",
      "trnV-UAC atpE\n",
      "48 atpE\n",
      "49 atpB\n",
      "50 rbcL\n",
      "51 accD\n",
      "52 psaI\n",
      "53 pafII\n",
      "54 cemA\n",
      "55 petA\n",
      "56 psbJ\n",
      "57 psbL\n",
      "58 psbF\n",
      "59 psbE\n",
      "60 petL\n",
      "61 petG\n",
      "62 trnW-CCA\n",
      "petG trnP-UGG\n",
      "63 trnP-UGG\n",
      "trnW-CCA psaJ\n",
      "64 psaJ\n",
      "65 rpl33\n",
      "66 rps18\n",
      "67 rpl20\n",
      "68 rps12\n",
      "69 clpP1\n",
      "70 psbB\n",
      "71 psbT\n",
      "72 pbf1\n",
      "73 psbH\n",
      "74 petB\n",
      "75 petD\n",
      "76 rpoA\n",
      "77 rps11\n",
      "78 rpl36\n",
      "79 infA\n",
      "80 rps8\n",
      "81 rpl14\n",
      "82 rpl16\n",
      "83 rps3\n",
      "84 rpl22\n",
      "85 rps19\n",
      "86 rpl2\n",
      "87 rpl23\n",
      "88 trnM\n",
      "rpl23 ycf2\n",
      "89 ycf2\n",
      "90 trnL-CAA\n",
      "ycf2 ndhB\n",
      "91 ndhB\n",
      "92 rps7\n",
      "93 rps12\n",
      "94 ycf15\n",
      "95 trnV-GAC\n",
      "ycf15 rrn16\n",
      "96 rrn16\n",
      "97 trnI-GAU\n",
      "rrn16 trnA-UGC\n",
      "98 trnA-UGC\n",
      "trnI-GAU rrn23\n",
      "99 rrn23\n",
      "100 rrn4.5\n",
      "101 rrn5\n",
      "102 trnR-ACG\n",
      "rrn5 trnN-GUU\n",
      "103 trnN-GUU\n",
      "trnR-ACG ycf1 fragment\n",
      "104 ycf1 fragment\n",
      "105 ndhF\n",
      "106 rpl32\n",
      "107 trnL-UAG\n",
      "rpl32 ccsA\n",
      "108 ccsA\n",
      "109 ndhD\n",
      "110 psaC\n",
      "111 ndhE\n",
      "112 ndhG\n",
      "113 ndhI\n",
      "114 ndhA\n",
      "115 ndhH\n",
      "116 rps15\n",
      "117 ycf1\n",
      "118 trnN-GUU\n",
      "ycf1 trnR-ACG\n",
      "119 trnR-ACG\n",
      "trnN-GUU rrn5\n",
      "120 rrn5\n",
      "121 rrn4.5\n",
      "122 rrn23\n",
      "123 trnA-UGC\n",
      "rrn23 trnI-GAU\n",
      "124 trnI-GAU\n",
      "trnA-UGC rrn16\n",
      "125 rrn16\n",
      "126 trnV-GAC\n",
      "rrn16 ycf15\n",
      "127 ycf15\n",
      "128 rps7\n",
      "129 ndhB\n",
      "130 trnL-CAA\n",
      "ndhB ycf2\n",
      "131 ycf2\n",
      "132 trnM\n",
      "ycf2 rpl23\n",
      "133 rpl23\n",
      "134 rpl2\n",
      "135 rps19 fragment\n",
      "136 trnH-GUG\n",
      "rps19 fragment psbA\n",
      "137 psbA\n",
      "138 trnK-UUU\n",
      "psbA matK\n",
      "139 matK\n",
      "140 rps16\n",
      "141 trnQ-UUG\n",
      "rps16 psbK\n",
      "142 psbK\n",
      "143 psbI\n",
      "144 trnS-GCU\n",
      "psbI trnC-GCA\n",
      "145 trnC-GCA\n",
      "trnS-GCU petN\n",
      "146 petN\n",
      "147 psbM\n",
      "148 trnD-GUC\n",
      "psbM trnY-GUA\n",
      "149 trnY-GUA\n",
      "trnD-GUC trnE-UUC\n",
      "150 trnE-UUC\n",
      "trnY-GUA rpoB\n",
      "151 rpoB\n",
      "152 rpoC1\n",
      "153 rpoC2\n",
      "154 rps2\n",
      "155 atpI\n",
      "156 atpH\n",
      "157 atpF\n",
      "158 atpA\n",
      "159 trnR-UCU\n",
      "atpA trnS-CGA\n",
      "160 trnS-CGA\n",
      "trnR-UCU trnT-GGU\n",
      "161 trnT-GGU\n",
      "trnS-CGA psbD\n",
      "162 psbD\n",
      "163 psbC\n",
      "164 trnS-UGA\n",
      "psbC psbZ\n",
      "165 psbZ\n",
      "166 trnG-GCC\n",
      "psbZ trnM-CAU\n",
      "167 trnM-CAU\n",
      "trnG-GCC rps14\n",
      "168 rps14\n",
      "169 psaB\n",
      "170 psaA\n",
      "171 pafI\n",
      "172 trnS-GGA\n",
      "pafI rps4\n",
      "173 rps4\n",
      "174 trnT-UGU\n",
      "rps4 trnL-UAA\n",
      "175 trnL-UAA\n",
      "trnT-UGU trnF-GAA\n",
      "176 trnF-GAA\n",
      "trnL-UAA ndhJ\n",
      "177 ndhJ\n",
      "178 ndhK\n",
      "179 psbG\n",
      "180 ndhC\n",
      "181 trnV-UAC\n",
      "ndhC trnM-CAU\n",
      "182 trnM-CAU\n",
      "trnV-UAC atpE\n",
      "183 atpE\n",
      "184 atpB\n",
      "185 rbcL\n",
      "186 accD\n",
      "187 psaI\n",
      "188 pafII\n",
      "189 cemA\n",
      "190 petA\n",
      "191 psbJ\n",
      "192 psbL\n",
      "193 psbF\n",
      "194 psbE\n",
      "195 petL\n",
      "196 petG\n",
      "197 trnW-CCA\n",
      "petG trnP-UGG\n",
      "198 trnP-UGG\n",
      "trnW-CCA psaJ\n",
      "199 psaJ\n",
      "200 rpl33\n",
      "201 rps18\n",
      "202 rpl20\n",
      "203 rps12\n",
      "204 clpP1\n",
      "205 psbB\n",
      "206 psbT\n",
      "207 pbf1\n",
      "208 psbH\n",
      "209 petB\n",
      "210 petD\n",
      "211 rpoA\n",
      "212 rps11\n",
      "213 rpl36\n",
      "214 infA\n",
      "215 rps8\n",
      "216 rpl14\n",
      "217 rpl16\n",
      "218 rps3\n",
      "219 rpl22\n",
      "220 rps19\n",
      "221 rpl2\n",
      "222 rpl23\n",
      "223 trnM-CAU\n",
      "rpl23 ycf2\n",
      "224 ycf2\n",
      "225 trnL-CAA\n",
      "ycf2 ndhB\n",
      "226 ndhB\n",
      "227 rps7\n",
      "228 rps12\n",
      "229 ycf15\n",
      "230 trnV-GAC\n",
      "ycf15 rrn16\n",
      "231 rrn16\n",
      "232 trnI-GAU\n",
      "rrn16 trnA-UGC\n",
      "233 trnA-UGC\n",
      "trnI-GAU rrn23\n",
      "234 rrn23\n",
      "235 rrn4.5\n",
      "236 rrn5\n",
      "237 trnR-ACG\n",
      "rrn5 trnN-GUU\n",
      "238 trnN-GUU\n",
      "trnR-ACG ycf1 fragment\n",
      "239 ycf1 fragment\n",
      "240 ndhF\n",
      "241 rpl32\n",
      "242 trnL-UAG\n",
      "rpl32 ccsA\n",
      "243 ccsA\n",
      "244 ndhD\n",
      "245 psaC\n",
      "246 ndhE\n",
      "247 ndhG\n",
      "248 ndhI\n",
      "249 ndhA\n",
      "250 ndhH\n",
      "251 rps15\n",
      "252 ycf1\n",
      "253 trnN-GUU\n",
      "ycf1 trnR-ACG\n",
      "254 trnR-ACG\n",
      "trnN-GUU rrn5\n",
      "255 rrn5\n",
      "256 rrn4.5\n",
      "257 rrn23\n",
      "258 trnA-UGC\n",
      "rrn23 trnI-GAU\n",
      "259 trnI-GAU\n",
      "trnA-UGC rrn16\n",
      "260 rrn16\n",
      "261 trnV-GAC\n",
      "rrn16 ycf15\n",
      "262 ycf15\n",
      "263 rps7\n",
      "264 ndhB\n",
      "265 trnL-CAA\n",
      "ndhB ycf2\n",
      "266 ycf2\n",
      "267 trnM-CAU\n",
      "ycf2 rpl23\n",
      "268 rpl23\n",
      "269 rpl2\n",
      "270 rps19 fragment\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(organisms)\n",
    "\n",
    "def get_neighbours(index: int, df):\n",
    "        idx = df.index.get_loc(index)\n",
    "        if idx > 0:\n",
    "            prev_gene = df.iloc[idx - 1]['gene']\n",
    "        else:\n",
    "            prev_gene = None\n",
    "        if idx < len(df) - 1:\n",
    "            next_gene = df.iloc[idx + 1]['gene']\n",
    "        else:\n",
    "            next_gene = None\n",
    "        return prev_gene, next_gene\n",
    "\n",
    "neighbours = {}\n",
    "\n",
    "for i in df.index:\n",
    "    \n",
    "    gene = df.iloc[i, 1]\n",
    "    print(i, gene)\n",
    "    #print(type(i))\n",
    "    if gene.startswith(\"trn\"):\n",
    "         prev, next = get_neighbours(i, df)\n",
    "         print(prev, next)\n",
    "\n",
    "# NB: There are problems in the code, look at the output for idx = 7 (psbK)\n",
    "#\n",
    "# It is OK, look at the `if gene.startswith('trn')` line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>organism</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Crepis callicephala</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>region</th>\n",
       "      <th>IRA</th>\n",
       "      <th>IRB</th>\n",
       "      <th>IRA</th>\n",
       "      <th>IRB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ndhB</th>\n",
       "      <td>2202.0</td>\n",
       "      <td>2202.0</td>\n",
       "      <td>2202.0</td>\n",
       "      <td>2202.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ndhF</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2226.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2226.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl2</th>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl23</th>\n",
       "      <td>282.0</td>\n",
       "      <td>282.0</td>\n",
       "      <td>282.0</td>\n",
       "      <td>282.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps19 fragment</th>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps7</th>\n",
       "      <td>468.0</td>\n",
       "      <td>468.0</td>\n",
       "      <td>468.0</td>\n",
       "      <td>468.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn16</th>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "      <td>1490.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn23</th>\n",
       "      <td>2810.0</td>\n",
       "      <td>2810.0</td>\n",
       "      <td>2810.0</td>\n",
       "      <td>2810.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn4.5</th>\n",
       "      <td>103.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>103.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rrn5</th>\n",
       "      <td>122.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>122.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnA-UGC</th>\n",
       "      <td>893.0</td>\n",
       "      <td>893.0</td>\n",
       "      <td>893.0</td>\n",
       "      <td>893.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnI-GAU</th>\n",
       "      <td>834.0</td>\n",
       "      <td>834.0</td>\n",
       "      <td>836.0</td>\n",
       "      <td>836.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL-CAA</th>\n",
       "      <td>81.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnM</th>\n",
       "      <td>75.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnM-CAU</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnN-GUU</th>\n",
       "      <td>72.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnR-ACG</th>\n",
       "      <td>74.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnV-GAC</th>\n",
       "      <td>73.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf1 fragment</th>\n",
       "      <td>NaN</td>\n",
       "      <td>480.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>480.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf15</th>\n",
       "      <td>192.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf2</th>\n",
       "      <td>6855.0</td>\n",
       "      <td>6855.0</td>\n",
       "      <td>6855.0</td>\n",
       "      <td>6855.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism       Crepis callicephala         Crepis purpurea        \n",
       "region                         IRA     IRB             IRA     IRB\n",
       "gene                                                              \n",
       "ndhB                        2202.0  2202.0          2202.0  2202.0\n",
       "ndhF                           NaN  2226.0             NaN  2226.0\n",
       "rpl2                        1490.0  1490.0          1490.0  1490.0\n",
       "rpl23                        282.0   282.0           282.0   282.0\n",
       "rps19 fragment                60.0     NaN            55.0     NaN\n",
       "rps7                         468.0   468.0           468.0   468.0\n",
       "rrn16                       1490.0  1490.0          1490.0  1490.0\n",
       "rrn23                       2810.0  2810.0          2810.0  2810.0\n",
       "rrn4.5                       103.0   103.0           103.0   103.0\n",
       "rrn5                         122.0   116.0           121.0   122.0\n",
       "trnA-UGC                     893.0   893.0           893.0   893.0\n",
       "trnI-GAU                     834.0   834.0           836.0   836.0\n",
       "trnL-CAA                      81.0    81.0            83.0    83.0\n",
       "trnM                          75.0    74.0             NaN     NaN\n",
       "trnM-CAU                       NaN     NaN            75.0    75.0\n",
       "trnN-GUU                      72.0    72.0            74.0    74.0\n",
       "trnR-ACG                      74.0    74.0            75.0    75.0\n",
       "trnV-GAC                      73.0    72.0            72.0    72.0\n",
       "ycf1 fragment                  NaN   480.0             NaN   480.0\n",
       "ycf15                        192.0   192.0           192.0   192.0\n",
       "ycf2                        6855.0  6855.0          6855.0  6855.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_repeats = df[df['region'].isin(['IRA', 'IRB'])]\n",
    "\n",
    "pivot_table = df_repeats.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism', 'region'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Reorder levels in the columns\n",
    "#pivot_table = df2.reorder_levels([1, 0], axis=1)\n",
    "\n",
    "# Sort the columns\n",
    "pivot_table = pivot_table.sort_index(axis=1)\n",
    "\n",
    "# Display the pivot table\n",
    "display(pivot_table)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking ycf1 to develope the logic for renaming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>ycf1</td>\n",
       "      <td>IRB</td>\n",
       "      <td>106267</td>\n",
       "      <td>106747</td>\n",
       "      <td>480</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>ycf1</td>\n",
       "      <td>SSC</td>\n",
       "      <td>120557</td>\n",
       "      <td>125527</td>\n",
       "      <td>4970</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>ycf1 fragment</td>\n",
       "      <td>IRB</td>\n",
       "      <td>106289</td>\n",
       "      <td>106769</td>\n",
       "      <td>480</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>ycf1</td>\n",
       "      <td>SSC</td>\n",
       "      <td>120578</td>\n",
       "      <td>125551</td>\n",
       "      <td>4973</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              organism           gene region   start     end  length  spanned\n",
       "0  Crepis callicephala           ycf1    IRB  106267  106747     480     True\n",
       "1  Crepis callicephala           ycf1    SSC  120557  125527    4970     True\n",
       "2      Crepis purpurea  ycf1 fragment    IRB  106289  106769     480     True\n",
       "3      Crepis purpurea           ycf1    SSC  120578  125551    4973     True"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO, SeqFeature\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/in/C_callicephala_curated_1_cleaned_meta.gb\"\n",
    "c_pur = \"plastomes/in/Crepis_purpurea_edited+.gb\"\n",
    "\n",
    "regions = {}\n",
    "dataset = []\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in c_cal, c_pur:\n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            features = record.features\n",
    "            # check if there features other than 'source'\n",
    "            if len(features) > 1:\n",
    "                for feature in features:\n",
    "                    # check for genome regions\n",
    "                    if feature.type in ['misc_feature', 'repeat_region']:\n",
    "                        label = feature.qualifiers.get('note', \"\")[0].split()[-1].strip(\"()\")\n",
    "                        regions[label] = [int(feature.location.start), int(feature.location.end)]\n",
    "                # filling the dataframe\n",
    "                organism = record.annotations.get('organism', \"\")\n",
    "                for feature in features:\n",
    "                    if feature.type == 'gene':\n",
    "                        gene = feature.qualifiers.get('gene', \"\")\n",
    "                        if \"ycf1\" in gene[0] and not \"5\" in gene[0]:\n",
    "                            start = int(feature.location.start)\n",
    "                            end = int(feature.location.end)\n",
    "\n",
    "                            for region in regions.items():\n",
    "                                ranges = region[-1]\n",
    "                                if ranges[0] <= start < ranges[-1]:\n",
    "                                    #print(f'{region[0]}\\t{gene}')\n",
    "                                    spanned = False\n",
    "                                    if end >= ranges[-1]:\n",
    "                                        spanned = True\n",
    "                                    dataset.append((organism, gene[0], region[0], start, end, len(feature), spanned))\n",
    "\n",
    "df = pd.DataFrame(dataset, columns=['organism', 'gene', 'region', 'start', 'end', 'length', 'spanned'])\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "checking trna fragmentation in inverted repeats: case of trnL-CAA fragment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnL-UAA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>46769</td>\n",
       "      <td>47293</td>\n",
       "      <td>524</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnL-CAA</td>\n",
       "      <td>IRB</td>\n",
       "      <td>91329</td>\n",
       "      <td>91410</td>\n",
       "      <td>81</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnL-UAG</td>\n",
       "      <td>SSC</td>\n",
       "      <td>110966</td>\n",
       "      <td>111046</td>\n",
       "      <td>80</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnL-CAA</td>\n",
       "      <td>IRA</td>\n",
       "      <td>140384</td>\n",
       "      <td>140465</td>\n",
       "      <td>81</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnL-UAA | trnL</td>\n",
       "      <td>LSC</td>\n",
       "      <td>46804</td>\n",
       "      <td>47329</td>\n",
       "      <td>525</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnL-CAA | trnL-fragment</td>\n",
       "      <td>IRB</td>\n",
       "      <td>91342</td>\n",
       "      <td>91425</td>\n",
       "      <td>83</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnL-UAG | trnL</td>\n",
       "      <td>SSC</td>\n",
       "      <td>110983</td>\n",
       "      <td>111063</td>\n",
       "      <td>80</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnL-CAA | trnL-fragment</td>\n",
       "      <td>IRA</td>\n",
       "      <td>140415</td>\n",
       "      <td>140498</td>\n",
       "      <td>83</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              organism                      gene region   start     end  \\\n",
       "0  Crepis callicephala                  trnL-UAA    LSC   46769   47293   \n",
       "1  Crepis callicephala                  trnL-CAA    IRB   91329   91410   \n",
       "2  Crepis callicephala                  trnL-UAG    SSC  110966  111046   \n",
       "3  Crepis callicephala                  trnL-CAA    IRA  140384  140465   \n",
       "4      Crepis purpurea           trnL-UAA | trnL    LSC   46804   47329   \n",
       "5      Crepis purpurea  trnL-CAA | trnL-fragment    IRB   91342   91425   \n",
       "6      Crepis purpurea           trnL-UAG | trnL    SSC  110983  111063   \n",
       "7      Crepis purpurea  trnL-CAA | trnL-fragment    IRA  140415  140498   \n",
       "\n",
       "   length  spanned  \n",
       "0     524    False  \n",
       "1      81    False  \n",
       "2      80    False  \n",
       "3      81    False  \n",
       "4     525    False  \n",
       "5      83    False  \n",
       "6      80    False  \n",
       "7      83    False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO, SeqFeature\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/in/C_callicephala_curated_1_cleaned_meta.gb\"\n",
    "c_pur = \"plastomes/in/Crepis_purpurea_edited+.gb\"\n",
    "\n",
    "regions = {}\n",
    "dataset = []\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in c_cal, c_pur:\n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            features = record.features\n",
    "            # check if there features other than 'source'\n",
    "            if len(features) > 1:\n",
    "                for feature in features:\n",
    "                    # check for genome regions\n",
    "                    if feature.type in ['misc_feature', 'repeat_region']:\n",
    "                        label = feature.qualifiers.get('note', \"\")[0].split()[-1].strip(\"()\")\n",
    "                        regions[label] = [int(feature.location.start), int(feature.location.end)]\n",
    "                # filling the dataframe\n",
    "                organism = record.annotations.get('organism', \"\")\n",
    "                for feature in features:\n",
    "                    if feature.type == 'gene':\n",
    "                        gene = feature.qualifiers.get('gene', \"\")\n",
    "                        if \"trnL\" in gene[0]:\n",
    "                            start = int(feature.location.start)\n",
    "                            end = int(feature.location.end)\n",
    "\n",
    "                            for region in regions.items():\n",
    "                                ranges = region[-1]\n",
    "                                if ranges[0] <= start < ranges[-1]:\n",
    "                                    #print(f'{region[0]}\\t{gene}')\n",
    "                                    spanned = False\n",
    "                                    if end >= ranges[-1]:\n",
    "                                        spanned = True\n",
    "                                    dataset.append((organism, gene[0], region[0], start, end, len(feature), spanned))\n",
    "\n",
    "df = pd.DataFrame(dataset, columns=['organism', 'gene', 'region', 'start', 'end', 'length', 'spanned'])\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare features within LSC after labels being normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with different values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>organism</th>\n",
       "      <th>Crepis callicephala</th>\n",
       "      <th>Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>petD</th>\n",
       "      <td>525</td>\n",
       "      <td>1183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>psaA</th>\n",
       "      <td>2253</td>\n",
       "      <td>1776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl16</th>\n",
       "      <td>1445</td>\n",
       "      <td>1440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl22</th>\n",
       "      <td>465</td>\n",
       "      <td>540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC1</th>\n",
       "      <td>2799</td>\n",
       "      <td>2520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC2</th>\n",
       "      <td>4125</td>\n",
       "      <td>2274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps16</th>\n",
       "      <td>1120</td>\n",
       "      <td>1125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism  Crepis callicephala  Crepis purpurea\n",
       "gene                                          \n",
       "petD                      525             1183\n",
       "psaA                     2253             1776\n",
       "rpl16                    1445             1440\n",
       "rpl22                     465              540\n",
       "rpoC1                    2799             2520\n",
       "rpoC2                    4125             2274\n",
       "rps16                    1120             1125"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA genes with different values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>organism</th>\n",
       "      <th>Crepis callicephala</th>\n",
       "      <th>Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>trnD-GUC</th>\n",
       "      <td>74.0</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnE-UUC</th>\n",
       "      <td>72.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnF-GAA</th>\n",
       "      <td>70.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnG-GCC</th>\n",
       "      <td>71.0</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnG-UCC</th>\n",
       "      <td>781.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnK</th>\n",
       "      <td>2600.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnK-UUU</th>\n",
       "      <td>37.0</td>\n",
       "      <td>2600.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL-UAA</th>\n",
       "      <td>524.0</td>\n",
       "      <td>525.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnP-UGG</th>\n",
       "      <td>75.0</td>\n",
       "      <td>76.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnQ-UUG</th>\n",
       "      <td>73.0</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnS-CGA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>782.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnT-UGU</th>\n",
       "      <td>73.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnY-GUA</th>\n",
       "      <td>84.0</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism  Crepis callicephala  Crepis purpurea\n",
       "gene                                          \n",
       "trnD-GUC                 74.0             77.0\n",
       "trnE-UUC                 72.0             74.0\n",
       "trnF-GAA                 70.0             74.0\n",
       "trnG-GCC                 71.0             72.0\n",
       "trnG-UCC                781.0              NaN\n",
       "trnK                   2600.0              NaN\n",
       "trnK-UUU                 37.0           2600.0\n",
       "trnL-UAA                524.0            525.0\n",
       "trnP-UGG                 75.0             76.0\n",
       "trnQ-UUG                 73.0             72.0\n",
       "trnS-CGA                  NaN            782.0\n",
       "trnT-UGU                 73.0             75.0\n",
       "trnY-GUA                 84.0             87.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_target = df[df['region'].isin(['LSC'])]\n",
    "\n",
    "non_tRNA = df_target[~df_target['gene'].str.startswith('trn')]\n",
    "tRNA = df_target[df_target['gene'].str.startswith('trn')]\n",
    "\n",
    "pt_tRNA = tRNA.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Sort the columns\n",
    "pt_tRNA = pt_tRNA.sort_index(axis=1)\n",
    "\n",
    "pt_nontRNA = non_tRNA.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Sort the columns\n",
    "pt_nontRNA = pt_nontRNA.sort_index(axis=1)\n",
    "\n",
    "\n",
    "# Display the pivot table\n",
    "#display(pt_tRNA, pt_nontRNA)\n",
    "#with pd.option_context('display.max_rows', None,):\n",
    "#    display(pt_nontRNA)\n",
    "\n",
    "# Find differences\n",
    "diff_nontRNA = pt_nontRNA['Crepis callicephala'] != pt_nontRNA['Crepis purpurea']\n",
    "print(\"Genes with different values:\")\n",
    "display(pt_nontRNA[diff_nontRNA])\n",
    "\n",
    "diff_tRNA = pt_tRNA['Crepis callicephala'] != pt_tRNA['Crepis purpurea']\n",
    "print(\"tRNA genes with different values:\")\n",
    "display(pt_tRNA[diff_tRNA])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare features which labels where normalized, within SSC region."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with different values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>organism</th>\n",
       "      <th>Crepis callicephala</th>\n",
       "      <th>Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ndhA</th>\n",
       "      <td>2147</td>\n",
       "      <td>2146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ndhD</th>\n",
       "      <td>1503</td>\n",
       "      <td>1521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf1</th>\n",
       "      <td>4970</td>\n",
       "      <td>4973</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "organism  Crepis callicephala  Crepis purpurea\n",
       "gene                                          \n",
       "ndhA                     2147             2146\n",
       "ndhD                     1503             1521\n",
       "ycf1                     4970             4973"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tRNA genes with different values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>organism</th>\n",
       "      <th>Crepis callicephala</th>\n",
       "      <th>Crepis purpurea</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Crepis callicephala, Crepis purpurea]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_target = df[df['region'].isin(['SSC'])]\n",
    "\n",
    "non_tRNA = df_target[~df_target['gene'].str.startswith('trn')]\n",
    "tRNA = df_target[df_target['gene'].str.startswith('trn')]\n",
    "\n",
    "pt_tRNA = tRNA.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Sort the columns\n",
    "pt_tRNA = pt_tRNA.sort_index(axis=1)\n",
    "\n",
    "pt_nontRNA = non_tRNA.pivot_table(\n",
    "    values='length',\n",
    "    index='gene',\n",
    "    columns=['organism'],\n",
    "    aggfunc='first'  # or 'mean', 'sum', etc. depending on your needs\n",
    ")\n",
    "\n",
    "# Sort the columns\n",
    "pt_nontRNA = pt_nontRNA.sort_index(axis=1)\n",
    "\n",
    "\n",
    "# Display the pivot table\n",
    "#display(pt_tRNA, pt_nontRNA)\n",
    "#with pd.option_context('display.max_rows', None,):\n",
    "#    display(pt_nontRNA)\n",
    "\n",
    "# Find differences\n",
    "diff_nontRNA = pt_nontRNA['Crepis callicephala'] != pt_nontRNA['Crepis purpurea']\n",
    "print(\"Genes with different values:\")\n",
    "display(pt_nontRNA[diff_nontRNA])\n",
    "\n",
    "diff_tRNA = pt_tRNA['Crepis callicephala'] != pt_tRNA['Crepis purpurea']\n",
    "print(\"tRNA genes with different values:\")\n",
    "display(pt_tRNA[diff_tRNA])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Applying changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/in/C_callicephala_curated_1_cleaned_meta.gb\n",
      "plastomes/in/C_callicephala_curated_1_cleaned_meta.gb plastomes/out/Crepis_callicephala.gb\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 348 <class 'list'>\n",
      "records: 149980 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "plastomes/in/Crepis_purpurea_edited+.gb\n",
      "plastomes/in/Crepis_purpurea_edited+.gb plastomes/out/Crepis_purpurea.gb\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n",
      "features: 353 <class 'list'>\n",
      "records: 150007 <class 'Bio.SeqRecord.SeqRecord'>\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnH-GUG</td>\n",
       "      <td>LSC</td>\n",
       "      <td>2</td>\n",
       "      <td>77</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>psbA</td>\n",
       "      <td>LSC</td>\n",
       "      <td>474</td>\n",
       "      <td>1536</td>\n",
       "      <td>1062</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK-UUU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>1758</td>\n",
       "      <td>4358</td>\n",
       "      <td>2600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>matK</td>\n",
       "      <td>LSC</td>\n",
       "      <td>2079</td>\n",
       "      <td>3591</td>\n",
       "      <td>1512</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>rps16</td>\n",
       "      <td>LSC</td>\n",
       "      <td>5149</td>\n",
       "      <td>6269</td>\n",
       "      <td>1120</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>ycf2</td>\n",
       "      <td>IRA</td>\n",
       "      <td>140899</td>\n",
       "      <td>147754</td>\n",
       "      <td>6855</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147865</td>\n",
       "      <td>147940</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rpl23</td>\n",
       "      <td>IRA</td>\n",
       "      <td>148104</td>\n",
       "      <td>148386</td>\n",
       "      <td>282</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rpl2</td>\n",
       "      <td>IRA</td>\n",
       "      <td>148404</td>\n",
       "      <td>149894</td>\n",
       "      <td>1490</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>rps19 fragment</td>\n",
       "      <td>IRA</td>\n",
       "      <td>149952</td>\n",
       "      <td>150007</td>\n",
       "      <td>55</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>270 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism            gene region   start     end  length  \\\n",
       "0    Crepis callicephala        trnH-GUG    LSC       2      77      75   \n",
       "1    Crepis callicephala            psbA    LSC     474    1536    1062   \n",
       "2    Crepis callicephala        trnK-UUU    LSC    1758    4358    2600   \n",
       "3    Crepis callicephala            matK    LSC    2079    3591    1512   \n",
       "4    Crepis callicephala           rps16    LSC    5149    6269    1120   \n",
       "..                   ...             ...    ...     ...     ...     ...   \n",
       "265      Crepis purpurea            ycf2    IRA  140899  147754    6855   \n",
       "266      Crepis purpurea        trnM-CAU    IRA  147865  147940      75   \n",
       "267      Crepis purpurea           rpl23    IRA  148104  148386     282   \n",
       "268      Crepis purpurea            rpl2    IRA  148404  149894    1490   \n",
       "269      Crepis purpurea  rps19 fragment    IRA  149952  150007      55   \n",
       "\n",
       "     spanned  \n",
       "0      False  \n",
       "1      False  \n",
       "2      False  \n",
       "3      False  \n",
       "4      False  \n",
       "..       ...  \n",
       "265    False  \n",
       "266    False  \n",
       "267    False  \n",
       "268    False  \n",
       "269     True  \n",
       "\n",
       "[270 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>36111</td>\n",
       "      <td>36185</td>\n",
       "      <td>74</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>51913</td>\n",
       "      <td>51986</td>\n",
       "      <td>73</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRB</td>\n",
       "      <td>83887</td>\n",
       "      <td>83961</td>\n",
       "      <td>74</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147833</td>\n",
       "      <td>147908</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>36147</td>\n",
       "      <td>36221</td>\n",
       "      <td>74</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>51948</td>\n",
       "      <td>52021</td>\n",
       "      <td>73</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRB</td>\n",
       "      <td>83900</td>\n",
       "      <td>83975</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnM-CAU</td>\n",
       "      <td>IRA</td>\n",
       "      <td>147865</td>\n",
       "      <td>147940</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism      gene region   start     end  length  spanned\n",
       "31   Crepis callicephala  trnM-CAU    LSC   36111   36185      74    False\n",
       "46   Crepis callicephala  trnM-CAU    LSC   51913   51986      73    False\n",
       "87   Crepis callicephala  trnM-CAU    IRB   83887   83961      74    False\n",
       "131  Crepis callicephala  trnM-CAU    IRA  147833  147908      75    False\n",
       "166      Crepis purpurea  trnM-CAU    LSC   36147   36221      74    False\n",
       "181      Crepis purpurea  trnM-CAU    LSC   51948   52021      73    False\n",
       "222      Crepis purpurea  trnM-CAU    IRB   83900   83975      75    False\n",
       "266      Crepis purpurea  trnM-CAU    IRA  147865  147940      75    False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>gene</th>\n",
       "      <th>region</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>spanned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Crepis callicephala</td>\n",
       "      <td>trnK-UUU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>1758</td>\n",
       "      <td>4358</td>\n",
       "      <td>2600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Crepis purpurea</td>\n",
       "      <td>trnK-UUU</td>\n",
       "      <td>LSC</td>\n",
       "      <td>1760</td>\n",
       "      <td>4360</td>\n",
       "      <td>2600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                organism      gene region  start   end  length  spanned\n",
       "2    Crepis callicephala  trnK-UUU    LSC   1758  4358    2600    False\n",
       "137      Crepis purpurea  trnK-UUU    LSC   1760  4360    2600    False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO, SeqFeature\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/in/C_callicephala_curated_1_cleaned_meta.gb\"\n",
    "c_pur = \"plastomes/in/Crepis_purpurea_edited+.gb\"\n",
    "\n",
    "exception = [\n",
    "    \"trnL-CAA fragment\",\n",
    "    \"trnT-GGU | trnT-fragment\",\n",
    "]\n",
    "\n",
    "regions = {}\n",
    "dataset = []\n",
    "new_record = []\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in c_cal, c_pur:\n",
    "    print(acc)\n",
    "    output = acc.replace(\"/in/\", \"/out/\")\n",
    "    output = output.replace(\"C_callicephala_curated_1_cleaned_meta\", \"Crepis_callicephala\")\n",
    "    output = output.replace(\"Crepis_purpurea_edited+\", \"Crepis_purpurea\")\n",
    "    print(acc, output)\n",
    "\n",
    "    new_features = []\n",
    "\n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            features = record.features\n",
    "            # check if there features other than 'source'\n",
    "            if len(features) > 1:\n",
    "                for feature in features:\n",
    "                    # check for genome regions\n",
    "                    if feature.type in ['misc_feature', 'repeat_region']:\n",
    "                        label = feature.qualifiers.get('note', \"\")[0].split()[-1].strip(\"()\")\n",
    "                        regions[label] = [int(feature.location.start), int(feature.location.end)]\n",
    "                # filling the dataframe\n",
    "                organism = record.annotations.get('organism', \"\")\n",
    "                for feature in features:\n",
    "                    if feature.type == 'gene':\n",
    "                        \n",
    "                        gene = feature.qualifiers.get('gene', \"\")\n",
    "                        start = int(feature.location.start)\n",
    "                        end = int(feature.location.end)\n",
    "\n",
    "                        # handling 'trnM' case\n",
    "                        if gene[0] == \"trnM\":\n",
    "                            gene[0] = \"trnM-CAU\"\n",
    "\n",
    "                        # renaming 'trnK'\n",
    "                        if gene[0] == \"trnK\":\n",
    "                            gene[0] = \"trnK-UUU\"\n",
    "\n",
    "                        # handling the case of excessive trnK\n",
    "                        if gene[0] == \"trnK-UUU\" and len(feature) < 100:\n",
    "                            pass\n",
    "                        else:\n",
    "                            for region in regions.items():\n",
    "                                gene_label = normalize_labels(gene[0], region[0], exception)\n",
    "\n",
    "                                feature.qualifiers['gene'] = gene_label\n",
    "\n",
    "                                ranges = region[-1]\n",
    "                                if ranges[0] <= start < ranges[-1]:\n",
    "                                    #print(f'{region[0]}\\t{gene}')\n",
    "                                    spanned = False\n",
    "                                    if end >= ranges[-1]:\n",
    "                                        spanned = True\n",
    "                                    dataset.append((organism, normalize_labels(gene[0], region[0], exception), region[0], start, end, len(feature), spanned))\n",
    "\n",
    "                    new_features.append(feature)\n",
    "                    new_record = SeqRecord.SeqRecord(\n",
    "                        seq=record.seq,\n",
    "                        id=record.id,\n",
    "                        name=record.name,\n",
    "                        description=record.description,\n",
    "                        features=new_features,\n",
    "                        annotations=record.annotations\n",
    "                    )\n",
    "\n",
    "                    print(\"features:\", len(features), type(features))\n",
    "                    print(\"records:\", len(record), type(record))\n",
    "                    SeqIO.write(new_record, output, \"genbank\")\n",
    "\n",
    "df = pd.DataFrame(dataset, columns=['organism', 'gene', 'region', 'start', 'end', 'length', 'spanned'])\n",
    "display(df)\n",
    "display(df[df['gene'].str.slice(0,4) == 'trnM'])\n",
    "display(df[df['gene'].str.slice(0,4) == 'trnK'])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Changing Reference field in genbank file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/out/Crepis_callicephala.gb\n",
      "Crepis callicephala Crepis_callicephala\n",
      "initial refs length: 0\n",
      "plastomes/w-refs/Crepis_callicephala.gb\n",
      "plastomes/w-refs\n",
      "final refs length: 0\n",
      "plastomes/out/Crepis_purpurea.gb\n",
      "Crepis purpurea Crepis_purpurea\n",
      "initial refs length: 1\n",
      "plastomes/w-refs/Crepis_purpurea.gb\n",
      "plastomes/w-refs\n",
      "final refs length: 0\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO, SeqFeature\n",
    "import os\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/out/Crepis_callicephala.gb\"\n",
    "c_pur = \"plastomes/out/Crepis_purpurea.gb\"\n",
    "cich = \"plastomes/test/cichorieae_seqrecords.gb\"\n",
    "\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in c_cal, c_pur:\n",
    "    print(acc)\n",
    "    \n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "\n",
    "            # Deleye Comment field if present\n",
    "            if 'comment' in record.annotations.keys():\n",
    "                record.annotations.pop('comment')\n",
    "            \n",
    "            refs = record.annotations.get('references', [])\n",
    "            print(record.annotations.get('organism', \"\"), record.id)\n",
    "            print(f'initial refs length: {len(refs)}')\n",
    "\n",
    "            authors = \"Asan O. Emirsaliev, Dmitry A. Afonnikov, Irina V. Mitrofanova, Elena A. Salina, Alex V. Kochetov\"\n",
    "            title = \"Phylogenetic analysis of rare Crepis species of Crimean flora (C. purpurea (Willd.) M. Bieb. and C. callicephala Juz.)\"\n",
    "            journal: \"Unpublished\"\n",
    "            department = \"Department of Plant Molecular Genetics, IC&G SB RAS. 10, Ac. Lavrentieva ave., Novosibirsk, 630090, Russia\"\n",
    "\n",
    "            for ref in refs:\n",
    "                #print(type(refs), type(ref), ref)\n",
    "                #print(ref)\n",
    "                if ref.title == \"GeSeq - versatile and accurate annotation of organelle genomes\":\n",
    "                    removed = refs.pop()\n",
    "                    #print(\"removed:\", removed)\n",
    "            \n",
    "            new_ref = SeqFeature.Reference()\n",
    "            new_ref.authors = authors\n",
    "            new_ref.title = title\n",
    "            new_ref.journal = \"Unpublished\"\n",
    "            \n",
    "            submit_ref = SeqFeature.Reference()\n",
    "            submit_ref.authors = authors\n",
    "            submit_ref.title = \"Direct Submission\"\n",
    "            submit_ref.journal = department\n",
    "\n",
    "            record.annotations['references'] = [new_ref, submit_ref]\n",
    "            #print(f'\\nREFS:')\n",
    "            for ref in refs:\n",
    "                #print(ref)\n",
    "                pass\n",
    "\n",
    "            output = acc.replace(\"/out/\", \"/w-refs/\")\n",
    "            print(output)\n",
    "\n",
    "            dir_path = os.path.dirname(output)\n",
    "            print(dir_path)\n",
    "            os.makedirs(dir_path, exist_ok=True)\n",
    "            SeqIO.write(record, output, \"genbank\")\n",
    "\n",
    "            print(f'final refs length: {len(refs)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparing gb file for submission.\n",
    "Sequin is not working, and GB2Sequin tool only modifies Sequin files, but not tab-delimited files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/w-refs/Crepis_callicephala.gb\n",
      "Crepis_callicephala Crepis callicephala 347\n",
      "plastomes/final/Crepis_callicephala.gb\n",
      "plastomes/final\n",
      "plastomes/w-refs/Crepis_purpurea.gb\n",
      "Crepis_purpurea Crepis purpurea 352\n",
      "plastomes/final/Crepis_purpurea.gb\n",
      "plastomes/final\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO, SeqFeature\n",
    "import os\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/w-refs/Crepis_callicephala.gb\"\n",
    "c_pur = \"plastomes/w-refs/Crepis_purpurea.gb\"\n",
    "cich = \"plastomes/test/cichorieae_seqrecords.gb\"\n",
    "\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in [c_cal, c_pur]:\n",
    "    print(acc)\n",
    "    \n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            if not \"angustana\" in record.annotations.get('organism', \"\"):\n",
    "                features = record.features\n",
    "                print(record.id, record.annotations.get('organism', \"\"), len(features) - 1)\n",
    "                \n",
    "                # check if there are features other than 'source'\n",
    "                if len(features) > 1:\n",
    "                    for feature in features:\n",
    "\n",
    "                        # remove 'annotator' and 'info' keys that are not valid\n",
    "                        if \"annotator\" in feature.qualifiers.keys():\n",
    "                            feature.qualifiers.pop('annotator')\n",
    "                        if \"info\" in feature.qualifiers.keys():\n",
    "                            feature.qualifiers.pop('info')\n",
    "\n",
    "                        # fix 'trans_splicin' error in both genes and CDSs\n",
    "                        if feature.qualifiers.get('gene', [\"\"])[0] == \"rps12\":\n",
    "                                    if 'trans_splicin' in feature.qualifiers.keys():\n",
    "                                        val = feature.qualifiers.get('trans_splicin', \"\")\n",
    "                                        feature.qualifiers.pop('trans_splicin')\n",
    "                                        feature.qualifiers['trans_splicing'] = val\n",
    "                                    #print(feature)\n",
    "\n",
    "                        t = feature.type\n",
    "                        match t:\n",
    "                            case \"CDS\":\n",
    "                                #print(t, feature.qualifiers)\n",
    "                                if not 'product' in feature.qualifiers.keys():\n",
    "                                    #print(f'BEFORE fixing:\\n{feature}')\n",
    "                                    gene_to_look = feature.qualifiers.get('gene', [\"\"])[0]\n",
    "                                    match gene_to_look:\n",
    "                                        case \"ycf1\":\n",
    "                                            feature.qualifiers['product'] = \"Putative membrane protein ycf1; RF1; hypothetical chloroplast RF1 (chloroplast)\"\n",
    "                                        case \"psbG\":\n",
    "                                            feature.qualifiers['product'] = \"PsbG; Photosystem II G protein\"\n",
    "                                        case \"accD\":\n",
    "                                            feature.qualifiers['product'] = \"Acetyl-coenzyme A carboxylase carboxyl transferase subunit beta, chloroplastic; ACCase subunit beta; Acetyl-CoA carboxylase carboxyltransferase subunit beta; AccD\"\n",
    "                                        case \"rps19-fragment\":\n",
    "                                            feature.qualifiers['product'] = \"Small ribosomal subunit protein uS19c; 30S ribosomal protein S19, chloroplastic; Ribosomal protein S19\"\n",
    "                                    #print(f'AFTER fixing:\\n{feature}')\n",
    "                                    pass\n",
    "                                # fixing translation table issue\n",
    "                                if \"transl_table\" in feature.qualifiers.keys():\n",
    "                                    #print(feature.qualifiers['transl_table'])\n",
    "                                    pass\n",
    "                                else:\n",
    "                                    feature.qualifiers['transl_table'] = 11\n",
    "                            case \"gene\":\n",
    "                                pass\n",
    "\n",
    "                output = acc.replace(\"w-refs\", \"final\")\n",
    "                print(output)\n",
    "\n",
    "                dir_path = os.path.dirname(output)\n",
    "                print(dir_path)\n",
    "                os.makedirs(dir_path, exist_ok=True)\n",
    "                SeqIO.write(record, output, \"genbank\")\n",
    "                        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List of errors in annotation file processing\n",
    "\n",
    "1. ERROR: valid [SEQ_FEAT.InternalStop] 1 internal stops. Genetic code [11] FEATURE: CDS: PsbG; Photosystem II G protein [lcl|Crepis_callicephala:c49645-49597] [lcl|Crepis_callicephala: raw, dna len= 149980] -> [lcl|Crepis_callicephala_26]\n",
    "\n",
    "**DESCRIPTION:** The CDS entity has no annotation. The possible solution would be checking if CDS would have translation, and add adding to the resultant annotation.\n",
    "\n",
    "2. ERROR: valid [SEQ_FEAT.NoStop] Missing stop codon FEATURE: CDS: PsbG; Photosystem II G protein [lcl|Crepis_callicephala:c49645-49597] [lcl|Crepis_callicephala: raw, dna len= 149980] -> [lcl|Crepis_callicephala_26]\n",
    "\n",
    "**DESCRIPTION:** There are no stop codon within expected length of the gene. There are several possibilities for optimal solution:\n",
    "- pseudogenize the gene.\n",
    "- expand untill the stop codon.\n",
    "- break on internal stop codon if exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/final/Crepis_callicephala.gb\n",
      "Crepis_callicephala Crepis callicephala 347\n",
      "type: CDS\n",
      "location: [49596:49645](-)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['psbG']\n",
      "    Key: product, Value: ['PsbG; Photosystem II G protein']\n",
      "    Key: transl_table, Value: ['11']\n",
      "\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 2\n",
      "Seq('ATGGTCTTAGCTCCTGAATATTAAGACAATAAAAAGAAAAAAGGAAAAA')\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 2\n",
      "Seq('ATGGTCTTAGCTCCTGAATATTAAGACAATAAAAAGAAAAAAGGAAAAAATTTA')\n",
      "ATGGTCTTAGCTCCTGAATATTCAGACAATAAAAAGAAAAAAGGAAAAAATTGA\n",
      "L. sativa: 54\n",
      "C.callicephala: 54\n"
     ]
    },
    {
     "ename": "TranslationError",
     "evalue": "Final codon 'TTA' is not a stop codon",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTranslationError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 35\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;28mprint\u001b[39m(ls)\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mL. sativa: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(ls)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mC.callicephala: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(b)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 35\u001b[0m c \u001b[38;5;241m=\u001b[39m \u001b[43mb\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtranslate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrecord\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m11\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;28mprint\u001b[39m(c)\n",
      "File \u001b[0;32m~/miniconda3/envs/biopython-env/lib/python3.12/site-packages/Bio/SeqFeature.py:506\u001b[0m, in \u001b[0;36mSeqFeature.translate\u001b[0;34m(self, parent_sequence, table, start_offset, stop_symbol, to_stop, cds, gap)\u001b[0m\n\u001b[1;32m    503\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m cds \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    504\u001b[0m     cds \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCDS\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m--> 506\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfeat_seq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtranslate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    507\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcodon_table\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    508\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstop_symbol\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop_symbol\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    509\u001b[0m \u001b[43m    \u001b[49m\u001b[43mto_stop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mto_stop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    510\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    511\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgap\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgap\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    512\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/biopython-env/lib/python3.12/site-packages/Bio/SeqRecord.py:1409\u001b[0m, in \u001b[0;36mSeqRecord.translate\u001b[0;34m(self, table, stop_symbol, to_stop, cds, gap, id, name, description, features, annotations, letter_annotations, dbxrefs)\u001b[0m\n\u001b[1;32m   1406\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprotein\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mannotations\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmolecule_type\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m   1407\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mProteins cannot be translated!\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1408\u001b[0m answer \u001b[38;5;241m=\u001b[39m SeqRecord(\n\u001b[0;32m-> 1409\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mseq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtranslate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1410\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtable\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop_symbol\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop_symbol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mto_stop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mto_stop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcds\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgap\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgap\u001b[49m\n\u001b[1;32m   1411\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1412\u001b[0m )\n\u001b[1;32m   1413\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mid\u001b[39m, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m   1414\u001b[0m     answer\u001b[38;5;241m.\u001b[39mid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mid\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/biopython-env/lib/python3.12/site-packages/Bio/Seq.py:1620\u001b[0m, in \u001b[0;36m_SeqAbstractBaseClass.translate\u001b[0;34m(self, table, stop_symbol, to_stop, cds, gap)\u001b[0m\n\u001b[1;32m   1612\u001b[0m         warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1613\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPartial codon, len(sequence) not a multiple of three. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1614\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThis may become an error in future.\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   1615\u001b[0m             BiopythonWarning,\n\u001b[1;32m   1616\u001b[0m         )\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m Seq(\u001b[38;5;28;01mNone\u001b[39;00m, n \u001b[38;5;241m/\u001b[39m\u001b[38;5;241m/\u001b[39m \u001b[38;5;241m3\u001b[39m)\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m(\n\u001b[0;32m-> 1620\u001b[0m     \u001b[43m_translate_str\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop_symbol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mto_stop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcds\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgap\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgap\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1621\u001b[0m )\n",
      "File \u001b[0;32m~/miniconda3/envs/biopython-env/lib/python3.12/site-packages/Bio/Seq.py:2871\u001b[0m, in \u001b[0;36m_translate_str\u001b[0;34m(sequence, table, stop_symbol, to_stop, cds, pos_stop, gap)\u001b[0m\n\u001b[1;32m   2867\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CodonTable\u001b[38;5;241m.\u001b[39mTranslationError(\n\u001b[1;32m   2868\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSequence length \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mn\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m is not a multiple of three\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   2869\u001b[0m     )\n\u001b[1;32m   2870\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mstr\u001b[39m(sequence[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m3\u001b[39m:])\u001b[38;5;241m.\u001b[39mupper() \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m stop_codons:\n\u001b[0;32m-> 2871\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CodonTable\u001b[38;5;241m.\u001b[39mTranslationError(\n\u001b[1;32m   2872\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFinal codon \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00msequence[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m3\u001b[39m:]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m is not a stop codon\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   2873\u001b[0m     )\n\u001b[1;32m   2874\u001b[0m \u001b[38;5;66;03m# Don't translate the stop symbol, and manually translate the M\u001b[39;00m\n\u001b[1;32m   2875\u001b[0m sequence \u001b[38;5;241m=\u001b[39m sequence[\u001b[38;5;241m3\u001b[39m:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m3\u001b[39m]\n",
      "\u001b[0;31mTranslationError\u001b[0m: Final codon 'TTA' is not a stop codon"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, SimpleLocation\n",
    "import os\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/final/Crepis_callicephala.gb\"\n",
    "c_pur = \"plastomes/final/Crepis_purpurea.gb\"\n",
    "cich = \"plastomes/test/cichorieae_seqrecords.gb\"\n",
    "\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in [c_cal]:\n",
    "    print(acc)\n",
    "    \n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            if not \"angustana\" in record.annotations.get('organism', \"\"):\n",
    "                features = record.features\n",
    "                print(record.id, record.annotations.get('organism', \"\"), len(features) - 1)\n",
    "                \n",
    "                # check if there are features other than 'source'\n",
    "                if len(features) > 1:\n",
    "                    for feature in features:\n",
    "                        if feature.type == \"CDS\" and feature.qualifiers.get('gene', [\"\"])[0] == \"psbG\":\n",
    "                            print(feature)\n",
    "                            a = feature.extract(record)\n",
    "                            print(a)\n",
    "                            b = SeqFeature(SimpleLocation(49591, 49645, strand=-1), type=\"CDS\")\n",
    "                            print(b.extract(record))\n",
    "                            ls = \"ATGGTCTTAGCTCCTGAATATTCAGACAATAAAAAGAAAAAAGGAAAAAATTGA\"\n",
    "                            print(ls)\n",
    "                            print(f'L. sativa: {len(ls)}\\nC.callicephala: {len(b)}')\n",
    "                            c = b.translate(record, table=11)\n",
    "                            print(c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aligning protein PsbG against plastid genome sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['bash', 'aln_aa2na.sh', '-f', 'plastomes/refs/psbG_A0A5C0PXP7.fasta', '-r', 'plastomes/final/gb2sequin_out/c_callicephala_old/Crepis_callicephala___149980_bp____DNA.fsa.txt', '-o', 'aln_PsbG_2_cc.txt', '-t', '2'], returncode=0)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "pept = \"plastomes/refs/psbG_A0A5C0PXP7.fasta\"\n",
    "ref = \"plastomes/final/gb2sequin_out/c_callicephala_old/Crepis_callicephala___149980_bp____DNA.fsa.txt\"\n",
    "\n",
    "cmd = f'bash aln_aa2na.sh -f {pept} -r {ref} -o output.txt -t 2'\n",
    "\n",
    "import subprocess\n",
    "\n",
    "subprocess.run(['bash', 'aln_aa2na.sh', '-f', pept, \n",
    "                '-r', ref, \n",
    "                '-o', 'aln_PsbG_2_cc.txt', '-t', '2'], check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Successfully aligned 10 aa with only mismatch in range of 49645..49616 (reverse strand)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aligning genic na sequence against plastome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['bash', 'aln_na2na.sh', '-f', 'plastomes/refs/psbG_lac.fna', '-r', 'plastomes/final/gb2sequin_out/c_callicephala_old/Crepis_callicephala___149980_bp____DNA.fsa.txt', '-o', 'aln_psbG-lac_2_cc.txt', '-t', '2'], returncode=0)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = \"plastomes/refs/psbG_lac.fna\"\n",
    "ref = \"plastomes/final/gb2sequin_out/c_callicephala_old/Crepis_callicephala___149980_bp____DNA.fsa.txt\"\n",
    "\n",
    "import subprocess\n",
    "\n",
    "subprocess.run(['bash', 'aln_na2na.sh', '-f', gene, \n",
    "                '-r', ref, \n",
    "                '-o', 'aln_psbG-lac_2_cc.txt', '-t', '2'], check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ð£ Lactuca sativa Ð³ÐµÐ½ Ð½Ð° 5'-ÐºÐ¾Ð½Ñ†Ðµ Ð¸Ð¼ÐµÐµÑ‚ Ð¿Ð¾Ñ‡Ñ‚Ð¸ Ñ‚Ñƒ Ð¶Ðµ Ð¿Ð¾ÑÐ»ÐµÐ´Ð¾Ð²Ð°Ñ‚ÐµÐ»ÑŒÐ½Ð¾ÑÑ‚ÑŒ (Ñ Ð¾Ð´Ð½Ð¾Ð¹ Ð·Ð°Ð¼ÐµÐ½Ð¾Ð¹), Ñ‡Ñ‚Ð¾ Ð¸ Ñƒ Crepis callicephala. Ð”ÐµÐ»ÐµÑ†Ð¸Ð¸, ÐºÐ°Ðº Ð² ÑÐ»ÑƒÑ‡Ð°Ðµ Ñ Hypohaeris, Ð½ÐµÑ‚. Ð”Ð¾Ð²Ð¾Ð´Ð¸Ð¼ ÐµÐ³Ð¾ Ð´Ð¾ Ñ‚Ð¾Ð¹ Ð¶Ðµ Ð´Ð»Ð¸Ð½Ñ‹, Ñ‡Ñ‚Ð¾ Ñƒ L. sativa, Ñ€ÐµÐ´Ð°ÐºÑ‚Ð¸Ñ€ÑƒÑ Ð³Ñ€Ð°Ð½Ð¸Ñ†Ñ‹ ÑÑƒÑ‰Ð½Ð¾ÑÑ‚Ð¸."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. ERROR: valid [SEQ_FEAT.StartCodon] Illegal start codon (and 4 internal stops). Probably wrong genetic code [11] FEATURE: CDS: Acetyl-coenzyme A carboxylase carboxyl transferase subunit beta, chloroplastic; ACCase subunit beta; Acetyl-CoA carboxylase carboxyltransferase subunit beta; AccD [lcl|Crepis_callicephala:56772-57064] [lcl|Crepis_callicephala: raw, dna len= 149980] -> [lcl|Crepis_callicephala_31]\n",
    "\n",
    "**DESCRIPTION:** the gene should be marked as pseudogene, and the CDS removed from the annotation. Qualiviers `pseudo` and `pseudogene=â€unknown\"` might be added to both gene and CDS features.\n",
    "\n",
    "> Qualifier `pseudo` indicates that this feature is a non-functional version of the element named by the feature key.\n",
    ">\n",
    ">The qualifier `pseudo` should be used to describe non-functional \n",
    ">                 genes that are not formally described as pseudogenes, e.g. CDS \n",
    ">                 has no translation due to other reasons than pseudogenisation events.\n",
    ">                 Other reasons may include sequencing or assembly errors.\n",
    ">                 In order to annotate pseudogenes the qualifier /pseudogene= must be\n",
    ">                 used indicating the TYPE which can be taken from the INSDC controlled vocabulary \n",
    ">                 for pseudogenes.\n",
    "\n",
    "> Qualifier `pseudogene=` indicates that this feature is a pseudogene of the element named by the feature key. \n",
    "> where TYPE is one of the following: processed, unprocessed, unitary, allelic, unknown.\n",
    "> \n",
    "> unknown: the submitter does not know the method of pseudogenisation.\n",
    "\n",
    "4. ERROR: valid [SEQ_FEAT.InternalStop] 4 internal stops (and illegal start codon). Genetic code [11] FEATURE: CDS: Acetyl-coenzyme A carboxylase carboxyl transferase subunit beta, chloroplastic; ACCase subunit beta; Acetyl-CoA carboxylase carboxyltransferase subunit beta; AccD [lcl|Crepis_callicephala:56772-57064] [lcl|Crepis_callicephala: raw, dna len= 149980] -> [lcl|Crepis_callicephala_31]\n",
    "\n",
    "5. ERROR: valid [SEQ_FEAT.NoStop] Missing stop codon FEATURE: CDS: Acetyl-coenzyme A carboxylase carboxyl transferase subunit beta, chloroplastic; ACCase subunit beta; Acetyl-CoA carboxylase carboxyltransferase subunit beta; AccD [lcl|Crepis_callicephala:56772-57064] [lcl|Crepis_callicephala: raw, dna len= 149980] -> [lcl|Crepis_callicephala_31]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/final/Crepis_callicephala.gb\n",
      "Crepis_callicephala Crepis callicephala 347\n",
      "type: gene\n",
      "location: [56771:57064](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['accD']\n",
      "\n",
      "type: CDS\n",
      "location: [56771:57064](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['accD']\n",
      "    Key: product, Value: ['Acetyl-coenzyme A carboxylase carboxyl transferase subunit beta, chloroplastic; ACCase subunit beta; Acetyl-CoA carboxylase carboxyltransferase subunit beta; AccD']\n",
      "    Key: transl_table, Value: ['11']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, SimpleLocation\n",
    "import os\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/final/Crepis_callicephala.gb\"\n",
    "c_pur = \"plastomes/final/Crepis_purpurea.gb\"\n",
    "cich = \"plastomes/test/cichorieae_seqrecords.gb\"\n",
    "\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in [c_cal]:\n",
    "    print(acc)\n",
    "    \n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            if not \"angustana\" in record.annotations.get('organism', \"\"):\n",
    "                features = record.features\n",
    "                print(record.id, record.annotations.get('organism', \"\"), len(features) - 1)\n",
    "                \n",
    "                # check if there are features other than 'source'\n",
    "                if len(features) > 1:\n",
    "                    for feature in features:\n",
    "                        if feature.qualifiers.get('gene', [\"\"])[0] == \"accD\":\n",
    "                            print(feature)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. ERROR: valid [SEQ_FEAT.StartCodon] Illegal start codon used. Wrong genetic code [11] or protein should be partial FEATURE: CDS: photosystem II subunit L [lcl|Crepis_callicephala:c61997-61881] [lcl|Crepis_callicephala: raw, dna len= 149980] -> [lcl|Crepis_callicephala_37]\n",
    "\n",
    "**DESCRIPTION:** the psbL gene ORF starts in ACG codon. There is no option to point to the alternative start codon in genbank file.\n",
    "\n",
    "7. ERROR: valid [SEQ_FEAT.StartCodon] Illegal start codon (and 6 internal stops). Probably wrong genetic code [11] FEATURE: CDS: ribosomal protein S12 [(lcl|Crepis_callicephala:c136794-136768, c136232-136002, c67482-67369)] [lcl|Crepis_callicephala: raw, dna len= 149980] -> [lcl|Crepis_callicephala_46]\n",
    "\n",
    "**DESCRIPTION:** \n",
    "\n",
    "8. ERROR: valid [SEQ_FEAT.InternalStop] 6 internal stops (and illegal start codon). Genetic code [11] FEATURE: CDS: ribosomal protein S12 [(lcl|Crepis_callicephala:c136794-136768, c136232-136002, c67482-67369)] [lcl|Crepis_callicephala: raw, dna len= 149980] -> [lcl|Crepis_callicephala_46]\n",
    "\n",
    "**DESCRIPTION:** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/final/Crepis_callicephala.gb\n",
      "Crepis_callicephala Crepis callicephala 347\n",
      "[61880:61997](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 2\n",
      "Seq('ACGACACAATCAAACCCAAACGAACAAAATGTTGAATTGAATCGTACCAGCCTC...TAA')\n",
      "[61880:61997](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 2\n",
      "Seq('ACGACACAATCAAACCCAAACGAACAAAATGTTGAATTGAATCGTACCAGCCTC...TAA')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, SimpleLocation\n",
    "import os\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/final/Crepis_callicephala.gb\"\n",
    "c_pur = \"plastomes/final/Crepis_purpurea.gb\"\n",
    "cich = \"plastomes/test/cichorieae_seqrecords.gb\"\n",
    "\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in [c_cal]:\n",
    "    print(acc)\n",
    "    \n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            if not \"angustana\" in record.annotations.get('organism', \"\"):\n",
    "                features = record.features\n",
    "                print(record.id, record.annotations.get('organism', \"\"), len(features) - 1)\n",
    "                \n",
    "                # check if there are features other than 'source'\n",
    "                if len(features) > 1:\n",
    "                    for feature in features:\n",
    "                        if feature.qualifiers.get('gene', [\"\"])[0] == \"psbL\":\n",
    "                            print(feature.location)\n",
    "                            #print(feature.translate(record, table=11))\n",
    "                            print(feature.extract(record))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/final/Crepis_callicephala.gb\n",
      "\n",
      " Crepis_callicephala Crepis callicephala 347\n",
      "\n",
      "gene\n",
      "join{[136001:136794](-), [67368:67482](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 4\n",
      "Seq('TTATTTTGGCTTTTTTACCCCATATTGTAGGGTGGATCTCGAAAGATATGAAAG...TAT')\n",
      "\n",
      "CDS\n",
      "join{[136767:136794](-), [136001:136232](-), [67368:67482](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 3\n",
      "Seq('TTATTTTGGCTTTTTTACCCCATATTGAGAACGCCCTTGTTGACGATCCTTTAC...TAT')\n",
      "\n",
      "exon\n",
      "[67368:67482](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAT')\n",
      "\n",
      "gene\n",
      "join{[67368:67482](-), [95000:95793](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 4\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAA')\n",
      "\n",
      "CDS\n",
      "join{[67368:67482](-), [95562:95793](-), [95000:95027](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 3\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAA')\n",
      "\n",
      "exon\n",
      "[95000:95027](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('CAATATGGGGTAAAAAAGCCAAAATAA')\n",
      "\n",
      "intron\n",
      "[95027:95562](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('AGTGCGTTGTAGATTCTTATCCAAGACTTGTATCATTTGATGATGCCATGTGAA...CTA')\n",
      "\n",
      "exon\n",
      "[95562:95793](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('ACTATCACCCCAAAAAAACCAAACTCTGCCTTACGTAAAGTTGCCAGAGTACGA...TCT')\n",
      "\n",
      "exon\n",
      "[136001:136232](+)\n",
      "ID: Crepis_callicephala\n",
      "Name: Crepis_callicephala\n",
      "Description: Crepis callicephala chloroplast\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "Seq('ACTATCACCCCAAAAAAACCAAACTCTGCCTTACGTAAAGTTGCCAGAGTACGA...TCT')\n",
      "\n",
      "intron\n",
      "[136232:136767](+)\n",
      "ID: Crepis_callicephala\n",
      "Name: Crepis_callicephala\n",
      "Description: Crepis callicephala chloroplast\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "Seq('AGTGCGTTGTAGATTCTTATCCAAGACTTGTATCATTTGATGATGCCATGTGAA...CTA')\n",
      "\n",
      "exon\n",
      "[136767:136794](+)\n",
      "ID: Crepis_callicephala\n",
      "Name: Crepis_callicephala\n",
      "Description: Crepis callicephala chloroplast\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "Seq('CAATATGGGGTAAAAAAGCCAAAATAA')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, SimpleLocation\n",
    "import os\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/final/Crepis_callicephala.gb\"\n",
    "c_pur = \"plastomes/final/Crepis_purpurea.gb\"\n",
    "cich = \"plastomes/test/cichorieae_seqrecords.gb\"\n",
    "\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in [c_cal]:\n",
    "    print(acc)\n",
    "    \n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            if not \"angustana\" in record.annotations.get('organism', \"\"):\n",
    "                features = record.features\n",
    "                print(\"\\n\", record.id, record.annotations.get('organism', \"\"), len(features) - 1)\n",
    "                \n",
    "                # check if there are features other than 'source'\n",
    "                if len(features) > 1:\n",
    "                    for feature in features:\n",
    "                        if feature.qualifiers.get('gene', [\"\"])[0] == \"rps12\":\n",
    "                            #print(f'{feature.location}')\n",
    "                            #print(feature.translate(record, table=11))\n",
    "                            ex = feature.extract(record)\n",
    "                            print(f'\\n{feature.type}\\n{feature.location}\\n{ex}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/test/cichorieae_seqrecords.gb\n",
      "\n",
      " NC_061693.1 Lactuca sativa var. angustana 262\n",
      "\n",
      "gene\n",
      "join{[69571:69685](-), [97777:98020](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAG')\n",
      "\n",
      "CDS\n",
      "join{[69571:69685](-), [97777:98020](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAG')\n",
      "\n",
      "gene\n",
      "join{[69571:69685](+), [138808:139051](+)}\n",
      "ID: NC_061693.1\n",
      "Name: NC_061693\n",
      "Description: Lactuca sativa var. angustana chloroplast, complete genome\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "Seq('ATACACCCTAGTACATGTTCCTCGACGCTGAGGGCATCCCCGAAGCGCGGGGGA...TAG')\n",
      "\n",
      "CDS\n",
      "join{[69571:69685](+), [138808:139051](+)}\n",
      "ID: NC_061693.1\n",
      "Name: NC_061693\n",
      "Description: Lactuca sativa var. angustana chloroplast, complete genome\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "Seq('ATACACCCTAGTACATGTTCCTCGACGCTGAGGGCATCCCCGAAGCGCGGGGGA...TAG')\n",
      "\n",
      " NC_007578.1 Lactuca sativa 256\n",
      "\n",
      "gene\n",
      "join{[69591:69705](-), [97797:98040](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAG')\n",
      "\n",
      "CDS\n",
      "join{[69591:69705](-), [97797:98040](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAG')\n",
      "\n",
      "gene\n",
      "join{[69591:69705](-), [138828:139071](+)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAG')\n",
      "\n",
      "CDS\n",
      "join{[69591:69705](-), [138828:139071](+)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACGATTAAACAACTTATTAGAAATACAAGACAGCCAATTAGAAATGTC...TAG')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, SimpleLocation\n",
    "import os\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/final/Crepis_callicephala.gb\"\n",
    "c_pur = \"plastomes/final/Crepis_purpurea.gb\"\n",
    "cich = \"plastomes/test/cichorieae_seqrecords.gb\"\n",
    "\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in [cich]:\n",
    "    print(acc)\n",
    "    \n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            if \"sativa\" in record.annotations.get('organism', \"\") and \"NC_\" in record.id:\n",
    "                features = record.features\n",
    "                print(\"\\n\", record.id, record.annotations.get('organism', \"\"), len(features) - 1)\n",
    "                \n",
    "                # check if there are features other than 'source'\n",
    "                if len(features) > 1:\n",
    "                    for feature in features:\n",
    "                        if feature.qualifiers.get('gene', [\"\"])[0] == \"rps12\":\n",
    "                            #print(f'{feature.location}\\n')\n",
    "                            #print(feature.translate(record, table=11))\n",
    "                            ex = feature.extract(record)\n",
    "                            print(f'\\n{feature.type}\\n{feature.location}\\n{ex}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plastomes/refs/arabidopsis_thaliana.gb\n",
      "\n",
      " NC_001879.2 Nicotiana tabacum 368\n",
      "\n",
      "gene\n",
      "join{[72223:72337](-), [100061:100855](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 4\n",
      "Seq('ATGCCAACTATTAAACAACTTATTAGAAATACAAGACAGCCAATCAGAAATGTC...TAA')\n",
      "\n",
      "CDS\n",
      "join{[72223:72337](-), [100623:100855](-), [100061:100087](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 3\n",
      "Seq('ATGCCAACTATTAAACAACTTATTAGAAATACAAGACAGCCAATCAGAAATGTC...TAA')\n",
      "\n",
      "gene\n",
      "join{[72223:72337](-), [141774:142568](+)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 3\n",
      "Seq('ATGCCAACTATTAAACAACTTATTAGAAATACAAGACAGCCAATCAGAAATGTC...TAA')\n",
      "\n",
      "CDS\n",
      "join{[72223:72337](-), [141774:142006](+), [142542:142568](+)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 3\n",
      "Seq('ATGCCAACTATTAAACAACTTATTAGAAATACAAGACAGCCAATCAGAAATGTC...TAA')\n",
      "\n",
      "exon\n",
      "[72223:72337](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('ATGCCAACTATTAAACAACTTATTAGAAATACAAGACAGCCAATCAGAAATGTC...TAT')\n",
      "\n",
      "exon\n",
      "[100061:100087](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('AATATGGGGTAAAAAAGCCAAAATAA')\n",
      "\n",
      "intron\n",
      "[100087:100623](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('GTGCGTTGTAGATTCTTATCCAAGACTTGTATCATTTGATGATGCCATGTGAAT...TAC')\n",
      "\n",
      "exon\n",
      "[100623:100855](-)\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 1\n",
      "Seq('ACTATCACCCCAAAAAAACCAAACTCTGCCTTACGTAAAGTTGCCAGAGTACGA...CTA')\n",
      "\n",
      "exon\n",
      "[141774:142006](+)\n",
      "ID: NC_001879.2\n",
      "Name: NC_001879\n",
      "Description: Nicotiana tabacum plastid, complete genome\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "Seq('ACTATCACCCCAAAAAAACCAAACTCTGCCTTACGTAAAGTTGCCAGAGTACGA...CTA')\n",
      "\n",
      "exon\n",
      "[142542:142568](+)\n",
      "ID: NC_001879.2\n",
      "Name: NC_001879\n",
      "Description: Nicotiana tabacum plastid, complete genome\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "Seq('AATATGGGGTAAAAAAGCCAAAATAA')\n",
      "plastomes/refs/nicotiana_tabacum.gb\n",
      "\n",
      " NC_000932.1 Arabidopsis thaliana 258\n",
      "\n",
      "gene\n",
      "join{[69610:69724](-), [97998:98793](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACCATTAAACAACTTATTAGAAATACAAGACAGCCAATCCGAAACGTC...TAA')\n",
      "\n",
      "CDS\n",
      "join{[69610:69724](-), [98561:98793](-), [97998:98024](-)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACCATTAAACAACTTATTAGAAATACAAGACAGCCAATCCGAAACGTC...TAA')\n",
      "\n",
      "gene\n",
      "join{[69610:69724](-), [139855:140650](+)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACCATTAAACAACTTATTAGAAATACAAGACAGCCAATCCGAAACGTC...TAA')\n",
      "\n",
      "CDS\n",
      "join{[69610:69724](-), [139855:140087](+), [140624:140650](+)}\n",
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('ATGCCAACCATTAAACAACTTATTAGAAATACAAGACAGCCAATCCGAAACGTC...TAA')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqRecord, SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, SimpleLocation\n",
    "import os\n",
    "\n",
    "# variables\n",
    "c_cal = \"plastomes/final/Crepis_callicephala.gb\"\n",
    "c_pur = \"plastomes/final/Crepis_purpurea.gb\"\n",
    "cich = \"plastomes/test/cichorieae_seqrecords.gb\"\n",
    "a_thal = \"plastomes/refs/arabidopsis_thaliana.gb\"\n",
    "n_tab = \"plastomes/refs/nicotiana_tabacum.gb\"\n",
    "\n",
    "\n",
    "# Do initial parsing\n",
    "for acc in [a_thal, n_tab]:\n",
    "    print(acc)\n",
    "    \n",
    "    with open(acc, \"r\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"genbank\"):\n",
    "            if \"NC_\" in record.id:\n",
    "                features = record.features\n",
    "                print(\"\\n\", record.id, record.annotations.get('organism', \"\"), len(features) - 1)\n",
    "                \n",
    "                # check if there are features other than 'source'\n",
    "                if len(features) > 1:\n",
    "                    for feature in features:\n",
    "                        if feature.qualifiers.get('gene', [\"\"])[0] == \"rps12\":\n",
    "                            #print(f'{feature.location}\\n')\n",
    "                            #print(feature.translate(record, table=11))\n",
    "                            ex = feature.extract(record)\n",
    "                            print(f'\\n{feature.type}\\n{feature.location}\\n{ex}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
