{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d8b4b753",
   "metadata": {},
   "source": [
    "This notebook is for troubleshooting plastomes using NCBI-provided software pipelines. It was a suggestion from NCBI team rejected the annotation twice after uploading."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d2cf034",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Defining variables for files\n",
    "template = 'plastomes/final/template.sbt'\n",
    "gb = 'plastomes/final/Crepis_callicephala.gb'\n",
    "fasta = 'plastomes/in/Crepis_callicephala.fasta'\n",
    "\n",
    "# command itself:\n",
    "# table2asn -t template.sbt -i sequence.fsa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bfd7c794",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File plastomes/in/Crepis_callicephala.fasta does exist.\n"
     ]
    }
   ],
   "source": [
    "# check for fasta\n",
    "\n",
    "from Bio import SeqIO, SeqRecord\n",
    "\n",
    "try:\n",
    "    with open(fasta, \"r\") as fasta_file:\n",
    "        print(f\"File {fasta} does exist.\")\n",
    "except FileNotFoundError:\n",
    "    print(f\"File '{fasta}' not found.\\nConverting genbank file '{gb}' to FASTA...\")\n",
    "    try:\n",
    "        first_record = next(SeqIO.parse(gb, \"genbank\"))\n",
    "        SeqIO.write(first_record, fasta, \"fasta\")\n",
    "        print(f\"Successfully created '{fasta}'.\")\n",
    "    except FileNotFoundError:\n",
    "        print(f\"Error: The source GenBank file '{gb}' was not found.\")\n",
    "    except StopIteration:\n",
    "        print(f\"Error: The GenBank file '{gb}' is empty.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35dde3bb",
   "metadata": {},
   "source": [
    "## Run `table2asn` program\n",
    "\n",
    "There are some recommendations on NCBI site that were not illuminated at Readme file of the program.\n",
    "`table2asn` will recognize files with **the same basename** as the input sequence file. Sequences that are part of a plasmid, or an organellar chromosome, or specific nuclear chromosomes need to have that information included in the fasta definition line, in these formats:\n",
    "\n",
    "- [location=mitochondrion]\n",
    "- [location=chloroplast]\n",
    "\n",
    "Sequences that are a complete circular chromosome or plasmid need to have the circular topology and the completeness included.\n",
    "\n",
    "- [topology=circular] [completeness=complete]\n",
    "- [topology=circular] gap at end, not circularized\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "04cf3c4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating symbolic links:\n",
      "Symlink for 'plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA.tbl' is already exist.\n"
     ]
    }
   ],
   "source": [
    "# preparing directory\n",
    "# all the files should be stored at the same directory\n",
    "import os\n",
    "\n",
    "project_dir = 'plastomes/table2asn'\n",
    "tbl_file = \"plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA.tbl\"\n",
    "sqn_file = \"plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA(1).sqn\"\n",
    "\n",
    "SHORT_NAMES = {\n",
    "    \"Crepis_callicephala\": \"cc\",\n",
    "    \"Crepis_purpurea\": \"cp\",\n",
    "}\n",
    "\n",
    "def create_symlinks(project_dir: str, sourcefile: str):\n",
    "    \"\"\"\n",
    "    Create symlinks for files to process.\n",
    "    \"\"\"\n",
    "    basename = os.path.basename(sourcefile)\n",
    "    ext= basename.rsplit(\".\")[-1]\n",
    "    #print(\"ext\", ext)\n",
    "\n",
    "    # handle sql and tbl names\n",
    "    basename = str(basename).rsplit(\"___\")[0]\n",
    "\n",
    "    #print(\"basename\", basename)\n",
    "    species = basename.rsplit(\".\")[0]\n",
    "    #print(\"species\", species)\n",
    "    \n",
    "    label = SHORT_NAMES[species]\n",
    "    #print(\"label\", label)\n",
    "    target_dir = os.path.join(project_dir, label)\n",
    "    #print(\"target_dir\", target_dir)\n",
    "    abs_target_dir = os.path.abspath(target_dir)\n",
    "    # filename should match sequence id in FASTA header\n",
    "    target_filename = os.path.join(abs_target_dir, f\"{species}.{ext}\")\n",
    "    \n",
    "    #print(\"target_filename\", target_filename)\n",
    "    abs_sourcefile = os.path.abspath(sourcefile)\n",
    "\n",
    "    try:\n",
    "        os.path.exists(sourcefile)\n",
    "        try:\n",
    "            # fixing first relative paths\n",
    "            os.symlink(abs_sourcefile, target_filename)\n",
    "            print(f\"Symlink '{target_filename}' for '{abs_sourcefile}' was successfully created.\")\n",
    "        except FileExistsError:\n",
    "            print(f\"Symlink for '{sourcefile}' is already exist.\")\n",
    "        except PermissionError:\n",
    "            print(f\"You have no permissions for this action.\")\n",
    "    except FileExistsError:\n",
    "        print(f\"Source file '{sourcefile}' does not exist.\")\n",
    "\n",
    "\n",
    "print(\"Creating symbolic links:\")\n",
    "# create_symlinks(project_dir, gb) #  this is output file\n",
    "# create_symlinks(project_dir, fasta) # the header line should be modified\n",
    "create_symlinks(project_dir, tbl_file)\n",
    "#create_symlinks(project_dir, sqn_file) #  this is output file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7558e3a5",
   "metadata": {},
   "source": [
    "The header line of fasta file should follow definition line conventions. To keep previous files intact but change required data, the file should be just copied instead of linking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a7cca631",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "destination: plastomes/table2asn/cc/Crepis_callicephala.fsa\n"
     ]
    }
   ],
   "source": [
    "import shutil\n",
    "\n",
    "destination_directory = f\"{project_dir}/cc\"\n",
    "dest_file = f\"{destination_directory}/{os.path.basename(fasta).rsplit(\".\")[0]}.fsa\"\n",
    "print(\"destination:\", dest_file)\n",
    "\n",
    "if not os.path.exists(dest_file):\n",
    "    try:\n",
    "        shutil.copy(fasta, dest_file)\n",
    "        print(f\"File '{fasta}' copied to '{destination_directory}' successfully.\")\n",
    "    except FileNotFoundError:\n",
    "        print(\"Source file not found.\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3bf6b370",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[organism=Crepis callicephala] [location=chloroplast] [topology=circular] [completeness=complete] [gcode=11]\n",
      "Checking input directory...\n",
      "  plastomes/table2asn/cc is present: True\n",
      "Checking input files...\n",
      "  plastomes/final/template.sbt is present: True\n",
      "  Crepis_callicephala.tbl is present: True\n",
      "  Crepis_callicephala.fsa is present: True\n",
      "Running: conda run -n table2asn-env table2asn -indir /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/table2asn/cc -t /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/final/template.sbt -j '[organism=Crepis callicephala] [location=chloroplast] [topology=circular] [completeness=complete] [gcode=11]' -V vb -verbose\n",
      "STDOUT: \n",
      "STDERR: This copy of table2asn is more than 1 year old. Please download the current version if it is newer.\n",
      "Will be using one threads\n",
      "Recognized annotation format: five-column feature table\n",
      "Falling back on built-in data for popular organisms.\n",
      "\n",
      "\n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "# testing table2asn\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "\n",
    "# Configuration\n",
    "CONDA_ENV_NAME = \"plastome_post-env\"\n",
    "INPUT_SUBDIR = \"cc\"\n",
    "\n",
    "location = \"[location=chloroplast]\"\n",
    "topology = \"[topology=circular]\"\n",
    "completeness = \"[completeness=complete]\"\n",
    "organism = \"[organism=Crepis callicephala]\"\n",
    "code = \"[gcode=11]\"\n",
    "pcode = \"[gpcode=11]\"\n",
    "\n",
    "src_qualifiers = f\"{organism} {location} {topology} {completeness} {code}\"\n",
    "print(src_qualifiers)\n",
    "\n",
    "input_dir = f\"{project_dir}/cc\"\n",
    "# Check inputs\n",
    "print(f\"Checking input directory...\\n  {input_dir} is present: {os.path.isdir(input_dir)}\")\n",
    "print(f\"Checking input files...\\n  {template} is present: {os.path.isfile(template)}\")\n",
    "present_flag = False\n",
    "for i in os.listdir(input_dir):\n",
    "    file = os.path.join(input_dir, i)\n",
    "    present_flag = os.path.exists(file)\n",
    "    print(f\"  {i} is present: {present_flag}\")\n",
    "    if present_flag == False:\n",
    "        print(f\"    relative path: {file}\")\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "# table2asn -indir {project_dir}/cc -t {template} -j {comment}\n",
    "\n",
    "cmd = [\n",
    "    \"conda\", \"run\", \"-n\", \"table2asn-env\",\n",
    "    \"table2asn\",\n",
    "    \"-indir\", str(os.path.abspath(input_dir)),\n",
    "    \"-t\", str(os.path.abspath(template)),\n",
    "    \"-j\", f\"'{src_qualifiers}'\",\n",
    "    \"-V\", \"vb\",\n",
    "    \"-verbose\",\n",
    "]\n",
    "\n",
    "cmd_version = [\n",
    "    \"conda\", \"run\", \"-n\", \"table2asn-env\",\n",
    "    \"table2asn\",\n",
    "    \"-version\",\n",
    "]\n",
    "\n",
    "print(f\"Running: {' '.join(cmd)}\")\n",
    "#print(f\"Running: {' '.join(cmd_version)}\")\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70c4e93",
   "metadata": {},
   "source": [
    "The `table2asn` did not performed well. The working directory does not contain any output files.\n",
    "There the \"This copy of table2asn is more than 1 year old. Please download the current version if it is newer.\" message appeared in STDERR. Trying to launch binary executable file from the fresh release might be a solution.\n",
    "\n",
    "## Solution\n",
    "Thwe problem was copying fasta file with `.fasta` file extension. After changing it to `.fsa` the `table2asn` performed well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e3282652",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: table2asn: 1.29.324\n",
      "\n",
      "STDERR: \n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "table2asn = \"bin/table2asn.linux64\"\n",
    "\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "cmd_version = [\n",
    "    table2asn,\n",
    "    \"-version\",\n",
    "]\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd_version, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1e2b054a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: bin/table2asn.linux64 -indir /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/table2asn/cc -t /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/final/template.sbt -V vb -verbose -Z\n",
      "STDOUT: \n",
      "STDERR: Will be using one threads\n",
      "Recognized annotation format: five-column feature table\n",
      "\n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "project_dir = 'plastomes/table2asn'\n",
    "input_dir = f\"{project_dir}/cc\"\n",
    "\n",
    "cmd = [\n",
    "    table2asn,\n",
    "    \"-indir\", str(os.path.abspath(input_dir)),\n",
    "    \"-t\", str(os.path.abspath(template)),\n",
    "    \"-V\", \"vb\",\n",
    "    \"-verbose\",\n",
    "    \"-Z\",\n",
    "]\n",
    "\n",
    "print(f\"Running: {' '.join(cmd)}\")\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "799fe050",
   "metadata": {},
   "source": [
    "## Summarizing results\n",
    "There are 4 errors in Crepis callicephala genome of two types. The reference genome annotations were validated via GB2Sequin and compared to C. callicephala plastid genome.\n",
    "### Errors\n",
    "\n",
    "| species | accession | SEQ_INST.BadProteinStart | SEQ_FEAT.StartCodon |\n",
    "| ------- | --------- | ------------------------ | ------------------- |\n",
    "| Crepis callicephala | - | psbL, ndhD | \"photosystem II subunit L\", \"NADH dehydrogenase subunit D\" |\n",
    "| Lactuca sativa | NC_007578 | psbL, ndhD | - |\n",
    "| Lactuca sativa cv. Ramosa | PP999684 | psbL | - |\n",
    "| Nicotiana tabacum | NC_001879 | psbL | - |\n",
    "| Arabidopsis thaliana | NC_000932 | ndhD | - |\n",
    "| Oryza sativa | NC_031333 | rpl2, rpl2 | - |\n",
    "\n",
    "At least each fatal error should be checked manually. One of the solution for 'illegal start codon' type error might be exception defining in annotation.\n",
    "\n",
    "### Warnings\n",
    "In Crepis callicephala genome validation report\n",
    "\n",
    "| Warning | Feature | Feature_description | Location |\n",
    "| - | - | - | - |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S12 <133> | rps12:[c67482-67369, 136002-136794] |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | Ycf15 protein <158> | ycf15:c96959-96768 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S7 <159> | rps7:c94947-94480 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | NADH dehydrogenase subunit B <160> | ndhB:c94189-91988 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | Ycf2 protein <161> | ycf2:84073-90927 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L23 <162> | rpl23:c83722-83441 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L2 <163> | rpl2:c83422-81933 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c7316-7244 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c7316-7244 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c8577-8490 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c8577-8490 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c11817-11734 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c11817-11734 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c11992-11921 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c11992-11921 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c29889-29818 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c29889-29818 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c35029-34947 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Gly | 35855-35925 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | 44778-44864 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | c83961-83888 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | 110967-111046 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | (c132067-132025, c131268-131234) |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | 147834-147908 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | c1794-1759 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | c4320-1795 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c4320-1795 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | c4358-4321 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c6229-5377 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 16596-17318 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 16596-17318 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 26949-27665 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 26949-27665 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c30852-30143 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=2 | c42897-42165 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c43820-43128 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 46807-47243 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 46807-47243 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | 47244-47293 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c51689-51122 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | c51727-51690 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=2 | c68490-67891 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=2 | c68490-67891 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c69581-68783 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 72584-73348 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c80268-79232 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c83031-82367 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c93412-92744 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c95562-95028 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c95562-95028 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | 99728-99770 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | 99771-100526 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 99771-100526 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 99771-100526 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | 100527-100561 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 100664-101483 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 100664-101483 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c118053-116999 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c118053-116999 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c131131-130312 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c131131-130312 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c132024-131269 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c132024-131269 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 136233-136767 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 136233-136767 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 138383-139051 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 148764-149428 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S16 <5> | (c6269-6230, c5376-5150) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | RNA polymerase subunit beta' <17> | (16165-16595, 17319-18963) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | RNA polymerase subunit beta' <17> | (16165-16595, 17319-18963) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | CF0 subunit I <27> | (26805-26948, 27666-28076) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | CF0 subunit I <27> | (26805-26948, 27666-28076) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | photosystem I assembly factor I <43> | (c43944-43821, c43127-42898, c42164-42012) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | photosystem I assembly factor I <43> | (c43944-43821, c43127-42898, c42164-42012) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | cytochrome b6 <99> | (72578-72583, 73349-73990) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L16 <115> | (c80277-80269, c79231-78833) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L2 <123> | (c83422-83032, c82366-81933) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit B <129> | (c94189-93413, c92743-91988) |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | NADH dehydrogenase subunit A <152> | (c118605-118054, c116998-116459) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit A <152> | (c118605-118054, c116998-116459) |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf15 protein <158> | 134836-135027 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S7 <159> | 136848-137315 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | NADH dehydrogenase subunit B <160> | (137606-138382, 139052-139807) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit B <160> | (137606-138382, 139052-139807) |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf2 protein <161> | c147722-140868 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L23 <162> | 148073-148354 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L2 <163> | (148373-148763, 149429-149862) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L2 <163> | (148373-148763, 149429-149862) |\n",
    "\n",
    "\n",
    "### Notes\n",
    "For C. Callicephala, lines of type 'Note' are absent in table2asn validation output.\n",
    "\n",
    "### Error types\n",
    "#### CDSgeneRange\n",
    "A CDS is overlapped by a gene feature, but is not completely contained by it.\n",
    "This may be an annotation error.\n",
    "\n",
    "#### GeneXrefWithoutGene\n",
    "This feature has a gene xref, but there is no equivalent gene feature anywhere\n",
    "on the record.\n",
    "\n",
    "#### GeneXrefStrandProblem\n",
    "This feature has a gene xref that points to a gene on the wrong strand.\n",
    "\n",
    "#### MissingTrnaAA\n",
    "The tRNA encoded amino acid is not set. *Explanation*: The amino acid that the tRNA carries is not included. *Suggestion*: Include the amino acid as the product of the tRNA. If the amino acid of a tRNA is unknown, use tRNA-Xxx as the product.\n",
    "\n",
    "#### SEQ_FEAT.StartCodon and SEQ_INST.BadProteinStart\n",
    "\n",
    "*Explanation*: The StartCodon and BadProteinStart errors are produced when the CDS is not marked as partial at its 5′ end and does not begin with a start codon.\n",
    "\n",
    "*Suggestion*: Use the correct genetic code to get the correct translations. For example, include \\[gcode=11\\] for prokaryotic genome submissions. Other possible fixes include: extend the CDS to the start codon, or mark the 5′ end as partial (and extend the CDS to the end of the sequence for prokaryotic sequences), or add the /pseudo qualifier to the gene to indicate that the CDS cannot be translated.\n",
    "\n",
    "##### StartCodon\n",
    "An illegal start codon was used. Some possible explanations are: (1) the\n",
    "wrong genetic code may have been selected; (2) the wrong reading frame may\n",
    "be in use; or (3) the coding region may be incomplete at the 5' end, in\n",
    "which case a partial location should be indicated.\n",
    "\n",
    "##### BadProteinStart\n",
    "A gap symbols was found at the start of this protein Bioseq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "125bc582",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S12 <133> | rps12:[c67482-67369, 136002-136794] |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | Ycf15 protein <158> | ycf15:c96959-96768 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S7 <159> | rps7:c94947-94480 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | NADH dehydrogenase subunit B <160> | ndhB:c94189-91988 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | Ycf2 protein <161> | ycf2:84073-90927 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L23 <162> | rpl23:c83722-83441 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L2 <163> | rpl2:c83422-81933 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c7316-7244 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c7316-7244 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c8577-8490 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c8577-8490 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c11817-11734 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c11817-11734 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c11992-11921 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c11992-11921 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c29889-29818 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c29889-29818 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c35029-34947 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Gly | 35855-35925 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | 44778-44864 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | c83961-83888 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | 110967-111046 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | (c132067-132025, c131268-131234) |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | 147834-147908 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | c1794-1759 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | c4320-1795 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c4320-1795 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | c4358-4321 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c6229-5377 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 16596-17318 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 16596-17318 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 26949-27665 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 26949-27665 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c30852-30143 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=2 | c42897-42165 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c43820-43128 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 46807-47243 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 46807-47243 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | 47244-47293 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c51689-51122 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | c51727-51690 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=2 | c68490-67891 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=2 | c68490-67891 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c69581-68783 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 72584-73348 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c80268-79232 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c83031-82367 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c93412-92744 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c95562-95028 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c95562-95028 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | 99728-99770 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | 99771-100526 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 99771-100526 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 99771-100526 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | 100527-100561 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 100664-101483 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 100664-101483 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c118053-116999 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c118053-116999 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c131131-130312 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c131131-130312 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c132024-131269 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c132024-131269 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 136233-136767 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 136233-136767 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 138383-139051 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 148764-149428 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S16 <5> | (c6269-6230, c5376-5150) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | RNA polymerase subunit beta' <17> | (16165-16595, 17319-18963) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | RNA polymerase subunit beta' <17> | (16165-16595, 17319-18963) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | CF0 subunit I <27> | (26805-26948, 27666-28076) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | CF0 subunit I <27> | (26805-26948, 27666-28076) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | photosystem I assembly factor I <43> | (c43944-43821, c43127-42898, c42164-42012) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | photosystem I assembly factor I <43> | (c43944-43821, c43127-42898, c42164-42012) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | cytochrome b6 <99> | (72578-72583, 73349-73990) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L16 <115> | (c80277-80269, c79231-78833) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L2 <123> | (c83422-83032, c82366-81933) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit B <129> | (c94189-93413, c92743-91988) |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | NADH dehydrogenase subunit A <152> | (c118605-118054, c116998-116459) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit A <152> | (c118605-118054, c116998-116459) |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf15 protein <158> | 134836-135027 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S7 <159> | 136848-137315 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | NADH dehydrogenase subunit B <160> | (137606-138382, 139052-139807) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit B <160> | (137606-138382, 139052-139807) |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf2 protein <161> | c147722-140868 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L23 <162> | 148073-148354 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L2 <163> | (148373-148763, 149429-149862) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L2 <163> | (148373-148763, 149429-149862) |\n"
     ]
    }
   ],
   "source": [
    "# parsing text file with validation warnings to print it in structured format.\n",
    "\n",
    "file_2_parse = 'plastomes/table2asn/cc/Crepis_callicephala.val'\n",
    "\n",
    "with open(file_2_parse, \"r\") as handle:\n",
    "    for line in handle:\n",
    "        warning = \"-\"\n",
    "        feature = \"-\"\n",
    "        gene = \"-\"\n",
    "        location = \"-\"\n",
    "        if \"Warning:\" in line and \"SEQ_FEAT\" in line:\n",
    "            line = line.split(\" \", maxsplit=3)\n",
    "            warning = line[2].strip(\"[]\")\n",
    "            feature = line[-1].rsplit('FEATURE: ')[-1].split(\": \")[0]\n",
    "            gene = line[-1].rsplit('FEATURE: ')[-1].split(\"[\")[0].split(\": \")[-1]\n",
    "            #print(line)\n",
    "            location = line[-1].split(\"[\", maxsplit=1)[1].split(\"] \")[0]\n",
    "            # sanitizing values\n",
    "            variables = [warning, feature, gene, location]\n",
    "            for i, value in enumerate(variables):\n",
    "                value = value.strip(\" \")\n",
    "                if not value or value.isspace():\n",
    "                    variables[i] = \"-\"\n",
    "                elif \"|\" in value:\n",
    "                    variables[i] = value.replace(\"lcl|Crepis_callicephala:\", \"\")\n",
    "                else:\n",
    "                    variables[i] = value\n",
    "            warning, feature, gene, location = variables\n",
    "            # printing in md-compatible format\n",
    "            print(f\"| {warning} | {feature} | {gene} | {location} |\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "422b2a04",
   "metadata": {},
   "source": [
    "## Checking tRNA data\n",
    "### Perform re-assess using Aragorn v1.2.41\n",
    "The code to use:\n",
    "```shell\n",
    "aragorn -v -e -gcbact -c -i -o $output $fasta\n",
    "```\n",
    "`-gcbact`   Use Bacterial/Plant chloroplast genetic code.\n",
    "\n",
    "`-c`        Assume that each sequence has a circular topology. Search wraps around each end.\n",
    "\n",
    "`-d`        Double. Search both strands of each sequence. Default setting\n",
    "\n",
    "`-i`        Search for tRNA genes with introns in anticodon loop with maximum length 3000 bases. Minimum intron length is 0 bases.\n",
    "\n",
    "`-e`        Print out score for each reported gene.\n",
    "\n",
    "`-v`        Verbose. Prints out information during search to STDERR.\n",
    "\n",
    "`-o <outfile>`    Print output to <outfile>. If <outfile> already exists, it is overwritten. By default all output goes to stdout."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "099edc7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: \n",
      "STDERR: Crepis_callicephala ' '.\n",
      "149980 nucleotides in sequence\n",
      "Mean G+C content = 37.7%\n",
      "Using bacterial/plant chloroplast genetic code\n",
      "Searching from 3120 to 16881\n",
      "tRNA-Cys(gca) at [9331,9402] (107.847)\n",
      "tRNA-Cys(gca) [9331,9402] (116.002) replacing tRNA-Cys(gca) [9331,9402] (107.847)\n",
      "tRNA-Glu(ttc) at c[11921,11992] (104.658)\n",
      "tRNA-Glu(ttc) c[11920,11993] (115.411) replacing tRNA-Glu(ttc) c[11921,11992] (104.658)\n",
      "tRNA-Tyr(gta) at c[11734,11817] (105.503)\n",
      "tRNA-Tyr(gta) c[11732,11818] (117.555) replacing tRNA-Tyr(gta) c[11734,11817] (105.503)\n",
      "tRNA-Asp(gtc) at c[11547,11620] (108.811)\n",
      "tRNA-Asp(gtc) c[11545,11621] (117.885) replacing tRNA-Asp(gtc) c[11547,11620] (108.811)\n",
      "tRNA-Ser(gct) at c[8491,8577] (104.217)\n",
      "tRNA-Ser(gct) c[8491,8578] (116.445) replacing tRNA-Ser(gct) c[8491,8577] (104.217)\n",
      "tRNA-Ser(gct) c[8490,8577] (123.247) replacing tRNA-Ser(gct) c[8491,8578] (116.445)\n",
      "tRNA-Gln(ttg) at c[7245,7316] (105.114)\n",
      "tRNA-Gln(ttg) c[7245,7316] (105.602) replacing tRNA-Gln(ttg) c[7245,7316] (105.114)\n",
      "tRNA-Gln(ttg) c[7245,7316] (116.786) replacing tRNA-Gln(ttg) c[7245,7316] (105.602)\n",
      "Searching from 16882 to 30643\n",
      "tRNA-Thr(cgt) at c[30094,30875] (101.165)\n",
      "tRNA-Ser(cga) c[30094,30875] (101.783) replacing tRNA-Thr(cgt) c[30094,30875] (101.165)\n",
      "tRNA-Arg(tct) at c[29818,29889] (108.377)\n",
      "tRNA-Arg(tct) c[29818,29889] (117.452) replacing tRNA-Arg(tct) c[29818,29889] (108.377)\n",
      "Searching from 30644 to 44405\n",
      "tRNA-Gly(gcc) at [35856,35927] (107.727)\n",
      "tRNA-Gly(gcc) [35855,35926] (114.476) replacing tRNA-Gly(gcc) [35856,35927] (107.727)\n",
      "tRNA-Met(cat) at c[36113,36184] (111.245)\n",
      "tRNA-Met(cat) c[36112,36185] (118.102) replacing tRNA-Met(cat) c[36113,36184] (111.245)\n",
      "tRNA-Ser(tga) at c[34948,35029] (102.014)\n",
      "tRNA-Ser(tga) c[34948,35030] (110.345) replacing tRNA-Ser(tga) c[34948,35029] (102.014)\n",
      "tRNA-Met(cat) at c[34947,36184] (102.315)\n",
      "tRNA-Ser(tga) c[34947,35029] (121.044) replacing tRNA-Ser(tga) c[34948,35030] (110.345)\n",
      "Searching from 44406 to 58167\n",
      "tRNA-Ser(gga) at [44778,44863] (107.361)\n",
      "tRNA-Ser(gga) [44777,44863] (116.992) replacing tRNA-Ser(gga) [44778,44863] (107.361)\n",
      "tRNA-Ala(agc) at [44139,44864] (103.107)\n",
      "tRNA-Ser(gga) [44778,44864] (121.845) replacing tRNA-Ser(gga) [44777,44863] (116.992)\n",
      "tRNA-Leu(taa) at [46770,47293] (100.172)\n",
      "tRNA-Leu(taa) [46769,47293] (105.152) replacing tRNA-Leu(taa) [46770,47293] (100.172)\n",
      "tRNA-Leu(taa) [46770,47293] (111.808) replacing tRNA-Leu(taa) [46769,47293] (105.152)\n",
      "tRNA-Phe(gaa) at [47609,47682] (106.321)\n",
      "tRNA-Phe(gaa) [47609,47682] (115.774) replacing tRNA-Phe(gaa) [47609,47682] (106.321)\n",
      "tRNA-Met(cat) at [51914,51986] (111.705)\n",
      "tRNA-Ile(aat) at c[51085,51727] (100.151)\n",
      "tRNA-Val(tac) c[51085,51727] (103.348) replacing tRNA-Ile(aat) c[51085,51727] (100.151)\n",
      "tRNA-Ile(aat) c[51085,51727] (107.89) replacing tRNA-Val(tac) c[51085,51727] (103.348)\n",
      "tRNA-Val(tac) c[51085,51727] (111.087) replacing tRNA-Ile(aat) c[51085,51727] (107.89)\n",
      "tRNA-Thr(tgt) at c[46146,46218] (108.306)\n",
      "tRNA-Glu(ctc) at c[46146,47213] (100.298)\n",
      "tRNA-Thr(tgt) c[46145,46219] (122.683) replacing tRNA-Thr(tgt) c[46146,46218] (108.306)\n",
      "Searching from 58168 to 71929\n",
      "tRNA-Pro(tgg) at c[64379,64454] (105.888)\n",
      "tRNA-Pro(tgg) c[64379,64454] (119.886) replacing tRNA-Pro(tgg) c[64379,64454] (105.888)\n",
      "tRNA-Trp(cca) at c[64142,64215] (106.244)\n",
      "tRNA-Trp(cca) c[64142,64215] (116.997) replacing tRNA-Trp(cca) c[64142,64215] (106.244)\n",
      "Searching from 71930 to 85691\n",
      "tRNA-Met(cat) at c[83887,83961] (102.061)\n",
      "tRNA-Met(cat) c[83887,83961] (108.809) replacing tRNA-Met(cat) c[83887,83961] (102.061)\n",
      "Searching from 85692 to 99453\n",
      "tRNA-Val(gac) at [97644,97715] (113.371)\n",
      "tRNA-Leu(caa) at c[91330,91410] (109.768)\n",
      "tRNA-Leu(caa) c[91329,91411] (118.842) replacing tRNA-Leu(caa) c[91330,91410] (109.768)\n",
      "Searching from 99454 to 113215\n",
      "tRNA-Thr(cgt) at [99728,100561] (101.702)\n",
      "tRNA-Glu(ttc) [99728,100561] (104.79) replacing tRNA-Thr(cgt) [99728,100561] (101.702)\n",
      "tRNA-Arg(tcg) [99728,100561] (106.389) replacing tRNA-Glu(ttc) [99728,100561] (104.79)\n",
      "tRNA-Ile(gat) [99728,100561] (108.501) replacing tRNA-Arg(tcg) [99728,100561] (106.389)\n",
      "tRNA-Thr(cgt) [99728,100561] (109.821) replacing tRNA-Ile(gat) [99728,100561] (108.501)\n",
      "tRNA-Glu(ttc) [99728,100561] (112.908) replacing tRNA-Thr(cgt) [99728,100561] (109.821)\n",
      "tRNA-Ala(tgc) at [100626,101518] (101.592)\n",
      "tRNA-Ala(tgc) [100626,101518] (101.595) replacing tRNA-Ala(tgc) [100626,101518] (101.592)\n",
      "tRNA-Leu(gag) [100625,101518] (101.613) replacing tRNA-Ala(tgc) [100626,101518] (101.595)\n",
      "tRNA-Ala(tgc) [100625,101518] (104.353) replacing tRNA-Leu(gag) [100625,101518] (101.613)\n",
      "tRNA-Ala(tgc) [100625,101518] (104.355) replacing tRNA-Ala(tgc) [100625,101518] (104.353)\n",
      "tRNA-Leu(gag) [100626,101518] (108.269) replacing tRNA-Ala(tgc) [100625,101518] (104.355)\n",
      "tRNA-Ala(tgc) [100626,101518] (111.009) replacing tRNA-Leu(gag) [100626,101518] (108.269)\n",
      "tRNA-Ala(tgc) [100626,101518] (111.011) replacing tRNA-Ala(tgc) [100626,101518] (111.009)\n",
      "tRNA-Arg(acg) at [105301,105375] (100.692)\n",
      "tRNA-Arg(acg) [105301,105375] (115.882) replacing tRNA-Arg(acg) [105301,105375] (100.692)\n",
      "tRNA-Leu(tag) at [110967,111046] (107.517)\n",
      "tRNA-Leu(tag) [110967,111046] (120.595) replacing tRNA-Leu(tag) [110967,111046] (107.517)\n",
      "tRNA-Asn(gtt) at c[105869,105940] (111.726)\n",
      "tRNA-Asn(gtt) c[105868,105941] (123.398) replacing tRNA-Asn(gtt) c[105869,105940] (111.726)\n",
      "Searching from 113216 to 126977\n",
      "tRNA-Asn(gtt) at [125855,125926] (111.726)\n",
      "tRNA-Asn(gtt) [125854,125927] (123.398) replacing tRNA-Asn(gtt) [125855,125926] (111.726)\n",
      "tRNA-Arg(acg) at c[126420,126494] (100.692)\n",
      "tRNA-Arg(acg) c[126420,126494] (115.882) replacing tRNA-Arg(acg) c[126420,126494] (100.692)\n",
      "Searching from 126978 to 140739\n",
      "tRNA-Leu(caa) at [140385,140465] (109.768)\n",
      "tRNA-Leu(caa) [140384,140466] (118.842) replacing tRNA-Leu(caa) [140385,140465] (109.768)\n",
      "tRNA-Val(gac) at c[134080,134151] (113.371)\n",
      "tRNA-Thr(cgt) at c[131234,132067] (101.702)\n",
      "tRNA-Glu(ttc) c[131234,132067] (104.79) replacing tRNA-Thr(cgt) c[131234,132067] (101.702)\n",
      "tRNA-Arg(tcg) c[131234,132067] (106.389) replacing tRNA-Glu(ttc) c[131234,132067] (104.79)\n",
      "tRNA-Ile(gat) c[131234,132067] (108.501) replacing tRNA-Arg(tcg) c[131234,132067] (106.389)\n",
      "tRNA-Thr(cgt) c[131234,132067] (109.821) replacing tRNA-Ile(gat) c[131234,132067] (108.501)\n",
      "tRNA-Glu(ttc) c[131234,132067] (112.908) replacing tRNA-Thr(cgt) c[131234,132067] (109.821)\n",
      "tRNA-Ala(tgc) at c[130277,131169] (101.592)\n",
      "tRNA-Ala(tgc) c[130277,131169] (101.595) replacing tRNA-Ala(tgc) c[130277,131169] (101.592)\n",
      "tRNA-Leu(gag) c[130277,131170] (101.613) replacing tRNA-Ala(tgc) c[130277,131169] (101.595)\n",
      "tRNA-Ala(tgc) c[130277,131170] (104.353) replacing tRNA-Leu(gag) c[130277,131170] (101.613)\n",
      "tRNA-Ala(tgc) c[130277,131170] (104.355) replacing tRNA-Ala(tgc) c[130277,131170] (104.353)\n",
      "tRNA-Leu(gag) c[130277,131169] (108.269) replacing tRNA-Ala(tgc) c[130277,131170] (104.355)\n",
      "tRNA-Ala(tgc) c[130277,131169] (111.009) replacing tRNA-Leu(gag) c[130277,131169] (108.269)\n",
      "tRNA-Ala(tgc) c[130277,131169] (111.011) replacing tRNA-Ala(tgc) c[130277,131169] (111.009)\n",
      "Searching from 140740 to 149980\n",
      "Searching from 1 to 3119\n",
      "tRNA-Met(cat) at [147834,147908] (102.061)\n",
      "tRNA-Met(cat) [147834,147908] (108.809) replacing tRNA-Met(cat) [147834,147908] (102.061)\n",
      "tRNA-Lys(ttt) at c[1759,4358] (103.307)\n",
      "tRNA-His(gtg) at c[2,76] (105.78)\n",
      "tRNA-His(gtg) c[3,77] (115.233) replacing tRNA-His(gtg) c[2,76] (105.78)\n",
      "Removing tRNA-Leu(aag) at c[34947,36184] (102.315)\n",
      "Removing tRNA-Arg(tct) at [44139,44864] (103.107)\n",
      "Removing tRNA-Lys(ttt) at c[46146,47213] (100.298)\n",
      "Crepis_callicephala ' '.\n",
      "Search Finished\n",
      "\n",
      "\n",
      "\n",
      "`ARAGORN` run completed successfully!\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import sys\n",
    "\n",
    "fasta = \"plastomes/table2asn/cc/Crepis_callicephala.fsa\"\n",
    "gb = \"plastomes/table2asn/cc/Crepis_callicephala.gbf\"\n",
    "aragorn_out = \"plastomes/table2asn/cc/Crepis_callicephala.aragorn.trnas.txt\"\n",
    "\n",
    "aragorn_cmd = [\n",
    "    \"conda\", \"run\", \"-n\", \"aragorn-env\",\n",
    "    \"aragorn\",\n",
    "    \"-v\", \"-e\",\n",
    "    \"-gcbact\", \n",
    "    \"-c\", \"-i\", \n",
    "    \"-o\", aragorn_out, \n",
    "    gb,\n",
    "]\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(aragorn_cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`ARAGORN` run completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`ARAGORN` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e46c0b6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GenBank', 'to', 'Aragorn', 'comparison']\n",
      "['GenBank', 'Aragorn']\n",
      "start line index: 1663\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>inconsistent</th>\n",
       "      <th>gb_label</th>\n",
       "      <th>gb_coord</th>\n",
       "      <th>arag_label</th>\n",
       "      <th>arag_coord</th>\n",
       "      <th>arag_score</th>\n",
       "      <th>arag_note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-His</td>\n",
       "      <td>c(3,77)</td>\n",
       "      <td>tRNA-His(gtg)</td>\n",
       "      <td>c[3,77]</td>\n",
       "      <td>115.233</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Lys</td>\n",
       "      <td>c(1759,1794)</td>\n",
       "      <td>tRNA-Lys(ttt)</td>\n",
       "      <td>c[1759,4358]</td>\n",
       "      <td>103.307</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(7244,7316)</td>\n",
       "      <td>tRNA-Gln(ttg)</td>\n",
       "      <td>c[7245,7316]</td>\n",
       "      <td>116.786</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(8490,8577)</td>\n",
       "      <td>tRNA-Ser(gct)</td>\n",
       "      <td>c[8490,8577]</td>\n",
       "      <td>123.247</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Cys</td>\n",
       "      <td>(9331,9402)</td>\n",
       "      <td>tRNA-Cys(gca)</td>\n",
       "      <td>[9331,9402]</td>\n",
       "      <td>116.002</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Asp</td>\n",
       "      <td>c(11547,11620)</td>\n",
       "      <td>tRNA-Asp(gtc)</td>\n",
       "      <td>c[11545,11621]</td>\n",
       "      <td>117.885</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(11734,11817)</td>\n",
       "      <td>tRNA-Tyr(gta)</td>\n",
       "      <td>c[11732,11818]</td>\n",
       "      <td>117.555</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(11921,11992)</td>\n",
       "      <td>tRNA-Glu(ttc)</td>\n",
       "      <td>c[11920,11993]</td>\n",
       "      <td>115.411</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(29818,29889)</td>\n",
       "      <td>tRNA-Arg(tct)</td>\n",
       "      <td>c[29818,29889]</td>\n",
       "      <td>117.452</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Gly</td>\n",
       "      <td>c(30095,30142)</td>\n",
       "      <td>tRNA-Ser(cga)</td>\n",
       "      <td>c[30094,30875]</td>\n",
       "      <td>101.783</td>\n",
       "      <td>amino acceptor and sequence length mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td>(31011,31082)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(34947,35029)</td>\n",
       "      <td>tRNA-Ser(tga)</td>\n",
       "      <td>c[34947,35029]</td>\n",
       "      <td>121.044</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Gly</td>\n",
       "      <td>(35855,35925)</td>\n",
       "      <td>tRNA-Gly(gcc)</td>\n",
       "      <td>[35855,35926]</td>\n",
       "      <td>114.476</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>c(36112,36185)</td>\n",
       "      <td>tRNA-Met(cat)</td>\n",
       "      <td>c[36112,36185]</td>\n",
       "      <td>118.102</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>(44778,44864)</td>\n",
       "      <td>tRNA-Ser(gga)</td>\n",
       "      <td>[44778,44864]</td>\n",
       "      <td>121.845</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td>c(46146,46218)</td>\n",
       "      <td>tRNA-Thr(tgt)</td>\n",
       "      <td>c[46145,46219]</td>\n",
       "      <td>122.683</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>(46770,46806)</td>\n",
       "      <td>tRNA-Leu(taa)</td>\n",
       "      <td>[46770,47293]</td>\n",
       "      <td>111.808</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>(47612,47681)</td>\n",
       "      <td>tRNA-Phe(gaa)</td>\n",
       "      <td>[47609,47682]</td>\n",
       "      <td>115.774</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>c(51085,51121)</td>\n",
       "      <td>tRNA-Val(tac)</td>\n",
       "      <td>c[51085,51727]</td>\n",
       "      <td>111.087</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>(51914,51986)</td>\n",
       "      <td>tRNA-Met(cat)</td>\n",
       "      <td>[51914,51986]</td>\n",
       "      <td>111.705</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Trp</td>\n",
       "      <td>c(64142,64215)</td>\n",
       "      <td>tRNA-Trp(cca)</td>\n",
       "      <td>c[64142,64215]</td>\n",
       "      <td>116.997</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td>c(64381,64455)</td>\n",
       "      <td>tRNA-Pro(tgg)</td>\n",
       "      <td>c[64379,64454]</td>\n",
       "      <td>119.886</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>c(83888,83961)</td>\n",
       "      <td>tRNA-Met(cat)</td>\n",
       "      <td>c[83887,83961]</td>\n",
       "      <td>108.809</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>c(91330,91410)</td>\n",
       "      <td>tRNA-Leu(caa)</td>\n",
       "      <td>c[91329,91411]</td>\n",
       "      <td>118.842</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>(97644,97715)</td>\n",
       "      <td>tRNA-Val(gac)</td>\n",
       "      <td>[97644,97715]</td>\n",
       "      <td>113.371</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Ile</td>\n",
       "      <td>(99728,99770)</td>\n",
       "      <td>tRNA-Glu(ttc)</td>\n",
       "      <td>[99728,100561]</td>\n",
       "      <td>112.908</td>\n",
       "      <td>amino acceptor and sequence length mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>(100626,100663)</td>\n",
       "      <td>tRNA-Ala(tgc)</td>\n",
       "      <td>[100626,101518]</td>\n",
       "      <td>111.011</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Arg</td>\n",
       "      <td>(105301,105374)</td>\n",
       "      <td>tRNA-Arg(acg)</td>\n",
       "      <td>[105301,105375]</td>\n",
       "      <td>115.882</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Asn</td>\n",
       "      <td>c(105869,105940)</td>\n",
       "      <td>tRNA-Asn(gtt)</td>\n",
       "      <td>c[105868,105941]</td>\n",
       "      <td>123.398</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>(110967,111046)</td>\n",
       "      <td>tRNA-Leu(tag)</td>\n",
       "      <td>[110967,111046]</td>\n",
       "      <td>120.595</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Asn</td>\n",
       "      <td>(125855,125926)</td>\n",
       "      <td>tRNA-Asn(gtt)</td>\n",
       "      <td>[125854,125927]</td>\n",
       "      <td>123.398</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Arg</td>\n",
       "      <td>c(126421,126494)</td>\n",
       "      <td>tRNA-Arg(acg)</td>\n",
       "      <td>c[126420,126494]</td>\n",
       "      <td>115.882</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>c(130277,130311)</td>\n",
       "      <td>tRNA-Ala(tgc)</td>\n",
       "      <td>c[130277,131169]</td>\n",
       "      <td>111.011</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(131234,131268)</td>\n",
       "      <td>tRNA-Glu(ttc)</td>\n",
       "      <td>c[131234,132067]</td>\n",
       "      <td>112.908</td>\n",
       "      <td>amino acceptor and sequence length mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>c(134080,134152)</td>\n",
       "      <td>tRNA-Val(gac)</td>\n",
       "      <td>c[134080,134151]</td>\n",
       "      <td>113.371</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>(140385,140465)</td>\n",
       "      <td>tRNA-Leu(caa)</td>\n",
       "      <td>[140384,140466]</td>\n",
       "      <td>118.842</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>(147834,147908)</td>\n",
       "      <td>tRNA-Met(cat)</td>\n",
       "      <td>[147834,147908]</td>\n",
       "      <td>108.809</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    inconsistent  gb_label          gb_coord     arag_label        arag_coord  \\\n",
       "0          False  tRNA-His           c(3,77)  tRNA-His(gtg)           c[3,77]   \n",
       "1           True  tRNA-Lys      c(1759,1794)  tRNA-Lys(ttt)      c[1759,4358]   \n",
       "2           True  tRNA-???      c(7244,7316)  tRNA-Gln(ttg)      c[7245,7316]   \n",
       "3           True  tRNA-???      c(8490,8577)  tRNA-Ser(gct)      c[8490,8577]   \n",
       "4          False  tRNA-Cys       (9331,9402)  tRNA-Cys(gca)       [9331,9402]   \n",
       "5          False  tRNA-Asp    c(11547,11620)  tRNA-Asp(gtc)    c[11545,11621]   \n",
       "6           True  tRNA-???    c(11734,11817)  tRNA-Tyr(gta)    c[11732,11818]   \n",
       "7           True  tRNA-???    c(11921,11992)  tRNA-Glu(ttc)    c[11920,11993]   \n",
       "8           True  tRNA-???    c(29818,29889)  tRNA-Arg(tct)    c[29818,29889]   \n",
       "9           True  tRNA-Gly    c(30095,30142)  tRNA-Ser(cga)    c[30094,30875]   \n",
       "10          True  tRNA-Thr     (31011,31082)           None              None   \n",
       "11          True  tRNA-???    c(34947,35029)  tRNA-Ser(tga)    c[34947,35029]   \n",
       "12         False  tRNA-Gly     (35855,35925)  tRNA-Gly(gcc)     [35855,35926]   \n",
       "13         False  tRNA-Met    c(36112,36185)  tRNA-Met(cat)    c[36112,36185]   \n",
       "14          True  tRNA-???     (44778,44864)  tRNA-Ser(gga)     [44778,44864]   \n",
       "15         False  tRNA-Thr    c(46146,46218)  tRNA-Thr(tgt)    c[46145,46219]   \n",
       "16          True  tRNA-Leu     (46770,46806)  tRNA-Leu(taa)     [46770,47293]   \n",
       "17         False  tRNA-Phe     (47612,47681)  tRNA-Phe(gaa)     [47609,47682]   \n",
       "18          True  tRNA-Val    c(51085,51121)  tRNA-Val(tac)    c[51085,51727]   \n",
       "19         False  tRNA-Met     (51914,51986)  tRNA-Met(cat)     [51914,51986]   \n",
       "20         False  tRNA-Trp    c(64142,64215)  tRNA-Trp(cca)    c[64142,64215]   \n",
       "21         False  tRNA-Pro    c(64381,64455)  tRNA-Pro(tgg)    c[64379,64454]   \n",
       "22         False  tRNA-Met    c(83888,83961)  tRNA-Met(cat)    c[83887,83961]   \n",
       "23         False  tRNA-Leu    c(91330,91410)  tRNA-Leu(caa)    c[91329,91411]   \n",
       "24         False  tRNA-Val     (97644,97715)  tRNA-Val(gac)     [97644,97715]   \n",
       "25          True  tRNA-Ile     (99728,99770)  tRNA-Glu(ttc)    [99728,100561]   \n",
       "26          True  tRNA-Ala   (100626,100663)  tRNA-Ala(tgc)   [100626,101518]   \n",
       "27         False  tRNA-Arg   (105301,105374)  tRNA-Arg(acg)   [105301,105375]   \n",
       "28         False  tRNA-Asn  c(105869,105940)  tRNA-Asn(gtt)  c[105868,105941]   \n",
       "29          True  tRNA-???   (110967,111046)  tRNA-Leu(tag)   [110967,111046]   \n",
       "30         False  tRNA-Asn   (125855,125926)  tRNA-Asn(gtt)   [125854,125927]   \n",
       "31         False  tRNA-Arg  c(126421,126494)  tRNA-Arg(acg)  c[126420,126494]   \n",
       "32          True  tRNA-Ala  c(130277,130311)  tRNA-Ala(tgc)  c[130277,131169]   \n",
       "33          True  tRNA-???  c(131234,131268)  tRNA-Glu(ttc)  c[131234,132067]   \n",
       "34         False  tRNA-Val  c(134080,134152)  tRNA-Val(gac)  c[134080,134151]   \n",
       "35         False  tRNA-Leu   (140385,140465)  tRNA-Leu(caa)   [140384,140466]   \n",
       "36         False  tRNA-Met   (147834,147908)  tRNA-Met(cat)   [147834,147908]   \n",
       "\n",
       "   arag_score                                    arag_note  \n",
       "0     115.233                                         None  \n",
       "1     103.307                     sequence length mismatch  \n",
       "2     116.786                      amino acceptor mismatch  \n",
       "3     123.247                      amino acceptor mismatch  \n",
       "4     116.002                                         None  \n",
       "5     117.885                                         None  \n",
       "6     117.555                      amino acceptor mismatch  \n",
       "7     115.411                      amino acceptor mismatch  \n",
       "8     117.452                      amino acceptor mismatch  \n",
       "9     101.783  amino acceptor and sequence length mismatch  \n",
       "10       None                                         None  \n",
       "11    121.044                      amino acceptor mismatch  \n",
       "12    114.476                                         None  \n",
       "13    118.102                                         None  \n",
       "14    121.845                      amino acceptor mismatch  \n",
       "15    122.683                                         None  \n",
       "16    111.808                     sequence length mismatch  \n",
       "17    115.774                                         None  \n",
       "18    111.087                     sequence length mismatch  \n",
       "19    111.705                                         None  \n",
       "20    116.997                                         None  \n",
       "21    119.886                                         None  \n",
       "22    108.809                                         None  \n",
       "23    118.842                                         None  \n",
       "24    113.371                                         None  \n",
       "25    112.908  amino acceptor and sequence length mismatch  \n",
       "26    111.011                     sequence length mismatch  \n",
       "27    115.882                                         None  \n",
       "28    123.398                                         None  \n",
       "29    120.595                      amino acceptor mismatch  \n",
       "30    123.398                                         None  \n",
       "31    115.882                                         None  \n",
       "32    111.011                     sequence length mismatch  \n",
       "33    112.908  amino acceptor and sequence length mismatch  \n",
       "34    113.371                                         None  \n",
       "35    118.842                                         None  \n",
       "36    108.809                                         None  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# parse Aragorn output\n",
    "import pandas as pd\n",
    "\n",
    "aragorn_out = \"plastomes/table2asn/cc/Crepis_callicephala.aragorn.trnas.txt\"\n",
    "list_parsed = []\n",
    "\n",
    "with open(aragorn_out, \"r\") as handle:\n",
    "    start_line = None\n",
    "    for i, line in enumerate(handle):\n",
    "        line_parts = []\n",
    "        if \"Aragorn\" in line:\n",
    "            if \"GenBank\" in line and \"Aragorn\" in line:\n",
    "                line_parts = line.strip(\" \").split()\n",
    "                print(line_parts)\n",
    "                if len(line_parts) == 2:\n",
    "                    start_line = i\n",
    "                    print(f\"start line index: {start_line}\")\n",
    "        if start_line:\n",
    "            if i == start_line:\n",
    "                pass\n",
    "            else:\n",
    "                gb_label = None\n",
    "                gb_coord = None\n",
    "                arag_note = None\n",
    "                arag_label = None\n",
    "                arag_coord = None\n",
    "                arag_score = None\n",
    "                line = line.split(\"\\n\")[0]\n",
    "                #print(i, line)\n",
    "                inconsistent = False\n",
    "                split_initial = line.split(\" \", maxsplit=1)\n",
    "                consistency_flag = split_initial[0]\n",
    "                if len(split_initial) < 2:\n",
    "                    break\n",
    "                else:\n",
    "                    remain_part = line.split(\" \", maxsplit=1)[1].strip()\n",
    "                    line_parts.append(consistency_flag)\n",
    "                    line_parts.extend(remain_part.split(maxsplit=5))\n",
    "                    #print(line_parts)\n",
    "                    if \"*\" in line_parts[0]:\n",
    "                        inconsistent = True\n",
    "                    gb_label = line_parts[1]\n",
    "                    gb_coord = line_parts[2]\n",
    "                    if 'Not' in line_parts[3]:\n",
    "                        pass\n",
    "                    else:\n",
    "                        arag_label = line_parts[3]\n",
    "                        arag_coord = line_parts[4]\n",
    "                        arag_score = line_parts[5]\n",
    "                        #print(f\"'line_parts' length: {len(line_parts)}\")\n",
    "                        if len(line_parts) > 6:\n",
    "                            arag_note = line_parts[-1]\n",
    "                    #print(f\"{inconsistent}\\t{gb_label}\\t{gb_coord}\\t{arag_label}\\t{arag_coord}\\t{arag_score}\\t{arag_note}\")\n",
    "                    line_dict = {\n",
    "                        'inconsistent': inconsistent,\n",
    "                        'gb_label': gb_label,\n",
    "                        'gb_coord': gb_coord,\n",
    "                        'arag_label': arag_label,\n",
    "                        'arag_coord': arag_coord,\n",
    "                        'arag_score': arag_score,\n",
    "                        'arag_note': arag_note\n",
    "                    }\n",
    "\n",
    "                    list_parsed.append(line_dict)\n",
    "                    #print(line_dict)\n",
    "\n",
    "df_aragorn = pd.DataFrame.from_dict(list_parsed)\n",
    "display(df_aragorn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e278ddd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>inconsistent</th>\n",
       "      <th>gb_label</th>\n",
       "      <th>gb_coord</th>\n",
       "      <th>arag_label</th>\n",
       "      <th>arag_coord</th>\n",
       "      <th>arag_score</th>\n",
       "      <th>arag_note</th>\n",
       "      <th>gb_strand</th>\n",
       "      <th>gb_start</th>\n",
       "      <th>gb_end</th>\n",
       "      <th>arag_strand</th>\n",
       "      <th>arag_start</th>\n",
       "      <th>arag_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-His</td>\n",
       "      <td>c(3,77)</td>\n",
       "      <td>tRNA-His(gtg)</td>\n",
       "      <td>c[3,77]</td>\n",
       "      <td>115.233</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>77</td>\n",
       "      <td>-</td>\n",
       "      <td>3.0</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Lys</td>\n",
       "      <td>c(1759,1794)</td>\n",
       "      <td>tRNA-Lys(ttt)</td>\n",
       "      <td>c[1759,4358]</td>\n",
       "      <td>103.307</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>1759</td>\n",
       "      <td>1794</td>\n",
       "      <td>-</td>\n",
       "      <td>1759.0</td>\n",
       "      <td>4358.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(7244,7316)</td>\n",
       "      <td>tRNA-Gln(ttg)</td>\n",
       "      <td>c[7245,7316]</td>\n",
       "      <td>116.786</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>7244</td>\n",
       "      <td>7316</td>\n",
       "      <td>-</td>\n",
       "      <td>7245.0</td>\n",
       "      <td>7316.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(8490,8577)</td>\n",
       "      <td>tRNA-Ser(gct)</td>\n",
       "      <td>c[8490,8577]</td>\n",
       "      <td>123.247</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>8490</td>\n",
       "      <td>8577</td>\n",
       "      <td>-</td>\n",
       "      <td>8490.0</td>\n",
       "      <td>8577.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Cys</td>\n",
       "      <td>(9331,9402)</td>\n",
       "      <td>tRNA-Cys(gca)</td>\n",
       "      <td>[9331,9402]</td>\n",
       "      <td>116.002</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>9331</td>\n",
       "      <td>9402</td>\n",
       "      <td>+</td>\n",
       "      <td>9331.0</td>\n",
       "      <td>9402.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Asp</td>\n",
       "      <td>c(11547,11620)</td>\n",
       "      <td>tRNA-Asp(gtc)</td>\n",
       "      <td>c[11545,11621]</td>\n",
       "      <td>117.885</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>11547</td>\n",
       "      <td>11620</td>\n",
       "      <td>-</td>\n",
       "      <td>11545.0</td>\n",
       "      <td>11621.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(11734,11817)</td>\n",
       "      <td>tRNA-Tyr(gta)</td>\n",
       "      <td>c[11732,11818]</td>\n",
       "      <td>117.555</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>11734</td>\n",
       "      <td>11817</td>\n",
       "      <td>-</td>\n",
       "      <td>11732.0</td>\n",
       "      <td>11818.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(11921,11992)</td>\n",
       "      <td>tRNA-Glu(ttc)</td>\n",
       "      <td>c[11920,11993]</td>\n",
       "      <td>115.411</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>11921</td>\n",
       "      <td>11992</td>\n",
       "      <td>-</td>\n",
       "      <td>11920.0</td>\n",
       "      <td>11993.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(29818,29889)</td>\n",
       "      <td>tRNA-Arg(tct)</td>\n",
       "      <td>c[29818,29889]</td>\n",
       "      <td>117.452</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>29818</td>\n",
       "      <td>29889</td>\n",
       "      <td>-</td>\n",
       "      <td>29818.0</td>\n",
       "      <td>29889.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Gly</td>\n",
       "      <td>c(30095,30142)</td>\n",
       "      <td>tRNA-Ser(cga)</td>\n",
       "      <td>c[30094,30875]</td>\n",
       "      <td>101.783</td>\n",
       "      <td>amino acceptor and sequence length mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>30095</td>\n",
       "      <td>30142</td>\n",
       "      <td>-</td>\n",
       "      <td>30094.0</td>\n",
       "      <td>30875.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td>(31011,31082)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>31011</td>\n",
       "      <td>31082</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(34947,35029)</td>\n",
       "      <td>tRNA-Ser(tga)</td>\n",
       "      <td>c[34947,35029]</td>\n",
       "      <td>121.044</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>34947</td>\n",
       "      <td>35029</td>\n",
       "      <td>-</td>\n",
       "      <td>34947.0</td>\n",
       "      <td>35029.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Gly</td>\n",
       "      <td>(35855,35925)</td>\n",
       "      <td>tRNA-Gly(gcc)</td>\n",
       "      <td>[35855,35926]</td>\n",
       "      <td>114.476</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>35855</td>\n",
       "      <td>35925</td>\n",
       "      <td>+</td>\n",
       "      <td>35855.0</td>\n",
       "      <td>35926.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>c(36112,36185)</td>\n",
       "      <td>tRNA-Met(cat)</td>\n",
       "      <td>c[36112,36185]</td>\n",
       "      <td>118.102</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>36112</td>\n",
       "      <td>36185</td>\n",
       "      <td>-</td>\n",
       "      <td>36112.0</td>\n",
       "      <td>36185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>(44778,44864)</td>\n",
       "      <td>tRNA-Ser(gga)</td>\n",
       "      <td>[44778,44864]</td>\n",
       "      <td>121.845</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "      <td>+</td>\n",
       "      <td>44778</td>\n",
       "      <td>44864</td>\n",
       "      <td>+</td>\n",
       "      <td>44778.0</td>\n",
       "      <td>44864.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td>c(46146,46218)</td>\n",
       "      <td>tRNA-Thr(tgt)</td>\n",
       "      <td>c[46145,46219]</td>\n",
       "      <td>122.683</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>46146</td>\n",
       "      <td>46218</td>\n",
       "      <td>-</td>\n",
       "      <td>46145.0</td>\n",
       "      <td>46219.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>(46770,46806)</td>\n",
       "      <td>tRNA-Leu(taa)</td>\n",
       "      <td>[46770,47293]</td>\n",
       "      <td>111.808</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "      <td>+</td>\n",
       "      <td>46770</td>\n",
       "      <td>46806</td>\n",
       "      <td>+</td>\n",
       "      <td>46770.0</td>\n",
       "      <td>47293.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Phe</td>\n",
       "      <td>(47612,47681)</td>\n",
       "      <td>tRNA-Phe(gaa)</td>\n",
       "      <td>[47609,47682]</td>\n",
       "      <td>115.774</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>47612</td>\n",
       "      <td>47681</td>\n",
       "      <td>+</td>\n",
       "      <td>47609.0</td>\n",
       "      <td>47682.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>c(51085,51121)</td>\n",
       "      <td>tRNA-Val(tac)</td>\n",
       "      <td>c[51085,51727]</td>\n",
       "      <td>111.087</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>51085</td>\n",
       "      <td>51121</td>\n",
       "      <td>-</td>\n",
       "      <td>51085.0</td>\n",
       "      <td>51727.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>(51914,51986)</td>\n",
       "      <td>tRNA-Met(cat)</td>\n",
       "      <td>[51914,51986]</td>\n",
       "      <td>111.705</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>51914</td>\n",
       "      <td>51986</td>\n",
       "      <td>+</td>\n",
       "      <td>51914.0</td>\n",
       "      <td>51986.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Trp</td>\n",
       "      <td>c(64142,64215)</td>\n",
       "      <td>tRNA-Trp(cca)</td>\n",
       "      <td>c[64142,64215]</td>\n",
       "      <td>116.997</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>64142</td>\n",
       "      <td>64215</td>\n",
       "      <td>-</td>\n",
       "      <td>64142.0</td>\n",
       "      <td>64215.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td>c(64381,64455)</td>\n",
       "      <td>tRNA-Pro(tgg)</td>\n",
       "      <td>c[64379,64454]</td>\n",
       "      <td>119.886</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>64381</td>\n",
       "      <td>64455</td>\n",
       "      <td>-</td>\n",
       "      <td>64379.0</td>\n",
       "      <td>64454.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>c(83888,83961)</td>\n",
       "      <td>tRNA-Met(cat)</td>\n",
       "      <td>c[83887,83961]</td>\n",
       "      <td>108.809</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>83888</td>\n",
       "      <td>83961</td>\n",
       "      <td>-</td>\n",
       "      <td>83887.0</td>\n",
       "      <td>83961.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>c(91330,91410)</td>\n",
       "      <td>tRNA-Leu(caa)</td>\n",
       "      <td>c[91329,91411]</td>\n",
       "      <td>118.842</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>91330</td>\n",
       "      <td>91410</td>\n",
       "      <td>-</td>\n",
       "      <td>91329.0</td>\n",
       "      <td>91411.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>(97644,97715)</td>\n",
       "      <td>tRNA-Val(gac)</td>\n",
       "      <td>[97644,97715]</td>\n",
       "      <td>113.371</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>97644</td>\n",
       "      <td>97715</td>\n",
       "      <td>+</td>\n",
       "      <td>97644.0</td>\n",
       "      <td>97715.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Ile</td>\n",
       "      <td>(99728,99770)</td>\n",
       "      <td>tRNA-Glu(ttc)</td>\n",
       "      <td>[99728,100561]</td>\n",
       "      <td>112.908</td>\n",
       "      <td>amino acceptor and sequence length mismatch</td>\n",
       "      <td>+</td>\n",
       "      <td>99728</td>\n",
       "      <td>99770</td>\n",
       "      <td>+</td>\n",
       "      <td>99728.0</td>\n",
       "      <td>100561.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>(100626,100663)</td>\n",
       "      <td>tRNA-Ala(tgc)</td>\n",
       "      <td>[100626,101518]</td>\n",
       "      <td>111.011</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "      <td>+</td>\n",
       "      <td>100626</td>\n",
       "      <td>100663</td>\n",
       "      <td>+</td>\n",
       "      <td>100626.0</td>\n",
       "      <td>101518.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Arg</td>\n",
       "      <td>(105301,105374)</td>\n",
       "      <td>tRNA-Arg(acg)</td>\n",
       "      <td>[105301,105375]</td>\n",
       "      <td>115.882</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>105301</td>\n",
       "      <td>105374</td>\n",
       "      <td>+</td>\n",
       "      <td>105301.0</td>\n",
       "      <td>105375.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Asn</td>\n",
       "      <td>c(105869,105940)</td>\n",
       "      <td>tRNA-Asn(gtt)</td>\n",
       "      <td>c[105868,105941]</td>\n",
       "      <td>123.398</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>105869</td>\n",
       "      <td>105940</td>\n",
       "      <td>-</td>\n",
       "      <td>105868.0</td>\n",
       "      <td>105941.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>(110967,111046)</td>\n",
       "      <td>tRNA-Leu(tag)</td>\n",
       "      <td>[110967,111046]</td>\n",
       "      <td>120.595</td>\n",
       "      <td>amino acceptor mismatch</td>\n",
       "      <td>+</td>\n",
       "      <td>110967</td>\n",
       "      <td>111046</td>\n",
       "      <td>+</td>\n",
       "      <td>110967.0</td>\n",
       "      <td>111046.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Asn</td>\n",
       "      <td>(125855,125926)</td>\n",
       "      <td>tRNA-Asn(gtt)</td>\n",
       "      <td>[125854,125927]</td>\n",
       "      <td>123.398</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>125855</td>\n",
       "      <td>125926</td>\n",
       "      <td>+</td>\n",
       "      <td>125854.0</td>\n",
       "      <td>125927.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Arg</td>\n",
       "      <td>c(126421,126494)</td>\n",
       "      <td>tRNA-Arg(acg)</td>\n",
       "      <td>c[126420,126494]</td>\n",
       "      <td>115.882</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>126421</td>\n",
       "      <td>126494</td>\n",
       "      <td>-</td>\n",
       "      <td>126420.0</td>\n",
       "      <td>126494.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>c(130277,130311)</td>\n",
       "      <td>tRNA-Ala(tgc)</td>\n",
       "      <td>c[130277,131169]</td>\n",
       "      <td>111.011</td>\n",
       "      <td>sequence length mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>130277</td>\n",
       "      <td>130311</td>\n",
       "      <td>-</td>\n",
       "      <td>130277.0</td>\n",
       "      <td>131169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>True</td>\n",
       "      <td>tRNA-???</td>\n",
       "      <td>c(131234,131268)</td>\n",
       "      <td>tRNA-Glu(ttc)</td>\n",
       "      <td>c[131234,132067]</td>\n",
       "      <td>112.908</td>\n",
       "      <td>amino acceptor and sequence length mismatch</td>\n",
       "      <td>-</td>\n",
       "      <td>131234</td>\n",
       "      <td>131268</td>\n",
       "      <td>-</td>\n",
       "      <td>131234.0</td>\n",
       "      <td>132067.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Val</td>\n",
       "      <td>c(134080,134152)</td>\n",
       "      <td>tRNA-Val(gac)</td>\n",
       "      <td>c[134080,134151]</td>\n",
       "      <td>113.371</td>\n",
       "      <td>None</td>\n",
       "      <td>-</td>\n",
       "      <td>134080</td>\n",
       "      <td>134152</td>\n",
       "      <td>-</td>\n",
       "      <td>134080.0</td>\n",
       "      <td>134151.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Leu</td>\n",
       "      <td>(140385,140465)</td>\n",
       "      <td>tRNA-Leu(caa)</td>\n",
       "      <td>[140384,140466]</td>\n",
       "      <td>118.842</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>140385</td>\n",
       "      <td>140465</td>\n",
       "      <td>+</td>\n",
       "      <td>140384.0</td>\n",
       "      <td>140466.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>False</td>\n",
       "      <td>tRNA-Met</td>\n",
       "      <td>(147834,147908)</td>\n",
       "      <td>tRNA-Met(cat)</td>\n",
       "      <td>[147834,147908]</td>\n",
       "      <td>108.809</td>\n",
       "      <td>None</td>\n",
       "      <td>+</td>\n",
       "      <td>147834</td>\n",
       "      <td>147908</td>\n",
       "      <td>+</td>\n",
       "      <td>147834.0</td>\n",
       "      <td>147908.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    inconsistent  gb_label          gb_coord     arag_label        arag_coord  \\\n",
       "0          False  tRNA-His           c(3,77)  tRNA-His(gtg)           c[3,77]   \n",
       "1           True  tRNA-Lys      c(1759,1794)  tRNA-Lys(ttt)      c[1759,4358]   \n",
       "2           True  tRNA-???      c(7244,7316)  tRNA-Gln(ttg)      c[7245,7316]   \n",
       "3           True  tRNA-???      c(8490,8577)  tRNA-Ser(gct)      c[8490,8577]   \n",
       "4          False  tRNA-Cys       (9331,9402)  tRNA-Cys(gca)       [9331,9402]   \n",
       "5          False  tRNA-Asp    c(11547,11620)  tRNA-Asp(gtc)    c[11545,11621]   \n",
       "6           True  tRNA-???    c(11734,11817)  tRNA-Tyr(gta)    c[11732,11818]   \n",
       "7           True  tRNA-???    c(11921,11992)  tRNA-Glu(ttc)    c[11920,11993]   \n",
       "8           True  tRNA-???    c(29818,29889)  tRNA-Arg(tct)    c[29818,29889]   \n",
       "9           True  tRNA-Gly    c(30095,30142)  tRNA-Ser(cga)    c[30094,30875]   \n",
       "10          True  tRNA-Thr     (31011,31082)           None              None   \n",
       "11          True  tRNA-???    c(34947,35029)  tRNA-Ser(tga)    c[34947,35029]   \n",
       "12         False  tRNA-Gly     (35855,35925)  tRNA-Gly(gcc)     [35855,35926]   \n",
       "13         False  tRNA-Met    c(36112,36185)  tRNA-Met(cat)    c[36112,36185]   \n",
       "14          True  tRNA-???     (44778,44864)  tRNA-Ser(gga)     [44778,44864]   \n",
       "15         False  tRNA-Thr    c(46146,46218)  tRNA-Thr(tgt)    c[46145,46219]   \n",
       "16          True  tRNA-Leu     (46770,46806)  tRNA-Leu(taa)     [46770,47293]   \n",
       "17         False  tRNA-Phe     (47612,47681)  tRNA-Phe(gaa)     [47609,47682]   \n",
       "18          True  tRNA-Val    c(51085,51121)  tRNA-Val(tac)    c[51085,51727]   \n",
       "19         False  tRNA-Met     (51914,51986)  tRNA-Met(cat)     [51914,51986]   \n",
       "20         False  tRNA-Trp    c(64142,64215)  tRNA-Trp(cca)    c[64142,64215]   \n",
       "21         False  tRNA-Pro    c(64381,64455)  tRNA-Pro(tgg)    c[64379,64454]   \n",
       "22         False  tRNA-Met    c(83888,83961)  tRNA-Met(cat)    c[83887,83961]   \n",
       "23         False  tRNA-Leu    c(91330,91410)  tRNA-Leu(caa)    c[91329,91411]   \n",
       "24         False  tRNA-Val     (97644,97715)  tRNA-Val(gac)     [97644,97715]   \n",
       "25          True  tRNA-Ile     (99728,99770)  tRNA-Glu(ttc)    [99728,100561]   \n",
       "26          True  tRNA-Ala   (100626,100663)  tRNA-Ala(tgc)   [100626,101518]   \n",
       "27         False  tRNA-Arg   (105301,105374)  tRNA-Arg(acg)   [105301,105375]   \n",
       "28         False  tRNA-Asn  c(105869,105940)  tRNA-Asn(gtt)  c[105868,105941]   \n",
       "29          True  tRNA-???   (110967,111046)  tRNA-Leu(tag)   [110967,111046]   \n",
       "30         False  tRNA-Asn   (125855,125926)  tRNA-Asn(gtt)   [125854,125927]   \n",
       "31         False  tRNA-Arg  c(126421,126494)  tRNA-Arg(acg)  c[126420,126494]   \n",
       "32          True  tRNA-Ala  c(130277,130311)  tRNA-Ala(tgc)  c[130277,131169]   \n",
       "33          True  tRNA-???  c(131234,131268)  tRNA-Glu(ttc)  c[131234,132067]   \n",
       "34         False  tRNA-Val  c(134080,134152)  tRNA-Val(gac)  c[134080,134151]   \n",
       "35         False  tRNA-Leu   (140385,140465)  tRNA-Leu(caa)   [140384,140466]   \n",
       "36         False  tRNA-Met   (147834,147908)  tRNA-Met(cat)   [147834,147908]   \n",
       "\n",
       "   arag_score                                    arag_note gb_strand  \\\n",
       "0     115.233                                         None         -   \n",
       "1     103.307                     sequence length mismatch         -   \n",
       "2     116.786                      amino acceptor mismatch         -   \n",
       "3     123.247                      amino acceptor mismatch         -   \n",
       "4     116.002                                         None         +   \n",
       "5     117.885                                         None         -   \n",
       "6     117.555                      amino acceptor mismatch         -   \n",
       "7     115.411                      amino acceptor mismatch         -   \n",
       "8     117.452                      amino acceptor mismatch         -   \n",
       "9     101.783  amino acceptor and sequence length mismatch         -   \n",
       "10       None                                         None         +   \n",
       "11    121.044                      amino acceptor mismatch         -   \n",
       "12    114.476                                         None         +   \n",
       "13    118.102                                         None         -   \n",
       "14    121.845                      amino acceptor mismatch         +   \n",
       "15    122.683                                         None         -   \n",
       "16    111.808                     sequence length mismatch         +   \n",
       "17    115.774                                         None         +   \n",
       "18    111.087                     sequence length mismatch         -   \n",
       "19    111.705                                         None         +   \n",
       "20    116.997                                         None         -   \n",
       "21    119.886                                         None         -   \n",
       "22    108.809                                         None         -   \n",
       "23    118.842                                         None         -   \n",
       "24    113.371                                         None         +   \n",
       "25    112.908  amino acceptor and sequence length mismatch         +   \n",
       "26    111.011                     sequence length mismatch         +   \n",
       "27    115.882                                         None         +   \n",
       "28    123.398                                         None         -   \n",
       "29    120.595                      amino acceptor mismatch         +   \n",
       "30    123.398                                         None         +   \n",
       "31    115.882                                         None         -   \n",
       "32    111.011                     sequence length mismatch         -   \n",
       "33    112.908  amino acceptor and sequence length mismatch         -   \n",
       "34    113.371                                         None         -   \n",
       "35    118.842                                         None         +   \n",
       "36    108.809                                         None         +   \n",
       "\n",
       "    gb_start  gb_end arag_strand  arag_start  arag_end  \n",
       "0          3      77           -         3.0      77.0  \n",
       "1       1759    1794           -      1759.0    4358.0  \n",
       "2       7244    7316           -      7245.0    7316.0  \n",
       "3       8490    8577           -      8490.0    8577.0  \n",
       "4       9331    9402           +      9331.0    9402.0  \n",
       "5      11547   11620           -     11545.0   11621.0  \n",
       "6      11734   11817           -     11732.0   11818.0  \n",
       "7      11921   11992           -     11920.0   11993.0  \n",
       "8      29818   29889           -     29818.0   29889.0  \n",
       "9      30095   30142           -     30094.0   30875.0  \n",
       "10     31011   31082           +         NaN       NaN  \n",
       "11     34947   35029           -     34947.0   35029.0  \n",
       "12     35855   35925           +     35855.0   35926.0  \n",
       "13     36112   36185           -     36112.0   36185.0  \n",
       "14     44778   44864           +     44778.0   44864.0  \n",
       "15     46146   46218           -     46145.0   46219.0  \n",
       "16     46770   46806           +     46770.0   47293.0  \n",
       "17     47612   47681           +     47609.0   47682.0  \n",
       "18     51085   51121           -     51085.0   51727.0  \n",
       "19     51914   51986           +     51914.0   51986.0  \n",
       "20     64142   64215           -     64142.0   64215.0  \n",
       "21     64381   64455           -     64379.0   64454.0  \n",
       "22     83888   83961           -     83887.0   83961.0  \n",
       "23     91330   91410           -     91329.0   91411.0  \n",
       "24     97644   97715           +     97644.0   97715.0  \n",
       "25     99728   99770           +     99728.0  100561.0  \n",
       "26    100626  100663           +    100626.0  101518.0  \n",
       "27    105301  105374           +    105301.0  105375.0  \n",
       "28    105869  105940           -    105868.0  105941.0  \n",
       "29    110967  111046           +    110967.0  111046.0  \n",
       "30    125855  125926           +    125854.0  125927.0  \n",
       "31    126421  126494           -    126420.0  126494.0  \n",
       "32    130277  130311           -    130277.0  131169.0  \n",
       "33    131234  131268           -    131234.0  132067.0  \n",
       "34    134080  134152           -    134080.0  134151.0  \n",
       "35    140385  140465           +    140384.0  140466.0  \n",
       "36    147834  147908           +    147834.0  147908.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# decomposing coords represented like \"c(3,77)\" or \"(9331,9402)\"\n",
    "# the coords data are stored at the 'gb_coord' column\n",
    "# in the string, \"c\" symbol is mentioned for complement-stranded coords\n",
    "import numpy as np\n",
    "\n",
    "# decomposing 'gb_coord'\n",
    "df_aragorn['gb_strand'] = np.where(\n",
    "    df_aragorn['gb_coord'].str.startswith(\"c\"), \n",
    "    \"-\", \"+\"\n",
    "    )\n",
    "\n",
    "coord = df_aragorn['gb_coord'].str.strip(\"[(c)] \").str.split(\",\", expand=True)\n",
    "df_aragorn[\"gb_start\"] = coord[0].astype(int)\n",
    "df_aragorn[\"gb_end\"] = coord[1].astype(int)\n",
    "\n",
    "# decomposing 'arag_coord'\n",
    "# one row contains None for 'arag_coord' field\n",
    "df_aragorn['arag_strand'] = np.where(\n",
    "    df_aragorn[\"arag_coord\"].notna() & \n",
    "    df_aragorn['arag_coord'].str.startswith(\"c\"), \n",
    "    \"-\", \"+\"\n",
    "    )\n",
    "\n",
    "coord = df_aragorn['arag_coord'].where(df_aragorn[\"arag_coord\"].notna()).str.lstrip(\"c\").str.strip(\"[] \").str.split(\",\", expand=True)\n",
    "df_aragorn[\"arag_start\"] = pd.to_numeric(coord[0], errors=\"coerce\")\n",
    "df_aragorn[\"arag_end\"] = pd.to_numeric(coord[1], errors=\"coerce\")\n",
    "\n",
    "display(df_aragorn)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "472661fe",
   "metadata": {},
   "source": [
    "### Assessing tRNAs using tRNAscan-SE v2.0.12\n",
    "The command to run:\n",
    "```shell\n",
    "tRNAscan-SE -B -o $output -m $trnascan_stat --detail --log $logfile -v --thread 4 $fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a7bdbaf",
   "metadata": {},
   "source": [
    "### Manually adding tRNA labels\n",
    "According to Aragorn output both product names were added and coordinates adjusted.\n",
    "Complicated thing is where amino acids for the same position mismathes. For example, in current annotation tRNA-Gly c(30095,30142), but in Aragorn annotation tRNA-Ser(cga) c[30094,30875]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "00773082",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: Status: Phase I: Searching for tRNAs with HMM-enabled Infernal\n",
      "Status: Phase II: Infernal verification of candidate tRNAs detected with first-pass scan\n",
      "\n",
      "\n",
      "STDERR: \n",
      "tRNAscan-SE v.2.0.12 (Nov 2022) - scan sequences for transfer RNAs\n",
      "Copyright (C) 2022 Patricia Chan and Todd Lowe\n",
      "                   University of California Santa Cruz\n",
      "Freely distributed under the GNU General Public License (GPLv3)\n",
      "\n",
      "------------------------------------------------------------\n",
      "Sequence file(s) to search:        plastomes/table2asn/cc/Crepis_callicephala.fsa\n",
      "Search Mode:                       Bacterial\n",
      "Results written to:                plastomes/table2asn/cc/Crepis_callicephala.trnascan.trnas.txt\n",
      "Output format:                     Tabular\n",
      "Searching with:                    Infernal First Pass->Infernal\n",
      "Isotype-specific model scan:       Yes\n",
      "Covariance model:                  /home/asan/miniconda3/envs/trnascan-env/lib/tRNAscan-SE/models/TRNAinf-bact.cm\n",
      "                                   /home/asan/miniconda3/envs/trnascan-env/lib/tRNAscan-SE/models/TRNAinf-bact-SeC.cm\n",
      "Infernal first pass cutoff score:  10\n",
      "\n",
      "Temporary directory:               /tmp\n",
      "Search log saved in:               plastomes/table2asn/cc/Crepis_callicephala.trnascan.log.txt\n",
      "Search statistics saved in:        plastomes/table2asn/cc/Crepis_callicephala.trnascan.stat.txt\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "`tRNAscan` run completed successfully!\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import sys\n",
    "\n",
    "fasta = \"plastomes/table2asn/cc/Crepis_callicephala.fsa\"\n",
    "gb = \"plastomes/table2asn/cc/Crepis_callicephala.gbf\"\n",
    "trnascan_out = \"plastomes/table2asn/cc/Crepis_callicephala.trnascan.trnas.txt\"\n",
    "logfile = \"plastomes/table2asn/cc/Crepis_callicephala.trnascan.log.txt\"\n",
    "trnascan_stat = \"plastomes/table2asn/cc/Crepis_callicephala.trnascan.stat.txt\"\n",
    "trnascan_verbose_out = \"plastomes/table2asn/cc/Crepis_callicephala.trnascan.verbose.txt\"\n",
    "\n",
    "\n",
    "trnascan_cmd = [\n",
    "    \"conda\", \"run\", \"-n\", \"trnascan-env\", \n",
    "    \"tRNAscan-SE\", \n",
    "    \"-B\", \n",
    "    \"-o\", trnascan_out, \n",
    "    \"-m\", trnascan_stat, \n",
    "    \"--detail\", \n",
    "    \"--log\", logfile, \n",
    "    \"-v\", trnascan_verbose_out,\n",
    "    fasta\n",
    "]\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(trnascan_cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`tRNAscan` run completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`tRNAscan` failed with exit code {e.returncode}\")\n",
    "    print(\"STDOUT:\", e.stdout)\n",
    "    print(\"STDERR:\", e.stderr)\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eb70778",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4548/4270166751.py:29: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  df = pd.read_csv(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tRNA_No</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>3_letter</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9331</td>\n",
       "      <td>9401</td>\n",
       "      <td>Cys</td>\n",
       "      <td>GCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>59.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>31016</td>\n",
       "      <td>31080</td>\n",
       "      <td>Thr</td>\n",
       "      <td>GGT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>35855</td>\n",
       "      <td>35925</td>\n",
       "      <td>Gly</td>\n",
       "      <td>GCC</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>44778</td>\n",
       "      <td>44864</td>\n",
       "      <td>Ser</td>\n",
       "      <td>GGA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>47609</td>\n",
       "      <td>47681</td>\n",
       "      <td>Phe</td>\n",
       "      <td>GAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tRNA_No  start    end 3_letter anticodon  intron_start  intron_end  score\n",
       "0        1   9331   9401      Cys       GCA             0           0   59.5\n",
       "1        2  31016  31080      Thr       GGT             0           0   20.0\n",
       "2        3  35855  35925      Gly       GCC             0           0   57.3\n",
       "3        4  44778  44864      Ser       GGA             0           0   67.7\n",
       "4        5  47609  47681      Phe       GAA             0           0   61.7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import io\n",
    "\n",
    "def parse_trnascan_output(file_path: str|os.PathLike):\n",
    "    \"\"\"\n",
    "    Parse tRNAscan-SE output file and return a pandas DataFrame.\n",
    "    \"\"\"\n",
    "    column_names = [\n",
    "        \"sequence_name\", \"tRNA_No\", \"start\", \"end\", \"3_letter\", \n",
    "        \"anticodon\", \"intron_start\", \"intron_end\", \"score\", \n",
    "        \"isotype_CM\", \"isotype_score\", \"note\",\n",
    "    ]\n",
    "    \n",
    "    with open(file_path, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "    \n",
    "    # Find the start of the data (should be \n",
    "    # after the dashed separator line)\n",
    "    data_start = 0\n",
    "    for i, line in enumerate(lines):\n",
    "        if line.startswith(\"--------\"):\n",
    "            data_start = i + 1\n",
    "            break\n",
    "    \n",
    "    data_lines = lines[data_start:]\n",
    "    data_str = \"\".join(data_lines)\n",
    "    \n",
    "    df = pd.read_csv(\n",
    "        io.StringIO(data_str),\n",
    "        delim_whitespace=True,\n",
    "        header=None,\n",
    "        names=column_names,\n",
    "        index_col=False\n",
    "    )\n",
    "    \n",
    "    # Fix column types\n",
    "    numeric_cols = [\n",
    "        \"tRNA_No\", \"start\", \"end\", \n",
    "        \"intron_start\", \"intron_end\", \n",
    "        \"score\", \"isotype_score\",\n",
    "        ]\n",
    "    for col in numeric_cols:\n",
    "        df[col] = pd.to_numeric(df[col], errors='coerce')\n",
    "    \n",
    "    return df\n",
    "\n",
    "df_trnascan = parse_trnascan_output(trnascan_out)\n",
    "filter_cols = [\n",
    "    \"tRNA_No\", \"start\", \"end\", \"3_letter\", \n",
    "    \"anticodon\", \"intron_start\", \"intron_end\", \"score\",\n",
    "]\n",
    "df_trnascan = df_trnascan[filter_cols]\n",
    "display(df_trnascan.head())\n",
    "\n",
    "# Optional: Save to CSV\n",
    "# df.to_csv(\"trnascan_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dade3c24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>3_letter_code</th>\n",
       "      <th>1_letter_code</th>\n",
       "      <th>3_letter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alanine</td>\n",
       "      <td>ALA</td>\n",
       "      <td>A</td>\n",
       "      <td>Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cysteine</td>\n",
       "      <td>CYS</td>\n",
       "      <td>C</td>\n",
       "      <td>Cys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aspartic Acid</td>\n",
       "      <td>ASP</td>\n",
       "      <td>D</td>\n",
       "      <td>Asp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glutamic Acid</td>\n",
       "      <td>GLU</td>\n",
       "      <td>E</td>\n",
       "      <td>Glu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Phenylalanine</td>\n",
       "      <td>PHE</td>\n",
       "      <td>F</td>\n",
       "      <td>Phe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Glycine</td>\n",
       "      <td>GLY</td>\n",
       "      <td>G</td>\n",
       "      <td>Gly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Histidine</td>\n",
       "      <td>HIS</td>\n",
       "      <td>H</td>\n",
       "      <td>His</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Isoleucine</td>\n",
       "      <td>ILE</td>\n",
       "      <td>I</td>\n",
       "      <td>Ile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Lysine</td>\n",
       "      <td>LYS</td>\n",
       "      <td>K</td>\n",
       "      <td>Lys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Leucine</td>\n",
       "      <td>LEU</td>\n",
       "      <td>L</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Methionine</td>\n",
       "      <td>MET</td>\n",
       "      <td>M</td>\n",
       "      <td>Met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Asparagine</td>\n",
       "      <td>ASN</td>\n",
       "      <td>N</td>\n",
       "      <td>Asn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Proline</td>\n",
       "      <td>PRO</td>\n",
       "      <td>P</td>\n",
       "      <td>Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Glutamine</td>\n",
       "      <td>GLN</td>\n",
       "      <td>Q</td>\n",
       "      <td>Gln</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Arginine</td>\n",
       "      <td>ARG</td>\n",
       "      <td>R</td>\n",
       "      <td>Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Serine</td>\n",
       "      <td>SER</td>\n",
       "      <td>S</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Threonine</td>\n",
       "      <td>THR</td>\n",
       "      <td>T</td>\n",
       "      <td>Thr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Valine</td>\n",
       "      <td>VAL</td>\n",
       "      <td>V</td>\n",
       "      <td>Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Tryptophan</td>\n",
       "      <td>TRP</td>\n",
       "      <td>W</td>\n",
       "      <td>Trp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Tyrosine</td>\n",
       "      <td>TYR</td>\n",
       "      <td>Y</td>\n",
       "      <td>Tyr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             name 3_letter_code 1_letter_code 3_letter\n",
       "0         Alanine           ALA             A      Ala\n",
       "1        Cysteine           CYS             C      Cys\n",
       "2   Aspartic Acid           ASP             D      Asp\n",
       "3   Glutamic Acid           GLU             E      Glu\n",
       "4   Phenylalanine           PHE             F      Phe\n",
       "5         Glycine           GLY             G      Gly\n",
       "6       Histidine           HIS             H      His\n",
       "7      Isoleucine           ILE             I      Ile\n",
       "8          Lysine           LYS             K      Lys\n",
       "9         Leucine           LEU             L      Leu\n",
       "10     Methionine           MET             M      Met\n",
       "11     Asparagine           ASN             N      Asn\n",
       "12        Proline           PRO             P      Pro\n",
       "13      Glutamine           GLN             Q      Gln\n",
       "14       Arginine           ARG             R      Arg\n",
       "15         Serine           SER             S      Ser\n",
       "16      Threonine           THR             T      Thr\n",
       "17         Valine           VAL             V      Val\n",
       "18     Tryptophan           TRP             W      Trp\n",
       "19       Tyrosine           TYR             Y      Tyr"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# checking for compatibility between \n",
    "# current and Aragorn-suggested tRNA annotations\n",
    "# comparing 3-letter and single-letter codes\n",
    "# The amino acid table was dowloaded \n",
    "# from https://thecodingbiologist.com/posts/amino-acid-table\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "aa_csv = 'amino_acid_table.csv'\n",
    "\n",
    "df_aa = pd.read_csv(aa_csv)\n",
    "aa_columns = ['name', '3_letter_code', '1_letter_code']\n",
    "df_aa = df_aa[aa_columns]\n",
    "df_aa['3_letter'] = df_aa['3_letter_code'].str.title()\n",
    "display(df_aa)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "plastome_post-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
