{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d8b4b753",
   "metadata": {},
   "source": [
    "This notebook is for troubleshooting plastomes using NCBI-provided software pipelines. It was a suggestion from NCBI team rejected the annotation twice after uploading."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3d2cf034",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Defining variables for files\n",
    "template = 'plastomes/final/template.sbt'\n",
    "gb = 'plastomes/final/Crepis_callicephala.gb'\n",
    "fasta = 'plastomes/in/Crepis_callicephala.fasta'\n",
    "\n",
    "# command itself:\n",
    "# table2asn -t template.sbt -i sequence.fsa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bfd7c794",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File plastomes/in/Crepis_callicephala.fasta does exist.\n"
     ]
    }
   ],
   "source": [
    "# check for fasta\n",
    "\n",
    "from Bio import SeqIO, SeqRecord\n",
    "\n",
    "try:\n",
    "    with open(fasta, \"r\") as fasta_file:\n",
    "        print(f\"File {fasta} does exist.\")\n",
    "except FileNotFoundError:\n",
    "    print(f\"File '{fasta}' not found.\\nConverting genbank file '{gb}' to FASTA...\")\n",
    "    try:\n",
    "        first_record = next(SeqIO.parse(gb, \"genbank\"))\n",
    "        SeqIO.write(first_record, fasta, \"fasta\")\n",
    "        print(f\"Successfully created '{fasta}'.\")\n",
    "    except FileNotFoundError:\n",
    "        print(f\"Error: The source GenBank file '{gb}' was not found.\")\n",
    "    except StopIteration:\n",
    "        print(f\"Error: The GenBank file '{gb}' is empty.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35dde3bb",
   "metadata": {},
   "source": [
    "## Run `table2asn` program\n",
    "\n",
    "There are some recommendations on NCBI site that were not illuminated at Readme file of the program.\n",
    "`table2asn` will recognize files with **the same basename** as the input sequence file. Sequences that are part of a plasmid, or an organellar chromosome, or specific nuclear chromosomes need to have that information included in the fasta definition line, in these formats:\n",
    "\n",
    "- [location=mitochondrion]\n",
    "- [location=chloroplast]\n",
    "\n",
    "Sequences that are a complete circular chromosome or plasmid need to have the circular topology and the completeness included.\n",
    "\n",
    "- [topology=circular] [completeness=complete]\n",
    "- [topology=circular] gap at end, not circularized\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "04cf3c4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating symbolic links:\n",
      "Symlink for 'plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA.tbl' is already exist.\n"
     ]
    }
   ],
   "source": [
    "# preparing directory\n",
    "# all the files should be stored at the same directory\n",
    "import os\n",
    "\n",
    "project_dir = 'plastomes/table2asn'\n",
    "tbl_file = \"plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA.tbl\"\n",
    "sqn_file = \"plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA(1).sqn\"\n",
    "\n",
    "SHORT_NAMES = {\n",
    "    \"Crepis_callicephala\": \"cc\",\n",
    "    \"Crepis_purpurea\": \"cp\",\n",
    "}\n",
    "\n",
    "def create_symlinks(project_dir: str, sourcefile: str):\n",
    "    \"\"\"\n",
    "    Create symlinks for files to process.\n",
    "    \"\"\"\n",
    "    basename = os.path.basename(sourcefile)\n",
    "    ext= basename.rsplit(\".\")[-1]\n",
    "    #print(\"ext\", ext)\n",
    "\n",
    "    # handle sql and tbl names\n",
    "    basename = str(basename).rsplit(\"___\")[0]\n",
    "\n",
    "    #print(\"basename\", basename)\n",
    "    species = basename.rsplit(\".\")[0]\n",
    "    #print(\"species\", species)\n",
    "    \n",
    "    label = SHORT_NAMES[species]\n",
    "    #print(\"label\", label)\n",
    "    target_dir = os.path.join(project_dir, label)\n",
    "    #print(\"target_dir\", target_dir)\n",
    "    abs_target_dir = os.path.abspath(target_dir)\n",
    "    # filename should match sequence id in FASTA header\n",
    "    target_filename = os.path.join(abs_target_dir, f\"{species}.{ext}\")\n",
    "    \n",
    "    #print(\"target_filename\", target_filename)\n",
    "    abs_sourcefile = os.path.abspath(sourcefile)\n",
    "\n",
    "    try:\n",
    "        os.path.exists(sourcefile)\n",
    "        try:\n",
    "            # fixing first relative paths\n",
    "            os.symlink(abs_sourcefile, target_filename)\n",
    "            print(f\"Symlink '{target_filename}' for '{abs_sourcefile}' was successfully created.\")\n",
    "        except FileExistsError:\n",
    "            print(f\"Symlink for '{sourcefile}' is already exist.\")\n",
    "        except PermissionError:\n",
    "            print(f\"You have no permissions for this action.\")\n",
    "    except FileExistsError:\n",
    "        print(f\"Source file '{sourcefile}' does not exist.\")\n",
    "\n",
    "\n",
    "print(\"Creating symbolic links:\")\n",
    "# create_symlinks(project_dir, gb) #  this is output file\n",
    "# create_symlinks(project_dir, fasta) # the header line should be modified\n",
    "create_symlinks(project_dir, tbl_file)\n",
    "#create_symlinks(project_dir, sqn_file) #  this is output file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7558e3a5",
   "metadata": {},
   "source": [
    "The header line of fasta file should follow definition line conventions. To keep previous files intact but change required data, the file should be just copied instead of linking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a7cca631",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "destination: plastomes/table2asn/cc/Crepis_callicephala.fsa\n",
      "File 'plastomes/in/Crepis_callicephala.fasta' copied to 'plastomes/table2asn/cc' successfully.\n"
     ]
    }
   ],
   "source": [
    "import shutil\n",
    "\n",
    "destination_directory = f\"{project_dir}/cc\"\n",
    "dest_file = f\"{destination_directory}/{os.path.basename(fasta).rsplit(\".\")[0]}.fsa\"\n",
    "print(\"destination:\", dest_file)\n",
    "\n",
    "if not os.path.exists(dest_file):\n",
    "    try:\n",
    "        shutil.copy(fasta, dest_file)\n",
    "        print(f\"File '{fasta}' copied to '{destination_directory}' successfully.\")\n",
    "    except FileNotFoundError:\n",
    "        print(\"Source file not found.\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3bf6b370",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[organism=Crepis callicephala] [location=chloroplast] [topology=circular] [completeness=complete] [gcode=11]\n",
      "Checking input directory...\n",
      "  plastomes/table2asn/cc is present: True\n",
      "Checking input files...\n",
      "  plastomes/final/template.sbt is present: True\n",
      "  Crepis_callicephala.tbl is present: True\n",
      "  Crepis_callicephala.fsa is present: True\n",
      "Running: conda run -n table2asn-env table2asn -indir /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/table2asn/cc -t /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/final/template.sbt -j '[organism=Crepis callicephala] [location=chloroplast] [topology=circular] [completeness=complete] [gcode=11]' -V vb -verbose\n",
      "STDOUT: \n",
      "STDERR: This copy of table2asn is more than 1 year old. Please download the current version if it is newer.\n",
      "Will be using one threads\n",
      "Recognized annotation format: five-column feature table\n",
      "Falling back on built-in data for popular organisms.\n",
      "\n",
      "\n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "# testing table2asn\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "\n",
    "# Configuration\n",
    "CONDA_ENV_NAME = \"plastome_post-env\"\n",
    "INPUT_SUBDIR = \"cc\"\n",
    "\n",
    "location = \"[location=chloroplast]\"\n",
    "topology = \"[topology=circular]\"\n",
    "completeness = \"[completeness=complete]\"\n",
    "organism = \"[organism=Crepis callicephala]\"\n",
    "code = \"[gcode=11]\"\n",
    "pcode = \"[gpcode=11]\"\n",
    "\n",
    "src_qualifiers = f\"{organism} {location} {topology} {completeness} {code}\"\n",
    "print(src_qualifiers)\n",
    "\n",
    "input_dir = f\"{project_dir}/cc\"\n",
    "# Check inputs\n",
    "print(f\"Checking input directory...\\n  {input_dir} is present: {os.path.isdir(input_dir)}\")\n",
    "print(f\"Checking input files...\\n  {template} is present: {os.path.isfile(template)}\")\n",
    "present_flag = False\n",
    "for i in os.listdir(input_dir):\n",
    "    file = os.path.join(input_dir, i)\n",
    "    present_flag = os.path.exists(file)\n",
    "    print(f\"  {i} is present: {present_flag}\")\n",
    "    if present_flag == False:\n",
    "        print(f\"    relative path: {file}\")\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "# table2asn -indir {project_dir}/cc -t {template} -j {comment}\n",
    "\n",
    "cmd = [\n",
    "    \"conda\", \"run\", \"-n\", \"table2asn-env\",\n",
    "    \"table2asn\",\n",
    "    \"-indir\", str(os.path.abspath(input_dir)),\n",
    "    \"-t\", str(os.path.abspath(template)),\n",
    "    \"-j\", f\"'{src_qualifiers}'\",\n",
    "    \"-V\", \"vb\",\n",
    "    \"-verbose\",\n",
    "]\n",
    "\n",
    "cmd_version = [\n",
    "    \"conda\", \"run\", \"-n\", \"table2asn-env\",\n",
    "    \"table2asn\",\n",
    "    \"-version\",\n",
    "]\n",
    "\n",
    "print(f\"Running: {' '.join(cmd)}\")\n",
    "#print(f\"Running: {' '.join(cmd_version)}\")\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70c4e93",
   "metadata": {},
   "source": [
    "The `table2asn` did not performed well. The working directory does not contain any output files.\n",
    "There the \"This copy of table2asn is more than 1 year old. Please download the current version if it is newer.\" message appeared in STDERR. Trying to launch binary executable file from the fresh release might be a solution.\n",
    "\n",
    "## Solution\n",
    "Thwe problem was copying fasta file with `.fasta` file extension. After changing it to `.fsa` the `table2asn` performed well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e3282652",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: table2asn: 1.29.324\n",
      "\n",
      "STDERR: \n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "table2asn = \"bin/table2asn.linux64\"\n",
    "\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "cmd_version = [\n",
    "    table2asn,\n",
    "    \"-version\",\n",
    "]\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd_version, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1e2b054a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: bin/table2asn.linux64 -indir /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/table2asn/cc -t /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/final/template.sbt -V vb -verbose\n",
      "STDOUT: \n",
      "STDERR: \n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "cmd = [\n",
    "    table2asn,\n",
    "    \"-indir\", str(os.path.abspath(input_dir)),\n",
    "    \"-t\", str(os.path.abspath(template)),\n",
    "    \"-V\", \"vb\",\n",
    "    \"-verbose\",\n",
    "]\n",
    "\n",
    "print(f\"Running: {' '.join(cmd)}\")\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "799fe050",
   "metadata": {},
   "source": [
    "## Summarizing results\n",
    "There are 4 errors in Crepis callicephala genome of two types. The reference genome annotations were validated via GB2Sequin and compared to C. callicephala plastid genome.\n",
    "### Errors\n",
    "\n",
    "| species | accession | SEQ_INST.BadProteinStart | SEQ_FEAT.StartCodon |\n",
    "| ------- | --------- | ------------------------ | ------------------- |\n",
    "| Crepis callicephala | - | psbL, ndhD | \"photosystem II subunit L\", \"NADH dehydrogenase subunit D\" |\n",
    "| Lactuca sativa | NC_007578 | psbL, ndhD | - |\n",
    "| Lactuca sativa cv. Ramosa | PP999684 | psbL | - |\n",
    "| Nicotiana tabacum | NC_001879 | psbL | - |\n",
    "| Arabidopsis thaliana | NC_000932 | ndhD | - |\n",
    "| Oryza sativa | NC_031333 | rpl2, rpl2 | - |\n",
    "\n",
    "At least each fatal error should be checked manually. One of the solution for 'illegal start codon' type error might be exception defining in annotation.\n",
    "\n",
    "### Warnings\n",
    "In Crepis callicephala genome validation report\n",
    "\n",
    "| Warning | Feature | Feature_description | Location |\n",
    "| - | - | - | - |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S12 <133> | rps12:[lcl\\|Crepis_callicephala:c67482-67369, 136002-136794] |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | Ycf15 protein <158> | ycf15:lcl\\|Crepis_callicephala:c96959-96768 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S7 <159> | rps7:lcl\\|Crepis_callicephala:c94947-94480 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | NADH dehydrogenase subunit B <160> | ndhB:lcl\\|Crepis_callicephala:c94189-91988 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | Ycf2 protein <161> | ycf2:lcl\\|Crepis_callicephala:84073-90927 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L23 <162> | rpl23:lcl\\|Crepis_callicephala:c83722-83441 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L2 <163> | rpl2:lcl\\|Crepis_callicephala:c83422-81933 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c7316-7244 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c7316-7244 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c8577-8490 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c8577-8490 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c11817-11734 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c11817-11734 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c11992-11921 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c11992-11921 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c29889-29818 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c29889-29818 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c35029-34947 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Gly | lcl\\|Crepis_callicephala:35855-35925 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:44778-44864 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | lcl\\|Crepis_callicephala:c83961-83888 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:110967-111046 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | (lcl\\|Crepis_callicephala:c132067-132025, c131268-131234) |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | lcl\\|Crepis_callicephala:147834-147908 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | lcl\\|Crepis_callicephala:c1794-1759 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | lcl\\|Crepis_callicephala:c4320-1795 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | lcl\\|Crepis_callicephala:c4358-4321 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | lcl\\|Crepis_callicephala:47244-47293 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | lcl\\|Crepis_callicephala:c51727-51690 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | lcl\\|Crepis_callicephala:99728-99770 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | lcl\\|Crepis_callicephala:99771-100526 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | lcl\\|Crepis_callicephala:100527-100561 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S12 <133> | (lcl\\|Crepis_callicephala:c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf15 protein <158> | lcl\\|Crepis_callicephala:134836-135027 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S7 <159> | lcl\\|Crepis_callicephala:136848-137315 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | NADH dehydrogenase subunit B <160> | (lcl\\|Crepis_callicephala:137606-138382, 139052-139807) |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf2 protein <161> | lcl\\|Crepis_callicephala:c147722-140868 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L23 <162> | lcl\\|Crepis_callicephala:148073-148354 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L2 <163> | (lcl\\|Crepis_callicephala:148373-148763, 149429-149862) |\n",
    "\n",
    "### Notes\n",
    "For C. Callicephala, lines of type 'Note' are absent in table2asn validation output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "125bc582",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S12 <133> | rps12:[lcl\\|Crepis_callicephala:c67482-67369, 136002-136794] |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | Ycf15 protein <158> | ycf15:lcl\\|Crepis_callicephala:c96959-96768 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S7 <159> | rps7:lcl\\|Crepis_callicephala:c94947-94480 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | NADH dehydrogenase subunit B <160> | ndhB:lcl\\|Crepis_callicephala:c94189-91988 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | Ycf2 protein <161> | ycf2:lcl\\|Crepis_callicephala:84073-90927 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L23 <162> | rpl23:lcl\\|Crepis_callicephala:c83722-83441 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L2 <163> | rpl2:lcl\\|Crepis_callicephala:c83422-81933 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c7316-7244 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c7316-7244 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c8577-8490 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c8577-8490 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c11817-11734 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c11817-11734 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c11992-11921 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c11992-11921 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | lcl\\|Crepis_callicephala:c29889-29818 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c29889-29818 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:c35029-34947 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Gly | lcl\\|Crepis_callicephala:35855-35925 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:44778-44864 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | lcl\\|Crepis_callicephala:c83961-83888 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | lcl\\|Crepis_callicephala:110967-111046 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | (lcl\\|Crepis_callicephala:c132067-132025, c131268-131234) |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | lcl\\|Crepis_callicephala:147834-147908 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | lcl\\|Crepis_callicephala:c1794-1759 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | lcl\\|Crepis_callicephala:c4320-1795 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | lcl\\|Crepis_callicephala:c4358-4321 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | lcl\\|Crepis_callicephala:47244-47293 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | lcl\\|Crepis_callicephala:c51727-51690 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | lcl\\|Crepis_callicephala:99728-99770 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | lcl\\|Crepis_callicephala:99771-100526 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | lcl\\|Crepis_callicephala:100527-100561 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S12 <133> | (lcl\\|Crepis_callicephala:c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf15 protein <158> | lcl\\|Crepis_callicephala:134836-135027 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S7 <159> | lcl\\|Crepis_callicephala:136848-137315 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | NADH dehydrogenase subunit B <160> | (lcl\\|Crepis_callicephala:137606-138382, 139052-139807) |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf2 protein <161> | lcl\\|Crepis_callicephala:c147722-140868 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L23 <162> | lcl\\|Crepis_callicephala:148073-148354 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L2 <163> | (lcl\\|Crepis_callicephala:148373-148763, 149429-149862) |\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:26: SyntaxWarning: \"\\|\" is an invalid escape sequence. Such sequences will not work in the future. Did you mean \"\\\\|\"? A raw string is also an option.\n",
      "<>:26: SyntaxWarning: \"\\|\" is an invalid escape sequence. Such sequences will not work in the future. Did you mean \"\\\\|\"? A raw string is also an option.\n",
      "/tmp/ipykernel_109528/2619673773.py:26: SyntaxWarning: \"\\|\" is an invalid escape sequence. Such sequences will not work in the future. Did you mean \"\\\\|\"? A raw string is also an option.\n",
      "  variables[i] = value.replace(\"|\", \"\\|\")\n"
     ]
    }
   ],
   "source": [
    "# parsing text file with validation warnings to print it in structured format.\n",
    "\n",
    "file_2_parse = 'plastomes/table2asn/cc/Crepis_callicephala.val'\n",
    "\n",
    "with open(file_2_parse, \"r\") as handle:\n",
    "    for line in handle:\n",
    "        warning = \"-\"\n",
    "        feature = \"-\"\n",
    "        gene = \"-\"\n",
    "        location = \"-\"\n",
    "        if \"Warning:\" in line:\n",
    "            line = line.split(\" \", maxsplit=3)\n",
    "            warning = line[2].strip(\"[]\")\n",
    "            feature = line[-1].rsplit('FEATURE: ')[-1].split(\": \")[0]\n",
    "            gene = line[-1].rsplit('FEATURE: ')[-1].split(\"[\")[0].split(\": \")[-1]\n",
    "            location = line[-1].split(\"[\", maxsplit=1)[1].split(\"] \")[0]\n",
    "            # sanitizing values\n",
    "            variables = [warning, feature, gene, location]\n",
    "            for i, value in enumerate(variables):\n",
    "                value = value.strip(\" \")\n",
    "                if not value or value.isspace():\n",
    "                    variables[i] = \"-\"\n",
    "                elif \"|\" in value:\n",
    "                    variables[i] = value.replace(\"|\", \"\\|\")\n",
    "                else:\n",
    "                    variables[i] = value\n",
    "            warning, feature, gene, location = variables\n",
    "            # printing in md-compatible format\n",
    "            print(f\"| {warning} | {feature} | {gene} | {location} |\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "422b2a04",
   "metadata": {},
   "source": [
    "## Checking tRNA data\n",
    "### Perform re-assess using Aragorn v1.2.41\n",
    "The code to use:\n",
    "```shell\n",
    "aragorn -v -e -gcbact -c -i -o $output $fasta\n",
    "```\n",
    "`-gcbact`   Use Bacterial/Plant chloroplast genetic code.\n",
    "\n",
    "`-c`        Assume that each sequence has a circular topology. Search wraps around each end.\n",
    "\n",
    "`-d`        Double. Search both strands of each sequence. Default setting\n",
    "\n",
    "`-i`        Search for tRNA genes with introns in anticodon loop with maximum length 3000 bases. Minimum intron length is 0 bases.\n",
    "\n",
    "`-e`        Print out score for each reported gene.\n",
    "\n",
    "`-v`        Verbose. Prints out information during search to STDERR.\n",
    "\n",
    "`-o <outfile>`    Print output to <outfile>. If <outfile> already exists, it is overwritten. By default all output goes to stdout."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "099edc7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import sys\n",
    "\n",
    "fasta = \"plastomes/table2asn/cc/Crepis_callicephala.fsa\"\n",
    "gb = \"plastomes/table2asn/cc/Crepis_callicephala.gbf\"\n",
    "aragorn_out = \"plastomes/table2asn/cc/Crepis_callicephala.aragorn.trnas.txt\"\n",
    "\n",
    "aragorn_cmd = [\n",
    "    \"conda\", \"run\", \"-n\", \"aragorn-env\",\n",
    "    \"aragorn\",\n",
    "    \"-v\", \"-e\",\n",
    "    \"-gcbact\", \n",
    "    \"-c\", \"-i\", \n",
    "    \"-o\", aragorn_out, \n",
    "    gb,\n",
    "]\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`ARAGORN` run completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`ARAGORN` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "plastome_post-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
