{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d8b4b753",
   "metadata": {},
   "source": [
    "This notebook is for troubleshooting plastomes using NCBI-provided software pipelines. It was a suggestion from NCBI team rejected the annotation twice after uploading."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d2cf034",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Defining variables for files\n",
    "template = 'plastomes/final/template.sbt'\n",
    "gb = 'plastomes/final/Crepis_callicephala.gb'\n",
    "fasta = 'plastomes/in/Crepis_callicephala.fasta'\n",
    "\n",
    "# command itself:\n",
    "# table2asn -t template.sbt -i sequence.fsa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bfd7c794",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File plastomes/in/Crepis_callicephala.fasta does exist.\n"
     ]
    }
   ],
   "source": [
    "# check for fasta\n",
    "\n",
    "from Bio import SeqIO, SeqRecord\n",
    "\n",
    "try:\n",
    "    with open(fasta, \"r\") as fasta_file:\n",
    "        print(f\"File {fasta} does exist.\")\n",
    "except FileNotFoundError:\n",
    "    print(f\"File '{fasta}' not found.\\nConverting genbank file '{gb}' to FASTA...\")\n",
    "    try:\n",
    "        first_record = next(SeqIO.parse(gb, \"genbank\"))\n",
    "        SeqIO.write(first_record, fasta, \"fasta\")\n",
    "        print(f\"Successfully created '{fasta}'.\")\n",
    "    except FileNotFoundError:\n",
    "        print(f\"Error: The source GenBank file '{gb}' was not found.\")\n",
    "    except StopIteration:\n",
    "        print(f\"Error: The GenBank file '{gb}' is empty.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35dde3bb",
   "metadata": {},
   "source": [
    "## Run `table2asn` program\n",
    "\n",
    "There are some recommendations on NCBI site that were not illuminated at Readme file of the program.\n",
    "`table2asn` will recognize files with **the same basename** as the input sequence file. Sequences that are part of a plasmid, or an organellar chromosome, or specific nuclear chromosomes need to have that information included in the fasta definition line, in these formats:\n",
    "\n",
    "- [location=mitochondrion]\n",
    "- [location=chloroplast]\n",
    "\n",
    "Sequences that are a complete circular chromosome or plasmid need to have the circular topology and the completeness included.\n",
    "\n",
    "- [topology=circular] [completeness=complete]\n",
    "- [topology=circular] gap at end, not circularized\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "04cf3c4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating symbolic links:\n",
      "Symlink for 'plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA.tbl' is already exist.\n"
     ]
    }
   ],
   "source": [
    "# preparing directory\n",
    "# all the files should be stored at the same directory\n",
    "import os\n",
    "\n",
    "project_dir = 'plastomes/table2asn'\n",
    "tbl_file = \"plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA.tbl\"\n",
    "sqn_file = \"plastomes/final/gb2sequin_out/c_callicephala_final/05/Crepis_callicephala___149980_bp____DNA(1).sqn\"\n",
    "\n",
    "SHORT_NAMES = {\n",
    "    \"Crepis_callicephala\": \"cc\",\n",
    "    \"Crepis_purpurea\": \"cp\",\n",
    "}\n",
    "\n",
    "def create_symlinks(project_dir: str, sourcefile: str):\n",
    "    \"\"\"\n",
    "    Create symlinks for files to process.\n",
    "    \"\"\"\n",
    "    basename = os.path.basename(sourcefile)\n",
    "    ext= basename.rsplit(\".\")[-1]\n",
    "    #print(\"ext\", ext)\n",
    "\n",
    "    # handle sql and tbl names\n",
    "    basename = str(basename).rsplit(\"___\")[0]\n",
    "\n",
    "    #print(\"basename\", basename)\n",
    "    species = basename.rsplit(\".\")[0]\n",
    "    #print(\"species\", species)\n",
    "    \n",
    "    label = SHORT_NAMES[species]\n",
    "    #print(\"label\", label)\n",
    "    target_dir = os.path.join(project_dir, label)\n",
    "    #print(\"target_dir\", target_dir)\n",
    "    abs_target_dir = os.path.abspath(target_dir)\n",
    "    # filename should match sequence id in FASTA header\n",
    "    target_filename = os.path.join(abs_target_dir, f\"{species}.{ext}\")\n",
    "    \n",
    "    #print(\"target_filename\", target_filename)\n",
    "    abs_sourcefile = os.path.abspath(sourcefile)\n",
    "\n",
    "    try:\n",
    "        os.path.exists(sourcefile)\n",
    "        try:\n",
    "            # fixing first relative paths\n",
    "            os.symlink(abs_sourcefile, target_filename)\n",
    "            print(f\"Symlink '{target_filename}' for '{abs_sourcefile}' was successfully created.\")\n",
    "        except FileExistsError:\n",
    "            print(f\"Symlink for '{sourcefile}' is already exist.\")\n",
    "        except PermissionError:\n",
    "            print(f\"You have no permissions for this action.\")\n",
    "    except FileExistsError:\n",
    "        print(f\"Source file '{sourcefile}' does not exist.\")\n",
    "\n",
    "\n",
    "print(\"Creating symbolic links:\")\n",
    "# create_symlinks(project_dir, gb) #  this is output file\n",
    "# create_symlinks(project_dir, fasta) # the header line should be modified\n",
    "create_symlinks(project_dir, tbl_file)\n",
    "#create_symlinks(project_dir, sqn_file) #  this is output file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7558e3a5",
   "metadata": {},
   "source": [
    "The header line of fasta file should follow definition line conventions. To keep previous files intact but change required data, the file should be just copied instead of linking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a7cca631",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "destination: plastomes/table2asn/cc/Crepis_callicephala.fsa\n"
     ]
    }
   ],
   "source": [
    "import shutil\n",
    "\n",
    "destination_directory = f\"{project_dir}/cc\"\n",
    "dest_file = f\"{destination_directory}/{os.path.basename(fasta).rsplit(\".\")[0]}.fsa\"\n",
    "print(\"destination:\", dest_file)\n",
    "\n",
    "if not os.path.exists(dest_file):\n",
    "    try:\n",
    "        shutil.copy(fasta, dest_file)\n",
    "        print(f\"File '{fasta}' copied to '{destination_directory}' successfully.\")\n",
    "    except FileNotFoundError:\n",
    "        print(\"Source file not found.\")\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3bf6b370",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[organism=Crepis callicephala] [location=chloroplast] [topology=circular] [completeness=complete] [gcode=11]\n",
      "Checking input directory...\n",
      "  plastomes/table2asn/cc is present: True\n",
      "Checking input files...\n",
      "  plastomes/final/template.sbt is present: True\n",
      "  Crepis_callicephala.tbl is present: True\n",
      "  Crepis_callicephala.fsa is present: True\n",
      "Running: conda run -n table2asn-env table2asn -indir /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/table2asn/cc -t /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/final/template.sbt -j '[organism=Crepis callicephala] [location=chloroplast] [topology=circular] [completeness=complete] [gcode=11]' -V vb -verbose\n",
      "STDOUT: \n",
      "STDERR: This copy of table2asn is more than 1 year old. Please download the current version if it is newer.\n",
      "Will be using one threads\n",
      "Recognized annotation format: five-column feature table\n",
      "Falling back on built-in data for popular organisms.\n",
      "\n",
      "\n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "# testing table2asn\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "\n",
    "# Configuration\n",
    "CONDA_ENV_NAME = \"plastome_post-env\"\n",
    "INPUT_SUBDIR = \"cc\"\n",
    "\n",
    "location = \"[location=chloroplast]\"\n",
    "topology = \"[topology=circular]\"\n",
    "completeness = \"[completeness=complete]\"\n",
    "organism = \"[organism=Crepis callicephala]\"\n",
    "code = \"[gcode=11]\"\n",
    "pcode = \"[gpcode=11]\"\n",
    "\n",
    "src_qualifiers = f\"{organism} {location} {topology} {completeness} {code}\"\n",
    "print(src_qualifiers)\n",
    "\n",
    "input_dir = f\"{project_dir}/cc\"\n",
    "# Check inputs\n",
    "print(f\"Checking input directory...\\n  {input_dir} is present: {os.path.isdir(input_dir)}\")\n",
    "print(f\"Checking input files...\\n  {template} is present: {os.path.isfile(template)}\")\n",
    "present_flag = False\n",
    "for i in os.listdir(input_dir):\n",
    "    file = os.path.join(input_dir, i)\n",
    "    present_flag = os.path.exists(file)\n",
    "    print(f\"  {i} is present: {present_flag}\")\n",
    "    if present_flag == False:\n",
    "        print(f\"    relative path: {file}\")\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "# table2asn -indir {project_dir}/cc -t {template} -j {comment}\n",
    "\n",
    "cmd = [\n",
    "    \"conda\", \"run\", \"-n\", \"table2asn-env\",\n",
    "    \"table2asn\",\n",
    "    \"-indir\", str(os.path.abspath(input_dir)),\n",
    "    \"-t\", str(os.path.abspath(template)),\n",
    "    \"-j\", f\"'{src_qualifiers}'\",\n",
    "    \"-V\", \"vb\",\n",
    "    \"-verbose\",\n",
    "]\n",
    "\n",
    "cmd_version = [\n",
    "    \"conda\", \"run\", \"-n\", \"table2asn-env\",\n",
    "    \"table2asn\",\n",
    "    \"-version\",\n",
    "]\n",
    "\n",
    "print(f\"Running: {' '.join(cmd)}\")\n",
    "#print(f\"Running: {' '.join(cmd_version)}\")\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70c4e93",
   "metadata": {},
   "source": [
    "The `table2asn` did not performed well. The working directory does not contain any output files.\n",
    "There the \"This copy of table2asn is more than 1 year old. Please download the current version if it is newer.\" message appeared in STDERR. Trying to launch binary executable file from the fresh release might be a solution.\n",
    "\n",
    "## Solution\n",
    "Thwe problem was copying fasta file with `.fasta` file extension. After changing it to `.fsa` the `table2asn` performed well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e3282652",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: table2asn: 1.29.324\n",
      "\n",
      "STDERR: \n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "table2asn = \"bin/table2asn.linux64\"\n",
    "\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "cmd_version = [\n",
    "    table2asn,\n",
    "    \"-version\",\n",
    "]\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd_version, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1e2b054a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running: bin/table2asn.linux64 -indir /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/table2asn/cc -t /home/asan/BIO_GENOMICS/notebooks/plastome_postprocessing/plastomes/final/template.sbt -V vb -verbose -Z\n",
      "STDOUT: \n",
      "STDERR: Will be using one threads\n",
      "Recognized annotation format: five-column feature table\n",
      "\n",
      "`table2asn` completed successfully!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "project_dir = 'plastomes/table2asn'\n",
    "input_dir = f\"{project_dir}/cc\"\n",
    "\n",
    "cmd = [\n",
    "    table2asn,\n",
    "    \"-indir\", str(os.path.abspath(input_dir)),\n",
    "    \"-t\", str(os.path.abspath(template)),\n",
    "    \"-V\", \"vb\",\n",
    "    \"-verbose\",\n",
    "    \"-Z\",\n",
    "]\n",
    "\n",
    "print(f\"Running: {' '.join(cmd)}\")\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`table2asn` completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`table2asn` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "799fe050",
   "metadata": {},
   "source": [
    "## Summarizing results\n",
    "There are 4 errors in Crepis callicephala genome of two types. The reference genome annotations were validated via GB2Sequin and compared to C. callicephala plastid genome.\n",
    "### Errors\n",
    "\n",
    "| species | accession | SEQ_INST.BadProteinStart | SEQ_FEAT.StartCodon |\n",
    "| ------- | --------- | ------------------------ | ------------------- |\n",
    "| Crepis callicephala | - | psbL, ndhD | \"photosystem II subunit L\", \"NADH dehydrogenase subunit D\" |\n",
    "| Lactuca sativa | NC_007578 | psbL, ndhD | - |\n",
    "| Lactuca sativa cv. Ramosa | PP999684 | psbL | - |\n",
    "| Nicotiana tabacum | NC_001879 | psbL | - |\n",
    "| Arabidopsis thaliana | NC_000932 | ndhD | - |\n",
    "| Oryza sativa | NC_031333 | rpl2, rpl2 | - |\n",
    "\n",
    "At least each fatal error should be checked manually. One of the solution for 'illegal start codon' type error might be exception defining in annotation.\n",
    "\n",
    "### Warnings\n",
    "In Crepis callicephala genome validation report\n",
    "\n",
    "| Warning | Feature | Feature_description | Location |\n",
    "| - | - | - | - |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S12 <133> | rps12:[c67482-67369, 136002-136794] |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | Ycf15 protein <158> | ycf15:c96959-96768 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S7 <159> | rps7:c94947-94480 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | NADH dehydrogenase subunit B <160> | ndhB:c94189-91988 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | Ycf2 protein <161> | ycf2:84073-90927 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L23 <162> | rpl23:c83722-83441 |\n",
    "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L2 <163> | rpl2:c83422-81933 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c7316-7244 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c7316-7244 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c8577-8490 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c8577-8490 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c11817-11734 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c11817-11734 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c11992-11921 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c11992-11921 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c29889-29818 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c29889-29818 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c35029-34947 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Gly | 35855-35925 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | 44778-44864 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | c83961-83888 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | 110967-111046 |\n",
    "| SEQ_FEAT.MissingTrnaAA | tRNA | - | (c132067-132025, c131268-131234) |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | 147834-147908 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | c1794-1759 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | c4320-1795 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c4320-1795 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | c4358-4321 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c6229-5377 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 16596-17318 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 16596-17318 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 26949-27665 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 26949-27665 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c30852-30143 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=2 | c42897-42165 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c43820-43128 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 46807-47243 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 46807-47243 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | 47244-47293 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c51689-51122 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | c51727-51690 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=2 | c68490-67891 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=2 | c68490-67891 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c69581-68783 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 72584-73348 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c80268-79232 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c83031-82367 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c93412-92744 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c95562-95028 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c95562-95028 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | 99728-99770 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | 99771-100526 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 99771-100526 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 99771-100526 |\n",
    "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | 100527-100561 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 100664-101483 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 100664-101483 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c118053-116999 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c118053-116999 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c131131-130312 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c131131-130312 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c132024-131269 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c132024-131269 |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 136233-136767 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 136233-136767 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 138383-139051 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 148764-149428 |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S16 <5> | (c6269-6230, c5376-5150) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | RNA polymerase subunit beta' <17> | (16165-16595, 17319-18963) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | RNA polymerase subunit beta' <17> | (16165-16595, 17319-18963) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | CF0 subunit I <27> | (26805-26948, 27666-28076) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | CF0 subunit I <27> | (26805-26948, 27666-28076) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | photosystem I assembly factor I <43> | (c43944-43821, c43127-42898, c42164-42012) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | photosystem I assembly factor I <43> | (c43944-43821, c43127-42898, c42164-42012) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | cytochrome b6 <99> | (72578-72583, 73349-73990) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L16 <115> | (c80277-80269, c79231-78833) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L2 <123> | (c83422-83032, c82366-81933) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit B <129> | (c94189-93413, c92743-91988) |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
    "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | NADH dehydrogenase subunit A <152> | (c118605-118054, c116998-116459) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit A <152> | (c118605-118054, c116998-116459) |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf15 protein <158> | 134836-135027 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S7 <159> | 136848-137315 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | NADH dehydrogenase subunit B <160> | (137606-138382, 139052-139807) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit B <160> | (137606-138382, 139052-139807) |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf2 protein <161> | c147722-140868 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L23 <162> | 148073-148354 |\n",
    "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L2 <163> | (148373-148763, 149429-149862) |\n",
    "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L2 <163> | (148373-148763, 149429-149862) |\n",
    "\n",
    "\n",
    "### Notes\n",
    "For C. Callicephala, lines of type 'Note' are absent in table2asn validation output.\n",
    "\n",
    "### Error types\n",
    "#### CDSgeneRange\n",
    "A CDS is overlapped by a gene feature, but is not completely contained by it.\n",
    "This may be an annotation error.\n",
    "\n",
    "#### GeneXrefWithoutGene\n",
    "This feature has a gene xref, but there is no equivalent gene feature anywhere\n",
    "on the record.\n",
    "\n",
    "#### GeneXrefStrandProblem\n",
    "This feature has a gene xref that points to a gene on the wrong strand.\n",
    "\n",
    "#### MissingTrnaAA\n",
    "The tRNA encoded amino acid is not set. *Explanation*: The amino acid that the tRNA carries is not included. *Suggestion*: Include the amino acid as the product of the tRNA. If the amino acid of a tRNA is unknown, use tRNA-Xxx as the product.\n",
    "\n",
    "#### SEQ_FEAT.StartCodon and SEQ_INST.BadProteinStart\n",
    "\n",
    "*Explanation*: The StartCodon and BadProteinStart errors are produced when the CDS is not marked as partial at its 5′ end and does not begin with a start codon.\n",
    "\n",
    "*Suggestion*: Use the correct genetic code to get the correct translations. For example, include \\[gcode=11\\] for prokaryotic genome submissions. Other possible fixes include: extend the CDS to the start codon, or mark the 5′ end as partial (and extend the CDS to the end of the sequence for prokaryotic sequences), or add the /pseudo qualifier to the gene to indicate that the CDS cannot be translated.\n",
    "\n",
    "##### StartCodon\n",
    "An illegal start codon was used. Some possible explanations are: (1) the\n",
    "wrong genetic code may have been selected; (2) the wrong reading frame may\n",
    "be in use; or (3) the coding region may be incomplete at the 5' end, in\n",
    "which case a partial location should be indicated.\n",
    "\n",
    "##### BadProteinStart\n",
    "A gap symbols was found at the start of this protein Bioseq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "125bc582",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S12 <133> | rps12:[c67482-67369, 136002-136794] |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | Ycf15 protein <158> | ycf15:c96959-96768 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein S7 <159> | rps7:c94947-94480 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | NADH dehydrogenase subunit B <160> | ndhB:c94189-91988 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | Ycf2 protein <161> | ycf2:84073-90927 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L23 <162> | rpl23:c83722-83441 |\n",
      "| SEQ_FEAT.CDSgeneRange | CDS | ribosomal protein L2 <163> | rpl2:c83422-81933 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c7316-7244 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c7316-7244 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c8577-8490 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c8577-8490 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c11817-11734 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c11817-11734 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c11992-11921 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c11992-11921 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | - | c29889-29818 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c29889-29818 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | c35029-34947 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Gly | 35855-35925 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | 44778-44864 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | c83961-83888 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | 110967-111046 |\n",
      "| SEQ_FEAT.MissingTrnaAA | tRNA | - | (c132067-132025, c131268-131234) |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | tRNA | Met | 147834-147908 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | c1794-1759 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | c4320-1795 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c4320-1795 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | c4358-4321 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c6229-5377 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 16596-17318 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 16596-17318 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 26949-27665 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 26949-27665 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c30852-30143 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=2 | c42897-42165 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c43820-43128 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 46807-47243 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 46807-47243 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | 47244-47293 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c51689-51122 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | c51727-51690 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=2 | c68490-67891 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=2 | c68490-67891 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c69581-68783 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 72584-73348 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c80268-79232 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c83031-82367 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c93412-92744 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c95562-95028 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c95562-95028 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=1 | 99728-99770 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | intron | /number=1 | 99771-100526 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 99771-100526 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 99771-100526 |\n",
      "| SEQ_FEAT.GeneXrefWithoutGene | exon | /number=2 | 100527-100561 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 100664-101483 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 100664-101483 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c118053-116999 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c118053-116999 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c131131-130312 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c131131-130312 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | c132024-131269 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | c132024-131269 |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | intron | /number=1 | 136233-136767 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 136233-136767 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 138383-139051 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | intron | /number=1 | 148764-149428 |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S16 <5> | (c6269-6230, c5376-5150) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | RNA polymerase subunit beta' <17> | (16165-16595, 17319-18963) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | RNA polymerase subunit beta' <17> | (16165-16595, 17319-18963) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | CF0 subunit I <27> | (26805-26948, 27666-28076) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | CF0 subunit I <27> | (26805-26948, 27666-28076) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | photosystem I assembly factor I <43> | (c43944-43821, c43127-42898, c42164-42012) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | photosystem I assembly factor I <43> | (c43944-43821, c43127-42898, c42164-42012) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ATP-dependent Clp protease proteolytic subunit 1 <89> | (c69652-69582, c68782-68491, c67890-67663) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | cytochrome b6 <99> | (72578-72583, 73349-73990) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L16 <115> | (c80277-80269, c79231-78833) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L2 <123> | (c83422-83032, c82366-81933) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit B <129> | (c94189-93413, c92743-91988) |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein S12 <133> | (c67482-67369, c95793-95563, c95027-95001) |\n",
      "| SEQ_FEAT.NotSpliceConsensusDonor | CDS | NADH dehydrogenase subunit A <152> | (c118605-118054, c116998-116459) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit A <152> | (c118605-118054, c116998-116459) |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf15 protein <158> | 134836-135027 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein S7 <159> | 136848-137315 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | NADH dehydrogenase subunit B <160> | (137606-138382, 139052-139807) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | NADH dehydrogenase subunit B <160> | (137606-138382, 139052-139807) |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | Ycf2 protein <161> | c147722-140868 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L23 <162> | 148073-148354 |\n",
      "| SEQ_FEAT.GeneXrefStrandProblem | CDS | ribosomal protein L2 <163> | (148373-148763, 149429-149862) |\n",
      "| SEQ_FEAT.NotSpliceConsensusAcceptor | CDS | ribosomal protein L2 <163> | (148373-148763, 149429-149862) |\n"
     ]
    }
   ],
   "source": [
    "# parsing text file with validation warnings to print it in structured format.\n",
    "\n",
    "file_2_parse = 'plastomes/table2asn/cc/Crepis_callicephala.val'\n",
    "\n",
    "with open(file_2_parse, \"r\") as handle:\n",
    "    for line in handle:\n",
    "        warning = \"-\"\n",
    "        feature = \"-\"\n",
    "        gene = \"-\"\n",
    "        location = \"-\"\n",
    "        if \"Warning:\" in line and \"SEQ_FEAT\" in line:\n",
    "            line = line.split(\" \", maxsplit=3)\n",
    "            warning = line[2].strip(\"[]\")\n",
    "            feature = line[-1].rsplit('FEATURE: ')[-1].split(\": \")[0]\n",
    "            gene = line[-1].rsplit('FEATURE: ')[-1].split(\"[\")[0].split(\": \")[-1]\n",
    "            #print(line)\n",
    "            location = line[-1].split(\"[\", maxsplit=1)[1].split(\"] \")[0]\n",
    "            # sanitizing values\n",
    "            variables = [warning, feature, gene, location]\n",
    "            for i, value in enumerate(variables):\n",
    "                value = value.strip(\" \")\n",
    "                if not value or value.isspace():\n",
    "                    variables[i] = \"-\"\n",
    "                elif \"|\" in value:\n",
    "                    variables[i] = value.replace(\"lcl|Crepis_callicephala:\", \"\")\n",
    "                else:\n",
    "                    variables[i] = value\n",
    "            warning, feature, gene, location = variables\n",
    "            # printing in md-compatible format\n",
    "            print(f\"| {warning} | {feature} | {gene} | {location} |\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "422b2a04",
   "metadata": {},
   "source": [
    "## Checking tRNA data\n",
    "### Perform re-assess using Aragorn v1.2.41\n",
    "The code to use:\n",
    "```shell\n",
    "aragorn -v -e -gcbact -c -i -o $output $fasta\n",
    "```\n",
    "`-gcbact`   Use Bacterial/Plant chloroplast genetic code.\n",
    "\n",
    "`-c`        Assume that each sequence has a circular topology. Search wraps around each end.\n",
    "\n",
    "`-d`        Double. Search both strands of each sequence. Default setting\n",
    "\n",
    "`-i`        Search for tRNA genes with introns in anticodon loop with maximum length 3000 bases. Minimum intron length is 0 bases.\n",
    "\n",
    "`-e`        Print out score for each reported gene.\n",
    "\n",
    "`-v`        Verbose. Prints out information during search to STDERR.\n",
    "\n",
    "`-o <outfile>`    Print output to <outfile>. If <outfile> already exists, it is overwritten. By default all output goes to stdout."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "099edc7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: \n",
      "STDERR: Crepis_callicephala ' '.\n",
      "149980 nucleotides in sequence\n",
      "Mean G+C content = 37.7%\n",
      "Using bacterial/plant chloroplast genetic code\n",
      "Searching from 3120 to 16881\n",
      "tRNA-Cys(gca) at [9331,9402] (107.847)\n",
      "tRNA-Cys(gca) [9331,9402] (116.002) replacing tRNA-Cys(gca) [9331,9402] (107.847)\n",
      "tRNA-Glu(ttc) at c[11921,11992] (104.658)\n",
      "tRNA-Glu(ttc) c[11920,11993] (115.411) replacing tRNA-Glu(ttc) c[11921,11992] (104.658)\n",
      "tRNA-Tyr(gta) at c[11734,11817] (105.503)\n",
      "tRNA-Tyr(gta) c[11732,11818] (117.555) replacing tRNA-Tyr(gta) c[11734,11817] (105.503)\n",
      "tRNA-Asp(gtc) at c[11547,11620] (108.811)\n",
      "tRNA-Asp(gtc) c[11545,11621] (117.885) replacing tRNA-Asp(gtc) c[11547,11620] (108.811)\n",
      "tRNA-Ser(gct) at c[8491,8577] (104.217)\n",
      "tRNA-Ser(gct) c[8491,8578] (116.445) replacing tRNA-Ser(gct) c[8491,8577] (104.217)\n",
      "tRNA-Ser(gct) c[8490,8577] (123.247) replacing tRNA-Ser(gct) c[8491,8578] (116.445)\n",
      "tRNA-Gln(ttg) at c[7245,7316] (105.114)\n",
      "tRNA-Gln(ttg) c[7245,7316] (105.602) replacing tRNA-Gln(ttg) c[7245,7316] (105.114)\n",
      "tRNA-Gln(ttg) c[7245,7316] (116.786) replacing tRNA-Gln(ttg) c[7245,7316] (105.602)\n",
      "Searching from 16882 to 30643\n",
      "tRNA-Thr(cgt) at c[30094,30875] (101.165)\n",
      "tRNA-Ser(cga) c[30094,30875] (101.783) replacing tRNA-Thr(cgt) c[30094,30875] (101.165)\n",
      "tRNA-Arg(tct) at c[29818,29889] (108.377)\n",
      "tRNA-Arg(tct) c[29818,29889] (117.452) replacing tRNA-Arg(tct) c[29818,29889] (108.377)\n",
      "Searching from 30644 to 44405\n",
      "tRNA-Gly(gcc) at [35856,35927] (107.727)\n",
      "tRNA-Gly(gcc) [35855,35926] (114.476) replacing tRNA-Gly(gcc) [35856,35927] (107.727)\n",
      "tRNA-Met(cat) at c[36113,36184] (111.245)\n",
      "tRNA-Met(cat) c[36112,36185] (118.102) replacing tRNA-Met(cat) c[36113,36184] (111.245)\n",
      "tRNA-Ser(tga) at c[34948,35029] (102.014)\n",
      "tRNA-Ser(tga) c[34948,35030] (110.345) replacing tRNA-Ser(tga) c[34948,35029] (102.014)\n",
      "tRNA-Met(cat) at c[34947,36184] (102.315)\n",
      "tRNA-Ser(tga) c[34947,35029] (121.044) replacing tRNA-Ser(tga) c[34948,35030] (110.345)\n",
      "Searching from 44406 to 58167\n",
      "tRNA-Ser(gga) at [44778,44863] (107.361)\n",
      "tRNA-Ser(gga) [44777,44863] (116.992) replacing tRNA-Ser(gga) [44778,44863] (107.361)\n",
      "tRNA-Ala(agc) at [44139,44864] (103.107)\n",
      "tRNA-Ser(gga) [44778,44864] (121.845) replacing tRNA-Ser(gga) [44777,44863] (116.992)\n",
      "tRNA-Leu(taa) at [46770,47293] (100.172)\n",
      "tRNA-Leu(taa) [46769,47293] (105.152) replacing tRNA-Leu(taa) [46770,47293] (100.172)\n",
      "tRNA-Leu(taa) [46770,47293] (111.808) replacing tRNA-Leu(taa) [46769,47293] (105.152)\n",
      "tRNA-Phe(gaa) at [47609,47682] (106.321)\n",
      "tRNA-Phe(gaa) [47609,47682] (115.774) replacing tRNA-Phe(gaa) [47609,47682] (106.321)\n",
      "tRNA-Met(cat) at [51914,51986] (111.705)\n",
      "tRNA-Ile(aat) at c[51085,51727] (100.151)\n",
      "tRNA-Val(tac) c[51085,51727] (103.348) replacing tRNA-Ile(aat) c[51085,51727] (100.151)\n",
      "tRNA-Ile(aat) c[51085,51727] (107.89) replacing tRNA-Val(tac) c[51085,51727] (103.348)\n",
      "tRNA-Val(tac) c[51085,51727] (111.087) replacing tRNA-Ile(aat) c[51085,51727] (107.89)\n",
      "tRNA-Thr(tgt) at c[46146,46218] (108.306)\n",
      "tRNA-Glu(ctc) at c[46146,47213] (100.298)\n",
      "tRNA-Thr(tgt) c[46145,46219] (122.683) replacing tRNA-Thr(tgt) c[46146,46218] (108.306)\n",
      "Searching from 58168 to 71929\n",
      "tRNA-Pro(tgg) at c[64379,64454] (105.888)\n",
      "tRNA-Pro(tgg) c[64379,64454] (119.886) replacing tRNA-Pro(tgg) c[64379,64454] (105.888)\n",
      "tRNA-Trp(cca) at c[64142,64215] (106.244)\n",
      "tRNA-Trp(cca) c[64142,64215] (116.997) replacing tRNA-Trp(cca) c[64142,64215] (106.244)\n",
      "Searching from 71930 to 85691\n",
      "tRNA-Met(cat) at c[83887,83961] (102.061)\n",
      "tRNA-Met(cat) c[83887,83961] (108.809) replacing tRNA-Met(cat) c[83887,83961] (102.061)\n",
      "Searching from 85692 to 99453\n",
      "tRNA-Val(gac) at [97644,97715] (113.371)\n",
      "tRNA-Leu(caa) at c[91330,91410] (109.768)\n",
      "tRNA-Leu(caa) c[91329,91411] (118.842) replacing tRNA-Leu(caa) c[91330,91410] (109.768)\n",
      "Searching from 99454 to 113215\n",
      "tRNA-Thr(cgt) at [99728,100561] (101.702)\n",
      "tRNA-Glu(ttc) [99728,100561] (104.79) replacing tRNA-Thr(cgt) [99728,100561] (101.702)\n",
      "tRNA-Arg(tcg) [99728,100561] (106.389) replacing tRNA-Glu(ttc) [99728,100561] (104.79)\n",
      "tRNA-Ile(gat) [99728,100561] (108.501) replacing tRNA-Arg(tcg) [99728,100561] (106.389)\n",
      "tRNA-Thr(cgt) [99728,100561] (109.821) replacing tRNA-Ile(gat) [99728,100561] (108.501)\n",
      "tRNA-Glu(ttc) [99728,100561] (112.908) replacing tRNA-Thr(cgt) [99728,100561] (109.821)\n",
      "tRNA-Ala(tgc) at [100626,101518] (101.592)\n",
      "tRNA-Ala(tgc) [100626,101518] (101.595) replacing tRNA-Ala(tgc) [100626,101518] (101.592)\n",
      "tRNA-Leu(gag) [100625,101518] (101.613) replacing tRNA-Ala(tgc) [100626,101518] (101.595)\n",
      "tRNA-Ala(tgc) [100625,101518] (104.353) replacing tRNA-Leu(gag) [100625,101518] (101.613)\n",
      "tRNA-Ala(tgc) [100625,101518] (104.355) replacing tRNA-Ala(tgc) [100625,101518] (104.353)\n",
      "tRNA-Leu(gag) [100626,101518] (108.269) replacing tRNA-Ala(tgc) [100625,101518] (104.355)\n",
      "tRNA-Ala(tgc) [100626,101518] (111.009) replacing tRNA-Leu(gag) [100626,101518] (108.269)\n",
      "tRNA-Ala(tgc) [100626,101518] (111.011) replacing tRNA-Ala(tgc) [100626,101518] (111.009)\n",
      "tRNA-Arg(acg) at [105301,105375] (100.692)\n",
      "tRNA-Arg(acg) [105301,105375] (115.882) replacing tRNA-Arg(acg) [105301,105375] (100.692)\n",
      "tRNA-Leu(tag) at [110967,111046] (107.517)\n",
      "tRNA-Leu(tag) [110967,111046] (120.595) replacing tRNA-Leu(tag) [110967,111046] (107.517)\n",
      "tRNA-Asn(gtt) at c[105869,105940] (111.726)\n",
      "tRNA-Asn(gtt) c[105868,105941] (123.398) replacing tRNA-Asn(gtt) c[105869,105940] (111.726)\n",
      "Searching from 113216 to 126977\n",
      "tRNA-Asn(gtt) at [125855,125926] (111.726)\n",
      "tRNA-Asn(gtt) [125854,125927] (123.398) replacing tRNA-Asn(gtt) [125855,125926] (111.726)\n",
      "tRNA-Arg(acg) at c[126420,126494] (100.692)\n",
      "tRNA-Arg(acg) c[126420,126494] (115.882) replacing tRNA-Arg(acg) c[126420,126494] (100.692)\n",
      "Searching from 126978 to 140739\n",
      "tRNA-Leu(caa) at [140385,140465] (109.768)\n",
      "tRNA-Leu(caa) [140384,140466] (118.842) replacing tRNA-Leu(caa) [140385,140465] (109.768)\n",
      "tRNA-Val(gac) at c[134080,134151] (113.371)\n",
      "tRNA-Thr(cgt) at c[131234,132067] (101.702)\n",
      "tRNA-Glu(ttc) c[131234,132067] (104.79) replacing tRNA-Thr(cgt) c[131234,132067] (101.702)\n",
      "tRNA-Arg(tcg) c[131234,132067] (106.389) replacing tRNA-Glu(ttc) c[131234,132067] (104.79)\n",
      "tRNA-Ile(gat) c[131234,132067] (108.501) replacing tRNA-Arg(tcg) c[131234,132067] (106.389)\n",
      "tRNA-Thr(cgt) c[131234,132067] (109.821) replacing tRNA-Ile(gat) c[131234,132067] (108.501)\n",
      "tRNA-Glu(ttc) c[131234,132067] (112.908) replacing tRNA-Thr(cgt) c[131234,132067] (109.821)\n",
      "tRNA-Ala(tgc) at c[130277,131169] (101.592)\n",
      "tRNA-Ala(tgc) c[130277,131169] (101.595) replacing tRNA-Ala(tgc) c[130277,131169] (101.592)\n",
      "tRNA-Leu(gag) c[130277,131170] (101.613) replacing tRNA-Ala(tgc) c[130277,131169] (101.595)\n",
      "tRNA-Ala(tgc) c[130277,131170] (104.353) replacing tRNA-Leu(gag) c[130277,131170] (101.613)\n",
      "tRNA-Ala(tgc) c[130277,131170] (104.355) replacing tRNA-Ala(tgc) c[130277,131170] (104.353)\n",
      "tRNA-Leu(gag) c[130277,131169] (108.269) replacing tRNA-Ala(tgc) c[130277,131170] (104.355)\n",
      "tRNA-Ala(tgc) c[130277,131169] (111.009) replacing tRNA-Leu(gag) c[130277,131169] (108.269)\n",
      "tRNA-Ala(tgc) c[130277,131169] (111.011) replacing tRNA-Ala(tgc) c[130277,131169] (111.009)\n",
      "Searching from 140740 to 149980\n",
      "Searching from 1 to 3119\n",
      "tRNA-Met(cat) at [147834,147908] (102.061)\n",
      "tRNA-Met(cat) [147834,147908] (108.809) replacing tRNA-Met(cat) [147834,147908] (102.061)\n",
      "tRNA-Lys(ttt) at c[1759,4358] (103.307)\n",
      "tRNA-His(gtg) at c[2,76] (105.78)\n",
      "tRNA-His(gtg) c[3,77] (115.233) replacing tRNA-His(gtg) c[2,76] (105.78)\n",
      "Removing tRNA-Leu(aag) at c[34947,36184] (102.315)\n",
      "Removing tRNA-Arg(tct) at [44139,44864] (103.107)\n",
      "Removing tRNA-Lys(ttt) at c[46146,47213] (100.298)\n",
      "Crepis_callicephala ' '.\n",
      "Search Finished\n",
      "\n",
      "\n",
      "\n",
      "`ARAGORN` run completed successfully!\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import sys\n",
    "\n",
    "fasta = \"plastomes/table2asn/cc/Crepis_callicephala.fsa\"\n",
    "gb = \"plastomes/table2asn/cc/Crepis_callicephala.gbf\"\n",
    "aragorn_out = \"plastomes/table2asn/cc/Crepis_callicephala.aragorn.trnas.txt\"\n",
    "\n",
    "aragorn_cmd = [\n",
    "    \"conda\", \"run\", \"-n\", \"aragorn-env\",\n",
    "    \"aragorn\",\n",
    "    \"-v\", \"-e\",\n",
    "    \"-gcbact\", \n",
    "    \"-c\", \"-i\", \n",
    "    \"-o\", aragorn_out, \n",
    "    gb,\n",
    "]\n",
    "\n",
    "try:\n",
    "    result = subprocess.run(aragorn_cmd, check=True, capture_output=True, text=True)\n",
    "    print(\"STDOUT:\", result.stdout, flush=True)\n",
    "    print(\"STDERR:\", result.stderr, flush=True)\n",
    "    print(\"`ARAGORN` run completed successfully!\", flush=True)\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"`ARAGORN` failed with exit code {e.returncode}\")\n",
    "    sys.exit(e.returncode)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "472661fe",
   "metadata": {},
   "source": [
    "### Assessing tRNAs using tRNAscan-SE v2.0.12\n",
    "The command to run:\n",
    "```shell\n",
    "tRNAscan-SE -B -o $output -m $trnascan_stat --detail --log $logfile -v --thread 4 $fasta\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "plastome_post-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
